---
title: "`r params$doc_title`"
author: | 
        | christoph.reich@med.uni-heidelberg.de 
        | 
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
params:
  date: !r Sys.Date()
  doc_title: "MSc Default Title"
output: 
  html_document:
    code_folding: hide
    theme: paper
    toc: true
    number_sections: true
    toc_float: true
    toc_depth: 4
    df_print: kable
    latex_engine: xelatex
bibliography: ./references/references.bib
editor_options: 
  chunk_output_type: inline
---


```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = FALSE)
figurespath <- paste0("./reports/", format(Sys.time(), "%Y%m%d"), '_figures/')
knitr::opts_chunk$set(fig.path = figurespath)  # reference is script! also creates new dirs

SAVE.files <- FALSE

library(dplyr)
library(purrr)
library(tidyr)
library(stringr)
library(tibble)
library(ggplot2)
library(ggdist)
library(gghalves)
library(ggthemes)
library(ggpubr)
library(gt)
library(scales)
#library(biomaRt)  # different script
library(kableExtra)
library(MendelianRandomization)
library(TwoSampleMR)
library(mr.raps)
library(MRMix)
library(MRPRESSO)
library(penalized)
# library(rsimsum)   do not use
library(conflicted)
source("helper/MR_lasso.R")  # Slob&Burgess 2020
source("helper/PVE_calculation_fns.R") 
source("helper/plotfigure01fn.R") 
conflict_prefer("select", "dplyr")
conflict_prefer("filter", "dplyr")
conflicts_prefer(kableExtra::kable)
# conflict_prefer("kable", "kableExtra")
```

# Discussion

-   NA

# Version Update

The new report version includes: 

-   All genetic associations with BMI > 0 (for association statistics < 0, sign of the association statistic with GSD changed respectively)


```{r load-data}
my_data <- as_tibble(read.delim("./data-boekstegers_felix/GST_BMI_association_results.txt"))
# GBC: new data 2022-11-04
GBC_dat <- as_tibble(read.delim("./data-boekstegers_felix/C23C24_BMI_association_results.txt"))

# FTO data
FTO_dat <- as_tibble(read.delim("./data-boekstegers_felix/2023-01-19_FTO/FTO_association_results.txt"))
```

# Data Harmonization

We start with harmonizing our data sets for the exposure 

```{r data-harmonization, message=FALSE}
##exposure_dat:
#SNP
#beta.exposure
#se.exposure
#effect_allele.exposure
#other_allele.exposure
#eaf.exposure
exposure_dat <- my_data %>% 
  select(dummyID, SNP, beta_exp, se_exp, a1, a2, eaf_exp, exposure) %>% 
  rename(
    id.exposure = dummyID,
    beta.exposure = beta_exp,
    se.exposure = se_exp,
    effect_allele.exposure = a1,
    other_allele.exposure = a2, 
    eaf.exposure = eaf_exp, 
    exposure = exposure
  )
##outcome_dat:
#SNP
#beta.outcome
#se.outcome
#effect_allele.outcome
#other_allele.outcome
#eaf.outcome
#outcome
outcome_dat <- GBC_dat %>% 
  select(dummyID, SNP, beta_out, se_out, a1, a2, eaf_out, outcome) %>% 
  rename(
    id.outcome = dummyID, 
    beta.outcome = beta_out,
    se.outcome = se_out,
    effect_allele.outcome = a1,
    other_allele.outcome = a2, 
    eaf.outcome = eaf_out,
    outcome = outcome
  )

harmonised_dat <- TwoSampleMR::harmonise_data(
  exposure_dat = exposure_dat, 
  outcome_dat = outcome_dat, 
  action = 2  # default
  ) # # used later for analysis with the {TwoSampleMR} package

sum_SNPs_remove <- sum(harmonised_dat$remove)
sum_SNPs_palindromic <- sum(harmonised_dat$palindromic)
sum_SNPs_ambiguous <- sum(harmonised_dat$ambiguous)
sum_SNPs_remove <- nrow(harmonised_dat)-sum(harmonised_dat$mr_keep)
```


```{r combine-dat}
my_data_harm <- 
  harmonised_dat %>% 
  rename(
    exposureBMI = exposure, 
    outcomeGBC = outcome,
    # estimates exposure BMI
    eaf.bmi=eaf.exposure,
    beta.bmi=beta.exposure,
    se.bmi=se.exposure, 
    
    # estimates outcome GBC
    eaf.gbc=eaf.outcome,
    beta.gbc=beta.outcome,  
    se.gbc=se.outcome,
  ) %>% 
  left_join(my_data %>% select(SNP, p_value_exp, outcome, p_value_out, eaf_out, beta_out, se_out),by = c("SNP"="SNP")) %>% 
  rename(
    # BMI only need to add p-val
    p.value.bmi = p_value_exp,
    
    # GSD rename to "confounder"
    confounderGSD = outcome, 
    eaf.gsd = eaf_out,
    beta.gsd = beta_out, 
    se.gsd = se_out, 
    p.value.gsd = p_value_out,
  ) %>% 
  select(
    SNP, 
    # DROP:   "effect_allele.exposure" "other_allele.exposure"  "effect_allele.outcome"  "other_allele.outcome"
    
    # EXPOSURE
    eaf.bmi,
    beta.bmi,
    se.bmi,
    p.value.bmi,
    
    # CONFOUNDER
    eaf.gsd,
    beta.gsd,
    se.gsd,
    p.value.gsd,
    # OUTCOME
    eaf.gbc,
    beta.gbc,
    se.gbc,
    # pvalue missing and needs to be added from GBC_dat
    # harmonize decision
    palindromic, ambiguous, mr_keep
  ) %>% 
  left_join(GBC_dat %>% select(SNP, p_value_out), by=c("SNP"="SNP")) %>% 
  rename(
    p.value.gbc = p_value_out
  ) %>% 
  relocate(
    p.value.gbc, .before = palindromic
  ) %>% 
  mutate(
    p.value.gsd = ifelse(p.value.gsd == "<.0001",  # convert from string
                         yes = "0.0001", 
                         no=p.value.gsd), 
    p.value.gsd = as.numeric(p.value.gsd)) %>% 
  as_tibble()
```


After harmonizing our data sets we remove `r sum_SNPs_remove` SNPs since they are ambiguous. 

```{r rm-ambigous-SNPs}
dropped_data <- my_data_harm %>% 
  filter(mr_keep==FALSE)

my_data_harm <- my_data_harm %>% 
  filter(mr_keep==TRUE | SNP == "rs1558902") %>%  # also include FTO SNP 
  # change sign of beta after checking consistency of data (see mail felix 2023-01-19)
  mutate(beta.bmi = ifelse(SNP == "rs1558902", -beta.bmi, beta.bmi)) 

my_data_harm <- my_data_harm %>% 
  mutate(
    beta.gsd = ifelse(beta.bmi<0, -beta.gsd, beta.gsd),
    beta.gbc = ifelse(beta.bmi<0, -beta.gbc, beta.gbc),
    # change last beta.bmi, otherwise not negative any more ;)
    beta.bmi = ifelse(beta.bmi<0, -beta.bmi, beta.bmi)
  )
```


## Get gene names from rs-ids

```{r highlight-fto-df}
# see script biomart_search.R
rsIds_GeneNames <- as_tibble(readRDS("./data/biomartBMIsearch.rds"))

filter_FTO <- rsIds_GeneNames %>% 
  filter(str_detect(associated_gene, pattern = "FTO") | str_detect(associated_gene, pattern = "MC4R")) %>% 
  distinct(refsnp_id, .keep_all = TRUE)

t(filter_FTO) %>%  kableExtra::kable() %>% 
  kable_styling(font_size=10)
```

After data harmonization we drop `r sum_SNPs_remove` IVs one of which is the SNP associated with the FTO gene. 

\clearpage

# Simulating on empirical data

**Calculate approximate SEs for decreasing sample sizes and corresponding p-vals**

-   se, pval and study size depend on each other, within each study
-   selection of IVs based on exposure study (BMI in our case)
-   Summary statistics on genetic associations with BMI based on results of the GIANT + GERA studies [@Hoffmann_2018] with $N=100,418 + 234,069 =334487$ and $SNPs=289$

**Steps**

1.  Calculate more accurate SEs using the betas and Pvals
2.  Calculate the approximate SEs for decreasing samples sizes
3.  Calculate the probability values considering the estimated betas and the corresponding SEs
4.  Exclude the IV if Pval $\ge$ genome-wide significance threshold (5 × 10−6)
5.  Calculate cumulative explained variances
6.  Represent figure similar to Fig1 in Chatterjee's paper @QiChatterjee_2021


```{r z-statistic}
## LOOP simulation ----------------------------------------------
n.orig <- 334487
n <- c(1000, 2000, 5000, seq(10000, 330000, by=10000), 334487)

# create vectors to store pvals_n in
colnames_pval <- c()
for (i in 1:length(n)) {
  colnames_pval[i] <- paste0("p_value_exp_sample_", format(n[i], scientific = FALSE))
}
## initialize
se_update <- c()
pvals_df <- data.frame(matrix(ncol = length(colnames_pval), nrow =  nrow(my_data_harm)))
colnames(pvals_df) <- colnames_pval
for (i in 1:nrow(my_data_harm)) {
  # 1) Calculate more accurates SEs using the provided betas and Pvals 
  se_update[i] <- my_data_harm[["beta.bmi"]][i]/-qnorm(p=(my_data_harm[["p.value.bmi"]][i]/2),mean=0)
  # 2) Calculate the approximate SEs for decreasing samples sizes
  ### approximation of sd (single value)
  sd <- se_update[i] * sqrt(n.orig)
  #### approximation of standard error simulating different sample size
  se_n <- sd/(sqrt(n))
  # 3) Calculate the probability values considering the estimated betas and the corresponding SEs
  for (j in 1:length(n)) {
    pvals_n <- 2*(1-pnorm(abs(my_data_harm[["beta.bmi"]][i])/abs(se_n[j]), mean=0, 
                          lower.tail = TRUE))
    pvals_df[i, j] <- pvals_n
  }
}

# 4. Exclude the IV if Pval > genome-wide significance threshold  (5 × 10−8) --> updated to 5 × 10−6
pvals_df %>% 
  summarize(across(.cols = everything(), function(x) sum(x < 5*10^(-6)))) %>% 
  t() %>% 
  cbind(n) %>% 
  base::as.data.frame() %>% 
  tibble::rownames_to_column() %>% 
  rename(sample_size = n, No_of_IVs=V1) %>% 
  select(sample_size, No_of_IVs) %>% 
  kable(caption = "Count of significant SNPs depending on sample size") %>% 
  kable_styling(font_size=10)

# pvals_df %>%
#   filter(p_value_exp_sample_100000< 5*10^(-6)) %>% 
#   count()
my_data_harm <- as_tibble(cbind(my_data_harm, pvals_df))
```


## Calculate proportion of variance in phenotype explained by a given SNP (PVE)

-   The variance in BMI explained by the IVs is calculated with the following formula: $explained\ variance = 2f(1-f)\beta^2$, where $f$ denotes the allele frequency and $\beta$ is the additive genetic effect.
-   The explained variance in liability to GSD and GBC is calculated as described by @So_2011
    -   Prevalence of disease $K$ must be set:
        -   [Prevalence GSD](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3343155/#:~:text=Gallstones%20are%20common.,develop%20stones%20at%20some%20time.&text=The%20majority%20will%20not%20develop,cholecystitis,%20cholangitis,%20or%20pancreatitis)
            - @Aerts_2003
        -   [Prevalence GBC](https://www.wcrf.org/cancer-trends/gallbladder-cancer-statistics/) 
            - Update prevalence $K=0.05$ from [Cancer Today](https://gco.iarc.fr/today/home)

We added a second formula to calculate the explained variance for numeric traits [@Teslovich_2010]. 

```{r calculate-explained-variance}
# outsourced to "./scripts/PVE_calculation_fns.R"

## CALCULATE explained variance for total sample size
my_data_harm <- my_data_harm %>%
  mutate(
    explained_variance_BMI = explained_variance_numeric(eaf = eaf.bmi, beta = beta.bmi),
    # USE THIS !
    explained_variance_BMI2 = explained_variance_numeric2(
      maf = eaf.bmi, beta = beta.bmi,
      se_beta = se.bmi,
      samplesize = n[length(n)]
    ),
    explained_variance_GSD = explained_variance_binary(
      PA = eaf.gsd,
      RR1 = exp(beta.gsd),
      RR2 = 1 + (2 * (exp(beta.gsd) - 1)),
      K = 0.1 # assumed prevalence in population
    )$Vg,
    explained_variance_GBC = explained_variance_binary(
      PA = eaf.gbc,
      RR1 = exp(beta.gbc),
      RR2 = 1 + (2 * (exp(beta.gbc) - 1)), # exp(beta.gbc)^2,
      K = 1.2 / 100000 # assumed prevalence in population
    )$Vg
  )

# sum(my_data_harm$explained_variance_GBC)
```


### %var(X) BMI formula comparison

```{r compare-formulas-numeric-pve}
ggplot(data=my_data_harm, aes(x=explained_variance_BMI, y=explained_variance_BMI2)) +
  geom_point(alpha=0.6)+
  geom_abline()+
  theme_few()
```


## Plot explained variance in BMI vs -log10 P-value from Hoffmann et al.

```{r PVE_BMI_vs_pval, warning=FALSE}
# labeling (only FTO label)
my_data_harm$gene_label <- NA
my_data_harm$gene_label[my_data_harm$SNP %in% filter_FTO$refsnp_id] <- "FTO/MC4R"
my_data_harm$gene_label[my_data_harm$SNP == "rs1558902"] <- "FTO"
my_data_harm$gene_label[my_data_harm$SNP == "rs6567160"] <- "MC4R"

# color labeling
my_data_harm$fto_col_annot <- "No FTO/MC4R variant"
my_data_harm$fto_col_annot[my_data_harm$SNP %in% filter_FTO$refsnp_id] <- "FTO/MC4R variant"

ggplot(data = my_data_harm, 
       mapping = aes(x=explained_variance_BMI*100, y= -log10(p.value.bmi),  
                    col=fto_col_annot, label=gene_label)
       )+
  geom_point(alpha=0.6)+
  ggrepel::geom_text_repel(show.legend = FALSE)+
  xlab("Explained variance in BMI in %")+
  ylab("-log10 P-val")+
  labs(title = "Explained variance exposure",
       col='SNP'
       ) +
  theme_few()+
  scale_colour_few() -> fig00
# fig00   # we use the sample size dependent formula

# second formula variance explained2
ggplot(data = my_data_harm, 
       mapping = aes(x=explained_variance_BMI2*100, y= -log10(p.value.bmi),  
                    col=fto_col_annot, label=gene_label)
       )+
  geom_point(alpha=0.6)+
  ggrepel::geom_text_repel(show.legend = FALSE)+
  xlab("Explained variance in BMI in %")+
  ylab("-log10 P-val")+
  labs(title = "Explained variance exposure",
       col='SNP'
       ) +
  theme_few()+
  scale_colour_few() -> fig00_b
fig00_b

# filenamefig00_a <- paste0("../output/plots/", Sys.Date(), "_explained_var_X_old_formula.png")
# ggsave(filename = filenamefig00_a, plot = fig00, 
#        width = 10, height = 6, 
#        units = "in"  # default
#        )


if (SAVE.files ==TRUE ){
  filenamefig00_b <- paste0("./output/plots/pve/", Sys.Date(), "_explained_var_X_new_sample_size_dep.svg")
  ggsave(filename = filenamefig00_b, plot = fig00_b, 
       width = 10, height = 6, 
       units = "in"  # default
       )
}


```

## Plot Figure 1 from Qi&Chatterjee 2021

```{r get-data-for-fig1, warning=FALSE}
# 4.       Exclude the IV if Pval > genome-wide significance threshold  (5 × 10−8) **NEW 5x10^(-6))**
# 5.       Calculate cumulative explained variances
# initialize df
df_cum_expl_var <- tibble(n=n, 
                          cum_expl_varX = rep(NA, length(n)), 
                          cum_expl_varX2 = rep(NA, length(n)), 
                          cum_expl_varU = rep(NA, length(n)), 
                          cum_expl_varY = rep(NA, length(n)),
                          count_IVs = rep(NA, length(n))
)

threshold <- 5*10^(-6)
for (i in 1:length(colnames_pval) ) {
  select_col <- colnames_pval[i]
  # calculate cum explained variance and count of IVs
  summarydat <- my_data_harm %>% 
    filter(!!sym(select_col) < threshold) %>% 
    select(explained_variance_BMI, explained_variance_BMI2, explained_variance_GSD, explained_variance_GBC) %>% 
    # cumulative PVE
    summarise(expl_variances =across(.cols = everything(), function(x) sum(x, na.rm = TRUE)),
              count_IVs = n())
  # fill up df
  df_cum_expl_var[i, 2] <- summarydat$expl_variances[1]
  df_cum_expl_var[i, 3] <- summarydat$expl_variances[2]
  df_cum_expl_var[i, 4] <- summarydat$expl_variances[3]
  df_cum_expl_var[i, 5] <- summarydat$expl_variances[4]
  df_cum_expl_var[i, 6] <- summarydat$count_IVs
}

# for the numeric traits we have to map over sample size and recalculate se_updated
explained_variance_BMI2_sample_size <- map(n, ~explained_variance_numeric2(maf = my_data_harm$eaf.bmi, beta = my_data_harm$beta.bmi, 
                                                                           se_beta = abs(my_data_harm$beta.bmi / -qnorm(p=(my_data_harm$p.value.bmi / 2), mean = 0) * sqrt(n.orig) / sqrt(.x)), 
                                                                           samplesize = .x))

cum_expl_varX2_sample_size <- c()
for (mysamples in 1:length(colnames_pval)) {
  temp_dat <- my_data_harm %>% 
    select(all_of(colnames_pval[mysamples])) %>% 
    mutate(explained_variance_BMI2_sample_size = explained_variance_BMI2_sample_size[[mysamples]]) %>% 
    filter(!!sym(colnames_pval[mysamples]) < threshold)
  #print(temp_dat)
  cum_expl_varX2_sample_size <- append(cum_expl_varX2_sample_size, sum(temp_dat$explained_variance_BMI2_sample_size))
}

df_cum_expl_var <- df_cum_expl_var %>% 
  cbind(cum_expl_varX2_sample_size) %>% 
  relocate(cum_expl_varX2_sample_size, .before = cum_expl_varU)

ggplot(df_cum_expl_var, aes(x=n))+
  geom_line(aes(y=cum_expl_varX2_sample_size*100), col="red")+  # slightly higher PVE if samplesize mapped into PVE_X formula
  geom_line(aes(y=cum_expl_varX2*100))
```


```{r updated-fig01, fig.height=6, fig.width=10, warning=FALSE}
# 6.       Represent figure similar to Fig1 in Chatterjee’s paper

# Dual Y axis: we have 284 IVs, scale them to 5% (y-axis [0;5%]): 3/284 
## https://r-graph-gallery.com/line-chart-dual-Y-axis-ggplot2.html
scale_factor <- 4.5/284
N_annotation <- c(10000, 100000, 200000, 300000)

df_cum_expl_var <- df_cum_expl_var %>% 
  mutate(
    cum_expl_varX = cum_expl_varX *100, # in percent 
    cum_expl_varX2 = cum_expl_varX2_sample_size *100,  #  CAVE CHANGED HERE TO INCLUDE SAMPLE SIZE DEPENDENT SE and PVE CALC!
    cum_expl_varU = cum_expl_varU *100,
    cum_expl_varY = cum_expl_varY *100,
    count_IVs_scaled = count_IVs*scale_factor)

plot_data1b <- df_cum_expl_var %>% 
  pivot_longer(cols = c(cum_expl_varX2, cum_expl_varU, cum_expl_varY)) %>% 
  mutate(name = factor(name, labels = c("% var(U)", "% var(X)", "% var(Y)")),
         name = factor(name, levels = c("% var(X)", "% var(Y)", "% var(U)") )
         )
## plot
ggplot(plot_data1b, aes(x=n, y=value, col=name))+
  geom_point(alpha=0.6)+
  geom_line(alpha=0.6)+
  geom_line(aes(x=n, y=count_IVs_scaled, col="Number of IVs"), linetype="dotdash")+
  scale_x_continuous(name = paste0("Size of sample ", as.roman(1)),  # 2-sample-MR or label "Study Size"
                     labels = comma, 
                     breaks = N_annotation, #c( 10000, n[7:length(n)]),
                     limits = c(10000, 300000))+
  #ylab("% variance explained by IVs")+
  labs(title = "Figure 1",
       col='', 
       caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  scale_y_continuous(
    # Features of the first axis
    name = "% variance explained by IVs",
    # Add a second axis and specify its features
    sec.axis = sec_axis(~./scale_factor, name="Number of IVs"),
    limits = c(0,5)
  ) + 
  theme_few()+
  #scale_colour_few() +
  #scale_color_manual(name = "", values = c("% var(U)" = "red", 
  #                                         "% var(X)" = "blue", 
  #                                         "% var(Y)" = "darkblue", 
  #                                         "Number of IVs" = "black"
  #                                         ))
  #
  scale_color_manual(name = "", values = c("% var(U)" = ggthemes_data$few$colors$Dark[2,2][[1]], 
                                           "% var(X)" = ggthemes_data$few$colors$Dark[3,2][[1]], 
                                           "% var(Y)" = ggthemes_data$few$colors$Dark[4,2][[1]], 
                                           "Number of IVs" = "black"),
                     # order
                     breaks = c( "Number of IVs", "% var(X)", "% var(Y)", "% var(U)")
  ) -> fig01b.update
fig01b.update  # USE THIS

# if (SAVE.files ==TRUE ){
#   filenamefig01_b <- paste0("../output/plots/", Sys.Date(), "_figure01_new_PVE_X_formula.png")
#   ggsave(filename = filenamefig01_b, plot = fig01b.update, 
#          width = 10, height = 6, 
#          units = "in"  # default
#          )
# }
```

Discussion: *Pleiotropy* with increasing sample size. INSIDE assumption. 


# Different genetic architectures of the exposure

## MAF vs P-val

For the ~200 IVs, please plot the minor allele frequency (x-axis) versus -log10Pval (y-axis). Calculate the median MAF and the median –log10Pval, and draw the corresponding vertical and horizontal lines. The four areas will define four genetic architectures for the simulations (e.g. relatively rare variants with relatively large exposure effects).

```{r MAF-Pval, warning=FALSE}
ggplot(data = my_data_harm, aes(x=eaf.bmi, y=-log10(p.value.bmi)))+
  geom_point(alpha=0.6)+
  geom_hline(yintercept=-log10(median(my_data$p_value_exp)), col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
  geom_vline(xintercept=median(my_data$eaf_exp), col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
  xlab("Effect allele frequency for exposure")+
  ylab("-log10(P-value) ")+
  labs(title = "MAF vs Pval",
       subtitle = "Define different genetic architectures",
       #col='', 
       #caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  theme_few()+
  scale_colour_few() ->fig02
fig02

# ggsave(filename = "../output/plots/figure02.png", plot = fig02)
```

## MAF vs betas

```{r mafVSbetas}
plot_data_2 <- my_data_harm %>% 
  mutate(maf.bmi = ifelse(eaf.bmi >0.5, 1-eaf.bmi, eaf.bmi)) %>% 
  select(maf.bmi, beta.bmi, gene_label, fto_col_annot)


ggplot(data = plot_data_2, aes(x=maf.bmi, y=abs(beta.bmi), 
                               col=fto_col_annot, label=gene_label))+
  geom_point(alpha=0.6)+
  ggrepel::geom_text_repel(show.legend = FALSE)+
  geom_hline(yintercept=median(abs(plot_data_2$beta.bmi)), 
             col=ggthemes_data$few$colors$Dark[6,2][[1]], alpha=0.6, linetype=2)+
  geom_vline(xintercept=median(plot_data_2$maf.bmi), 
             col=ggthemes_data$few$colors$Dark[6,2][[1]], alpha=0.6, linetype=2)+
  annotate(geom = "text", x = 0.15, y = 0.045, 
           label="relatively rare\nrelatively strong", 
           color= ggthemes_data$few$colors$Dark[2,2][[1]]
             )+
  annotate(geom = "text", x = 0.42, y = 0.05, 
           label="relatively common\nrelatively strong", 
           color=ggthemes_data$few$colors$Dark[3,2][[1]])+
  annotate(geom = "text", x = 0.1, y = 0.015, 
           label="relatively rare\nrelatively weak", color=ggthemes_data$few$colors$Dark[4,2][[1]])+
  annotate("segment", x=0.44,xend=0.42,y=0.025,yend=0.013,arrow=arrow(), 
           color=ggthemes_data$few$colors$Dark[5,2][[1]])+
  annotate(geom = "text", x = 0.44, y = 0.03, 
           label="relatively common\nrelatively weak", 
           color=ggthemes_data$few$colors$Dark[5,2][[1]])+
  #geom_text(x=0.1, y=0.05, label="hello", family = "Times New Roman",
  #          size=6, col="red")+
  xlab("Minor allele frequency")+
  ylab("Exposure effect size")+  # abs(beta_BMI)
  labs(title = "",
       subtitle = "Define different genetic architectures",
       col='SNP', 
       #caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  theme_few()+
  scale_colour_few() -> fig02.update
fig02.update


# # SAVE
# if (SAVE.files ==TRUE ){
#   filenamefig02 <- paste0("../output/plots/", Sys.Date(), "_figure02.png")
#   ggsave(filename = filenamefig02, plot = fig02.update)
# }
```


We want to repeat Figure 01 for each of the 4 defined quadrants in the MAF vs beta plot above. 

```{r figure1quadrants, fig.show="hold", out.width="50%", warning=FALSE}
## plotfigure sourced from helper functions ----------------------------------
# create all combis
effect_size <- c("strong", "weak")
maf <- c("rare", "common")
all_quadrants <- tidyr::crossing(effect_size, maf)

####### PLOT NOW  ---
for (i in 1:nrow(all_quadrants)) {
  plotfigure01(data = my_data_harm, 
               effect_size = all_quadrants$effect_size[i], 
               maf = all_quadrants$maf[i], 
               n=n
               )
}

```

## FINAL PLOT

```{r FINAL-PLOT-1-huray, fig.height=10, fig.width=14, message=FALSE, warning=FALSE}
fig01b.update + 
  labs(title= "", caption = "") -> fig01b.update.arrange

fig02.update+
  labs(title= "", subtitle="", caption = "") -> fig02.update.arrange

figure01_EFFECT_SIZE_strong_VARIANTS_common+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[3,2][[1]])) -> figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange

figure01_EFFECT_SIZE_strong_VARIANTS_rare+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[2,2][[1]])) -> 
  figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_common+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[5,2][[1]])) -> figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_rare+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[4,2][[1]])) -> figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange

ggpubr::ggarrange(fig02.update.arrange, fig01b.update.arrange,
                  figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange, figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange,
                  figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange, figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange, 
                  ncol=2, nrow=3, 
                  heights = c(2, 1.5, 1.5),
                  legend = "bottom", common.legend = TRUE, align = "hv"
                  ) -> final_plot1.arrange


figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange+
  # delete y-labels
  scale_y_continuous(name="% variance explained by IVs", sec.axis = sec_axis(~./scale_factor, name=""), limits = c(0,2)) -> figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange

figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange+
  # delete y-labels
  scale_y_continuous(name="", sec.axis = sec_axis(~./scale_factor, name=""), limits = c(0,2)) -> figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange+
  # delete y-labels
  scale_y_continuous(name="",sec.axis = sec_axis(~./scale_factor, name=""), limits = c(0,2)) -> figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange+
  # delete y-labels
  scale_y_continuous(name="", sec.axis = sec_axis(~./scale_factor, name="Number of IVs"), limits = c(0,2)) -> figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange

ggpubr::ggarrange(ggarrange(fig02.update.arrange, fig01b.update.arrange, ncol=2, labels = c("A", "B"), legend = "none", align = "hv"),
                  ggarrange(figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange,
                            figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange, 
                            figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange,
                            figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange, 
                            ncol=4, labels = c("C", "D", "E", "F"), 
                            legend="bottom", common.legend = TRUE),
                  nrow=2
                  #heights = c(2, 1.5, 1.5),
                  #legend = "bottom", common.legend = TRUE, align = "hv"
                  ) -> final_plot1.arrange2

# final_plot1.arrange   # OLD
final_plot1.arrange2   # use this!
if (SAVE.files ==TRUE ){
  filename_final_plot1 <- paste0("./output/plots/pve/", Sys.Date(), "_final_plot1.arrange_OLD.svg")
  filename_final_plot1b <- paste0("./output/plots/pve/", Sys.Date(), "_final_plot1b.arrange_NEW.svg")
  
  # ggsave(filename = filename_final_plot1, plot = final_plot1.arrange, 
  #          width = 14, height = 10, 
  #          units = "in"  # default
  #          )
  ggsave(filename = filename_final_plot1b, plot = final_plot1.arrange2, 
           width = 14, height = 10, 
           units = "in"  # default
           )
}
```


\clearpage

# A *Gallstone Disease* Simulation Framework

## Get MR-estimates first

### Using {MendelianRandomization} package

```{r MendelianRandomization-analysis-GSD, warning=FALSE}
# outsourced calculation to `perform_mr.R`
# data FOR estimates in `perform_mr.R` script and simulation in `simstudy.R` script

if (SAVE.files ==TRUE ){
  saveRDS(my_data_harm, paste0("./output/Rdata/harmonized-dat/", Sys.Date(), "_my_data_harm.rds"))
}

## load results from script `perform_mr.R`
est.bmi.gsd <- readRDS(dplyr::last(list.files("./output/Rdata/mr-results/", pattern = "_est_BMI_GSD.rds", full.names = TRUE))) 

# biology plausibility FTO & MC4R
# ## The mr_ivw function implements the inverse-variance method, informally 
# known as the "Toby Johnson" method. With a single genetic variant, this 
# is simply the ratio method.
fto_wald <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$beta.gsd[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.gsd[my_data_harm$SNP == "rs1558902"]
)

fto_wald_ukbiobank <- mr_wald_ratio(
  b_exp = exp(FTO_dat$beta[3])-1, # BMI UK Biobank log-transformed  (before Hoffmann GERA + GIANT)
  b_out = FTO_dat$beta[2], # GSD UKB as before
  se_exp = exp(FTO_dat$se[3])-1, 
  se_out = FTO_dat$se[2]
  )

fto_wald_ukbiobank_noexp <- mr_wald_ratio(
  b_exp = (FTO_dat$beta[3]), # BMI UK Biobank log-transformed  (before Hoffmann GERA + GIANT)
  b_out = FTO_dat$beta[2], # GSD UKB as before
  se_exp = (FTO_dat$se[3]), 
  se_out = FTO_dat$se[2]
  )

mc4r_wald <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$beta.gsd[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.gsd[my_data_harm$SNP == "rs6567160"]
)
# add rows
est.bmi.gsd <- est.bmi.gsd %>% 
  add_row(mr_methods = "FTO_Wald", estimate = fto_wald$b, se.estimate=fto_wald$se) %>% 
  add_row(mr_methods = "MC4R_Wald", estimate = mc4r_wald$b, se.estimate=mc4r_wald$se) %>% 
  arrange(desc(estimate))

est.bmi.gsd %>% 
  mutate(OR=exp(estimate), CI.lower=exp(estimate-se.estimate),CI.upper=exp(estimate+se.estimate)) %>%  
  mutate(across(where(is.numeric), round, 3)) %>% 
  arrange(tolower(mr_methods)) %>% 
  relocate(time, .after = last_col()) -> est.bmi.gsd.table

# est.bmi.gsd.table %>% 
#   kable(digits=3, caption = "MR Results for GSD (outcome) using genetic variants as instrumental variables for BMI") %>% 
#   kable_classic(full_width = FALSE, html_font = "Cambria") 

est.bmi.gsd.table %>% 
  gt() %>% 
  cols_merge(
      columns = c(estimate, se.estimate),
      pattern = "{1} [\u00b1 {2}]"  # +- sign : "\u00b1"
    ) %>% 
  cols_merge(
      columns = c(OR, CI.lower, CI.upper),
      pattern = "{1} [{2}; {3}]"  
    ) %>% 
  cols_label(
    mr_methods = md("Methods"),
    estimate = md("Estimate [\u00b1 StdErr]"),
    OR = md("OR [95% CI]"),
    time=md("Computational time [sec]")
  ) %>% 
  fmt_number(decimals = 3, use_seps=TRUE) %>% 
  tab_source_note(
    source_note = md("MR Results for GSD (outcome) using genetic variants as instrumental variables for BMI")
  ) %>% 
  opt_stylize(style=1, color="gray") -> gt.tables.est.bmi.gsd


if (SAVE.files ==TRUE ){
 filename.gt.est.bmi.gsd <- paste0("./output/gt-tables/mr-estimates/", Sys.Date(), 
                                       "_est_bmi_gsd")
 
 gt.tables.est.bmi.gsd %>% 
   gtsave(paste0(filename.gt.est.bmi.gsd, ".docx"))   # alternative .rtf, .tex, .png, .html
 gt.tables.est.bmi.gsd %>% 
   gtsave(paste0(filename.gt.est.bmi.gsd, ".html")) 
}
```


### Scatter Plot

```{r scatter-MendelianRandomization-GSD, fig.width=10, fig.height=6, warning=FALSE, message=FALSE}
# create mr.object
mr.obj = MendelianRandomization::mr_input(
    # exposure
    bx = my_data_harm$beta.bmi, 
    bxse = my_data_harm$se.bmi, 
    # outcome
    by = my_data_harm$beta.gsd, 
    byse = my_data_harm$se.gsd, 
    snps = my_data_harm$SNP,
    exposure = "Body mass index",
    outcome = "Gallstone disease"
  )

p1.gsd <- mr_plot(mr.obj, error = TRUE, 
                  line = "ivw", 
                  orientate = FALSE, 
                  interactive = FALSE, 
                  labels = FALSE)

p1.gsd$layers[[1]]$aes_params$size <- 1
p1.gsd$layers[[1]]$aes_params$alpha <- 0.5

p1.gsd$layers[[4]]$aes_params$colour <- ggthemes_data$few$colors$Dark[1,2][[1]]
p1.gsd$layers[[5]]$aes_params$colour <- ggthemes_data$few$colors$Dark[1,2][[1]]
p1.gsd$layers[[4]]$aes_params$alpha <- 0.3
p1.gsd$layers[[5]]$aes_params$alpha <- 0.3

p1.gsd$layers[[6]]$aes_params$colour <- ggthemes_data$few$colors$Dark[2,2][[1]]

p1.gsd+
  ggthemes::theme_few() -> p1.gsd

p1.gsd

# SAVE
if (SAVE.files ==TRUE ){
  filename.mr.scatter.gsd <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_mrscatter_gsd.svg")
  ggsave(filename = filename.mr.scatter.gsd, plot = p1.gsd, 
           width = 10, height = 6, 
           units = "in"  # default
           )
}
```


### Forest Plot

```{r forestplot-MendelianRandomization-GSD, fig.width=5, fig.height=27, warning=FALSE}
p2.gsd <- mr_forest(
  mr.obj,
  alpha = 0.05,
  snp_estimates = TRUE,
  methods = "ivw",
  ordered = TRUE
) +
  ggthemes::theme_few()

#SAVE
if (SAVE.files ==TRUE ){
  filename.mr.forest.gsd <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_mrforest_GSD.svg")
  ggsave(filename = filename.mr.forest.gsd, plot = p2.gsd, 
         width = 5, 
         height = 24, 
           #units = "in"  # default
           )
}
```

### Funnel Plot

```{r funnel-plot-GSD, fig.height=3, fig.width=5}
p4.gsd <- mr_funnel(mr.obj, CI=TRUE)

p4.gsd$layers[[1]]$aes_params$size <- 1
p4.gsd$layers[[1]]$aes_params$alpha <- 0.8
# linerange layer
p4.gsd$layers[[2]]$aes_params$alpha <- 0.2
# xintercept layer - estimate
p4.gsd$layers[[4]]$aes_params$colour <- ggthemes_data$few$colors$Dark[9,2][[1]]
# delete linerange layer
p4.gsd$layers <- p4.gsd$layers[-2]  

# plot
p4.gsd <- p4.gsd+
  ggthemes::theme_few()
p4.gsd

ggarrange(p1.gsd, p4.gsd, nrow = 1, labels = c("A", "B")) -> gsd_scatter_funnel_arrange

#SAVE
if (SAVE.files ==TRUE ){
  filename.mr.funnel.gsd <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_mrfunnel_gsd.svg")
  filename.mr.scatterfunnel.gsd <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_scatter_mrfunnel_arrange_gsd.svg")
  ggsave(filename = filename.mr.funnel.gsd, plot = p4.gsd, 
         width = 8, 
         height = 4, 
         units = "in"  # default
           )
  
  ggsave(filename.mr.scatterfunnel.gsd, gsd_scatter_funnel_arrange, width=8, height=4, units="in")
}



```


\clearpage


# B *Gallbladder Cancer* Simulation Framework

## Get MR-estimates first

### Using {TwoSampleMR} package

```{r TwoSampleMR-analysis}
harmonised_dat4TwoSampleMR <- harmonised_dat %>% 
  mutate(id.exposure = "BMI", id.outcome = "GBC")  # https://mrcieu.github.io/TwoSampleMR/articles/perform_mr.html

res <- TwoSampleMR::mr(harmonised_dat4TwoSampleMR)
res <- res %>% mutate(OR = exp(b), CI.lower= exp(b-se), CI.upper=exp(b+se)) %>% 
  arrange(desc(b))

res %>% 
  kable(digits = 3, 
        caption = "MR Results for GBC (outcome) using genetic variants as instrumental variables for BMI"
        ) %>% 
  kable_classic(full_width = FALSE, html_font = "Cambria")
```

### Scatter Plot

```{r scatter-TwoSampleMR, warning=FALSE, message=FALSE}
# mr_method_list()
res.selected_methods <- TwoSampleMR::mr(harmonised_dat4TwoSampleMR, 
                                        method_list = c("mr_egger_regression", "mr_ivw"))
p1 <- mr_scatter_plot(res.selected_methods, harmonised_dat4TwoSampleMR)
p1[[1]] #+
  #ggthemes::theme_few()

# SAVE
if (SAVE.files ==TRUE ){
  filename.mr.scatter <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_mrscatter.svg")
  ggsave(filename = filename.mr.scatter, plot = p1[[1]], 
           #width = 10, height = 6, 
           #units = "in"  # default
           )
}
```

### Forest Plot

```{r forestplot-TwoSampleMR-GBC, fig.width=5, fig.height=24, warning=FALSE}
res_single <- TwoSampleMR::mr_singlesnp(harmonised_dat4TwoSampleMR, all_method = c("mr_egger_regression", "mr_ivw"))
p2 <- mr_forest_plot(res_single,  exponentiate = FALSE)
p2[[1]]

#SAVE
if (SAVE.files ==TRUE ){
  filename.mr.forest <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_mrforest_GBC.svg")
  ggsave(filename = filename.mr.forest, plot = p2[[1]], 
         width = 5, 
         height = 24, 
           #units = "in"  # default
           )
}
```

### Funnel Plot

```{r funnel-plot, fig.height=3, fig.width=5}
p4 <- mr_funnel_plot(res_single)
p4[[1]] +
  ggthemes::theme_few() -> p4

p4

#SAVE
if (SAVE.files ==TRUE ){
  filename.mr.funnel.gbc <- paste0("./output/plots/mr-analysis/", Sys.Date(), "_figure_mrfunnel_gbc.svg")
  ggsave(filename = filename.mr.funnel.gbc, plot = p4, 
         width = 8, 
         height = 4, 
         units = "in"  # default
           )
}
```


### Using {MendelianRandomization} package

```{r MendelianRandomization-analysis, warning=FALSE}
## load results
est <- readRDS(file = dplyr::last(list.files("./output/Rdata/mr-results/", pattern = "_est.rds", full.names = TRUE)))

# biology plausibility FTO & MC4R
# ## The mr_ivw function implements the inverse-variance method, informally 
# known as the "Toby Johnson" method. With a single genetic variant, this 
# is simply the ratio method.
fto_wald.gbc <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$beta.gbc[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.gbc[my_data_harm$SNP == "rs1558902"]
)
mc4r_wald.gbc <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$beta.gbc[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.gbc[my_data_harm$SNP == "rs6567160"]
)
# add rows
est <- est %>% 
  add_row(mr_methods = "FTO_Wald", estimate = fto_wald.gbc$b, se.estimate=fto_wald.gbc$se) %>% 
  add_row(mr_methods = "MC4R_Wald", estimate = mc4r_wald.gbc$b, se.estimate=mc4r_wald.gbc$se)


est %>% 
  mutate(OR=exp(estimate), CI.lower=exp(estimate-se.estimate),CI.upper=exp(estimate+se.estimate)) %>%  
  arrange(desc(estimate)) %>% 
  kable(digits=3, caption = "MR Results for GBC (outcome) using genetic variants as instrumental variables for BMI") %>% 
  kable_classic(full_width = FALSE, html_font = "Cambria")
```



# Simulation

We simulate betas from a normal distribution. The script is outsourced to `simstudy.R`. 

```{r simulatesimulatesimulate, warning=FALSE}
# save simulation-vars for simulation `simstudy.R` script
sim_vars <- list(n=n, colnames_pval=colnames_pval)
if (SAVE.files ==TRUE ){
  saveRDS(sim_vars, paste0("./output/Rdata/", Sys.Date(), "_sim_vars.rds"))
}
## SIMULATION OUTSOURCED TO "./scripts/simstudy.R"

# read dat from "simstudy.R"  #------------------------------------------------
#est_sim <- readRDS(dplyr::last(list.files(path = "../output/Rdata", pattern = "_est_sim_NSIM_", full.names = TRUE)))
# from "purrr-simstudy.R"
est_sim <- readRDS(dplyr::last(list.files(path = "./output/Rdata/sim-results/", pattern = "_est_sim_purrr_NSIM_", full.names = TRUE)))
est_sim_architectures <- readRDS(dplyr::last(list.files(path = "./output/Rdata/sim-results/", pattern = "_est_sim_architectures_purrr_NSIM_", full.names = TRUE)))
nsim <- nrow(est_sim)
nsim_architectures <- nrow(est_sim_architectures[[1]])
```


## PVE with simulated data Plot

```{r simulated-pves, fig.height=6, fig.width=10, warning=FALSE}
# retrieve PVE colnames
colname_pve_X <-  colnames(est_sim)[str_detect(colnames(est_sim), pattern = "pve_X_sim")]
# PVE_Y, PVE_U removed on 2023-01-24 (see github)

## for performance measures later
# retrieve IVW/ MODE/ CONMIX/... cols for BMI -> GSD 
#1
colname_IVW_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "IVW_theta_sim_BMI_GSD_")]
colname_IVW_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "IVW_theta_se_sim_BMI_GSD_")]
#2
colname_MODE_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MODE_theta_sim_BMI_GSD_")]
colname_MODE_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MODE_theta_se_sim_BMI_GSD_")]
#3
colname_CONMIX_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "CONMIX_theta_sim_BMI_GSD_")]
colname_CONMIX_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "CONMIX_theta_se_sim_BMI_GSD_")]
#4
colname_MRMIX_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MRMIX_theta_sim_BMI_GSD_")]
colname_MRMIX_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MRMIX_theta_se_sim_BMI_GSD_")]
#5
colname_MEDIAN_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MEDIAN_theta_sim_BMI_GSD_")]
colname_MEDIAN_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MEDIAN_theta_se_sim_BMI_GSD_")]
#6
colname_EGGER_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "EGGER_theta_sim_BMI_GSD_")]
colname_EGGER_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "EGGER_theta_se_sim_BMI_GSD_")]
#7
colname_ROBUST_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "ROBUST_theta_sim_BMI_GSD_")]
colname_ROBUST_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "ROBUST_theta_se_sim_BMI_GSD_")]
#8
colname_PRESSO_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "PRESSO_theta_sim_BMI_GSD_")]
colname_PRESSO_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "PRESSO_theta_se_sim_BMI_GSD_")]
#9
colname_RAPS_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "RAPS_theta_sim_BMI_GSD_")]
colname_RAPS_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "RAPS_theta_se_sim_BMI_GSD_")]
#10
colname_LASSO_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "LASSO_theta_sim_BMI_GSD_")]
colname_LASSO_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "LASSO_theta_se_sim_BMI_GSD_")]

# retrieve IVW/ MODE/ CONMIX cols for BMI -> GBC
# removed on 2023-01-24 (see github)

# filter out sample sizes (IVW only used as example, could be any other col)
sample_size_filtered <- as.numeric(str_extract(colname_IVW_theta_sim_BMI_GSD, pattern = "(\\d)+"))

# EXPOSURE PVE ------------------------------------------------------------------------------------------------------
est_sim %>% 
  select(!!!(colname_pve_X)) %>% 
  pivot_longer(everything(), names_to = "PVE_X", values_to = "value_X") %>% 
  cbind(sample_size_filtered) %>% as_tibble() -> pve_X_sim

# mean PVE per sample size
est_sim %>% 
  select(!!!(colname_pve_X)) %>% 
  map(~mean(.x)) %>% 
  unlist() -> PVE_X_sim_mean

boxplot_overlay_dat <- pve_X_sim %>% 
  # cbind(pve_U_sim %>% select(-sample_size_filtered)) %>% 
  # cbind(pve_Y_sim %>% select(-sample_size_filtered)) %>% 
  as_tibble()

## plot
ggplot(plot_data1b, aes(x=n, y=value, col=name))+    # plot_data1b from section Simlating on empirical data (PVE)
  geom_point(alpha=0.6)+
  geom_line(alpha=0.6)+
  geom_line(aes(x=n, y=count_IVs_scaled, col="Number of IVs"), linetype="dotdash")+
  scale_x_continuous(name = paste0("Size of sample ", as.roman(1)),  # 2-sample-MR or label "Study Size"
                     labels = comma, 
                     breaks = N_annotation, #c( 10000, n[7:length(n)]),
                     limits = c(10000, 300000))+
  #ylab("% variance explained by IVs")+
  labs(title = "PVE with simulation added",
       col='', 
       caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  scale_y_continuous(
    # Features of the first axis
    name = "% variance explained by IVs",
    # Add a second axis and specify its features
    sec.axis = sec_axis(~./scale_factor, name="Number of IVs"),
    limits = c(0,7.5)
  ) + 
  #geom_boxplot(data=boxplot_overlay_dat, aes(x=sample_size_filtered, y=value_X*100, group=sample_size_filtered))
  geom_boxplot(data = boxplot_overlay_dat, aes(x=sample_size_filtered, y=value_X*100, # *100 to get percent values
                                group=sample_size_filtered), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2)+
  # U/Y removed on 2023-01-24 (see github)
  theme_few()+
  #scale_colour_few() +
  #scale_color_manual(name = "", values = c("% var(U)" = "red", 
  #                                         "% var(X)" = "blue", 
  #                                         "% var(Y)" = "darkblue", 
  #                                         "Number of IVs" = "black"
  #                                         ))
  #
  scale_color_manual(name = "", values = c("% var(U)" = ggthemes_data$few$colors$Dark[2,2][[1]], 
                                           "% var(X)" = ggthemes_data$few$colors$Dark[3,2][[1]], 
                                           "% var(Y)" = ggthemes_data$few$colors$Dark[4,2][[1]], 
                                           "Number of IVs" = "black"),
                     # order
                     breaks = c( "Number of IVs", "% var(X)", "% var(Y)", "% var(U)")
  ) -> plot_pve_simulated
#plot_pve_simulated

# if (SAVE.files ==TRUE ){
#   filename_pveplot_sim <- paste0("../output/plots/pve/", Sys.Date(), "_figure_simplot_pve.png")
#   ggsave(filename = filename_pveplot_sim, plot = plot_pve_simulated, 
#          width = 10, height = 6, 
#          units = "in"  # default
#          )
# }
```

```{r simulated-pves-raincloud-plot, fig.height=6, fig.width=10, warning=FALSE}
ggplot(plot_data1b, aes(x=n, y=value, col=name))+    # plot_data1b from section Simlating on empirical data (PVE)
  geom_point(alpha=0.6)+
  geom_line(alpha=0.6)+
  geom_line(aes(x=n, y=count_IVs_scaled, col="Number of IVs"), linetype="dotdash")+
  scale_x_continuous(name = paste0("Size of sample ", as.roman(1)),  # 2-sample-MR or label "Study Size"
                     labels = comma, 
                     breaks = N_annotation, #c( 10000, n[7:length(n)]),
                     #limits = c(10000, 300000)  # throws an error when used with ggdist -> use coord_cartesian() instead!
                     )+
  #ylab("% variance explained by IVs")+
  labs(title = "PVE with simulation added",
       col='', 
       caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
  ) +
  scale_y_continuous(
    # Features of the first axis
    name = "% variance explained by IVs",
    # Add a second axis and specify its features
    sec.axis = sec_axis(~./scale_factor, name="Number of IVs"),
    limits = c(0,7.5)
  ) + 
  ## Rainclouds
  # https://z3tt.github.io/Rainclouds/
  # exposure simulation
  geom_boxplot(data = boxplot_overlay_dat, aes(x=sample_size_filtered, y=value_X*100, # *100 to get percent values
                                               group=sample_size_filtered), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = boxplot_overlay_dat, mapping=aes(x=sample_size_filtered, y=value_X*100, group=sample_size_filtered), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = boxplot_overlay_dat, mapping=aes(x=sample_size_filtered-2000, y=value_X*100, group=sample_size_filtered), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  coord_cartesian(xlim=c(10000, 300000))+
  theme_few()+
  #scale_colour_few() +
  #scale_color_manual(name = "", values = c("% var(U)" = "red", 
  #                                         "% var(X)" = "blue", 
  #                                         "% var(Y)" = "darkblue", 
  #                                         "Number of IVs" = "black"
  #                                         ))
  #
  scale_color_manual(name = "", values = c("% var(U)" = ggthemes_data$few$colors$Dark[2,2][[1]], 
                                           "% var(X)" = ggthemes_data$few$colors$Dark[3,2][[1]], 
                                           "% var(Y)" = ggthemes_data$few$colors$Dark[4,2][[1]], 
                                           "Number of IVs" = "black"),
                     # order
                     breaks = c( "Number of IVs", "% var(X)", "% var(Y)", "% var(U)")
  ) -> plot_pve_simulated.raincloud
plot_pve_simulated.raincloud

if (SAVE.files ==TRUE ){
  filename_pveplot_sim.raincloud <- paste0("./output/plots/pve/", Sys.Date(), "_figure_simplot_pve_raincloud.svg")
  ggsave(filename = filename_pveplot_sim.raincloud, plot = plot_pve_simulated.raincloud, 
         width = 10, height = 6, 
         units = "in"  # default
         )
}
```

## Performance Metrics

### Calculate BIAS, MSE, MEAN, EMP_SE

```{r calculate-performance-metrics}
# fns
bias <- function(thetaobs,theta, ...) mean(thetaobs)- theta
mse <- function(thetaobs, theta, ...) mean((thetaobs-theta)^2)
emp_se <- function(nsim, thetaobs, theta, ...) sqrt( 1/(nsim-1) * sum((thetaobs-mean(thetaobs))^2) )  # == sd (with ...)


## get performance metrics    # https://stackoverflow.com/questions/47415072/pass-multiple-functions-to-purrrmap
funs <- c(mean = mean, emp_se=emp_se, bias = bias, mse = mse)
## args gsd
#1
args.gsd.ivw <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "IVW"], 
                     nsim=nrow(est_sim)
                     ) # trim argument non-matching and ignored for median
#2
args.gsd.mode <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "mode"], 
                     nsim=nrow(est_sim)
                     )
#3
args.gsd.conmix <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "conmix"], 
                     nsim=nrow(est_sim)
                     )
#4
args.gsd.mrmix <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "MRMix"], 
                     nsim=nrow(est_sim)
                     )
#5
args.gsd.median <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "median"], 
                     nsim=nrow(est_sim)
                     )
#6
args.gsd.egger <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "egger"], 
                     nsim=nrow(est_sim)
                     )
#7
args.gsd.robust <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "Robust"], 
                     nsim=nrow(est_sim)
                     )
#8
args.gsd.presso <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "PRESSO"], 
                     nsim=nrow(est_sim)
                     )
#9
args.gsd.raps <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "RAPS"], 
                     nsim=nrow(est_sim)
                     )
#10
args.gsd.lasso <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "Lasso"], 
                     nsim=nrow(est_sim)
                     )
## args gbc
#deleted 2023-01-29

## Standard Deviation
# est_sim %>%    # equivalent to `emp_se()` function, but `sd()` misses dotdotdot argument 
#   select(!!!(colname_theta)) %>% 
#   map_dbl(.f = ~sd(.))
```


```{r CALC-perf-metrics-GSD}
#1
IVW_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.ivw), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#2
MODE_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.mode), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#3
CONMIX_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.conmix), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#4
MRMIX_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.mrmix), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#5
MEDIAN_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_MEDIAN_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.median), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#6
EGGER_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_EGGER_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.egger), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#7
ROBUST_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_ROBUST_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.robust), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#8
PRESSO_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_PRESSO_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.presso), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#9
RAPS_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_RAPS_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.raps), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
#10
LASSO_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_LASSO_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.lasso), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
```

```{r CALC-perf-metrics-GBC}
# deleted 2023-01-29
```

### Calculate COVERAGE, COVERAGE-BIAS-ELIMINATED, TYPE1ERROR

```{r}
# print("95% Confidence Interval: [{:.2f}, {:.2f}]".format(lower_bound, upper_bound))
```


```{r further-metrics}
# coverage
coverage <- function(thetaobs, thetaobs_se, theta, ...) {
  # https://www.r-bloggers.com/2021/11/calculate-confidence-intervals-in-r/
  lower_ci <- thetaobs - 1.96*thetaobs_se
  upper_ci <- thetaobs + 1.96*thetaobs_se
  coverage <- mean(lower_ci <= theta & upper_ci >= theta)
  return(coverage)
}

# coverage bias eliminated
coverage_bias_eliminated <- function(thetaobs, thetaobs_se, ...) {
  lower_ci <- thetaobs - 1.96*thetaobs_se
  upper_ci <- thetaobs + 1.96*thetaobs_se
  mean_thetaobs <- mean(thetaobs)
  coverage_bias_eliminated <- mean(lower_ci <= mean_thetaobs & upper_ci >=mean_thetaobs)
  return(coverage_bias_eliminated)
}
# type1error
type1error <- function(pvals) sum(pvals<=0.05)/length(pvals)  # no causal_effect detected

# GSD
#1
IVW_GSD_coverage_calc <- c()
IVW_GSD_coverage_bias_eliminated_calc <- c()
IVW_GSD_type1error_calc <- c()
#2
MODE_GSD_coverage_calc <- c()
MODE_GSD_coverage_bias_eliminated_calc <- c()
MODE_GSD_type1error_calc <- c()
#3
CONMIX_GSD_coverage_calc <- c()
CONMIX_GSD_coverage_bias_eliminated_calc <- c()
CONMIX_GSD_type1error_calc <- c()
#4
MRMIX_GSD_coverage_calc <- c()
MRMIX_GSD_coverage_bias_eliminated_calc <- c()
MRMIX_GSD_type1error_calc <- c()
#5
MEDIAN_GSD_coverage_calc <- c()
MEDIAN_GSD_coverage_bias_eliminated_calc <- c()
MEDIAN_GSD_type1error_calc <- c()
#6
EGGER_GSD_coverage_calc <- c()
EGGER_GSD_coverage_bias_eliminated_calc <- c()
EGGER_GSD_type1error_calc <- c()
#7
ROBUST_GSD_coverage_calc <- c()
ROBUST_GSD_coverage_bias_eliminated_calc <- c()
ROBUST_GSD_type1error_calc <- c()
#8
PRESSO_GSD_coverage_calc <- c()
PRESSO_GSD_coverage_bias_eliminated_calc <- c()
PRESSO_GSD_type1error_calc <- c()
#9
RAPS_GSD_coverage_calc <- c()
RAPS_GSD_coverage_bias_eliminated_calc <- c()
RAPS_GSD_type1error_calc <- c()
#10
LASSO_GSD_coverage_calc <- c()
LASSO_GSD_coverage_bias_eliminated_calc <- c()
LASSO_GSD_type1error_calc <- c()

for (i in 1:length(sample_size_filtered)) {
  sample_size <- sample_size_filtered[i]
  # GSD additional metrics ----------------------------------------------------
  ## 1/ IVW
  col_IVW_GSD_theta <- paste0("IVW_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_IVW_GSD_theta_se <- paste0("IVW_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_IVW_GSD_mr_pval <- paste0("IVW_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  IVW_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_IVW_GSD_theta], thetaobs_se = est_sim[col_IVW_GSD_theta_se], theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "IVW"])
  IVW_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_IVW_GSD_theta]], thetaobs_se = est_sim[[col_IVW_GSD_theta_se]])
  IVW_GSD_type1error_calc[i] <- type1error(est_sim[[col_IVW_GSD_mr_pval]])
  ## 2/ MODE
  col_MODE_GSD_theta <- paste0("MODE_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MODE_GSD_theta_se <- paste0("MODE_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MODE_GSD_mr_pval <- paste0("MODE_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  MODE_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MODE_GSD_theta], 
                                       thetaobs_se = est_sim[col_MODE_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "mode"])
  MODE_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MODE_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_MODE_GSD_theta_se]])
  MODE_GSD_type1error_calc[i] <- type1error(est_sim[[col_MODE_GSD_mr_pval]])
  ## 3/ CONMIX
  col_CONMIX_GSD_theta <- paste0("CONMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_CONMIX_GSD_theta_se <- paste0("CONMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_CONMIX_GSD_mr_pval <- paste0("CONMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  CONMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_CONMIX_GSD_theta], 
                                       thetaobs_se = est_sim[col_CONMIX_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "conmix"])
  CONMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_CONMIX_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_CONMIX_GSD_theta_se]])
  CONMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_CONMIX_GSD_mr_pval]])  
  ## 4/ MRMIX
  col_MRMIX_GSD_theta <- paste0("MRMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MRMIX_GSD_theta_se <- paste0("MRMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MRMIX_GSD_mr_pval <- paste0("MRMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  MRMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MRMIX_GSD_theta], 
                                       thetaobs_se = est_sim[col_MRMIX_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "MRMix"])
  MRMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MRMIX_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_MRMIX_GSD_theta_se]])
  MRMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_MRMIX_GSD_mr_pval]])
  ## 5/ MEDIAN
  col_MEDIAN_GSD_theta <- paste0("MEDIAN_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MEDIAN_GSD_theta_se <- paste0("MEDIAN_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MEDIAN_GSD_mr_pval <- paste0("MEDIAN_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  MEDIAN_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MEDIAN_GSD_theta], 
                                       thetaobs_se = est_sim[col_MEDIAN_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "median"])
  MEDIAN_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MEDIAN_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_MEDIAN_GSD_theta_se]])
  MEDIAN_GSD_type1error_calc[i] <- type1error(est_sim[[col_MEDIAN_GSD_mr_pval]])
  ## 6/ EGGER
  col_EGGER_GSD_theta <- paste0("EGGER_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_EGGER_GSD_theta_se <- paste0("EGGER_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_EGGER_GSD_mr_pval <- paste0("EGGER_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  EGGER_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_EGGER_GSD_theta], 
                                       thetaobs_se = est_sim[col_EGGER_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "egger"])
  EGGER_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_EGGER_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_EGGER_GSD_theta_se]])
  EGGER_GSD_type1error_calc[i] <- type1error(est_sim[[col_EGGER_GSD_mr_pval]])  
  ## 7/ ROBUST
  col_ROBUST_GSD_theta <- paste0("ROBUST_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_ROBUST_GSD_theta_se <- paste0("ROBUST_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_ROBUST_GSD_mr_pval <- paste0("ROBUST_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  ROBUST_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_ROBUST_GSD_theta], 
                                       thetaobs_se = est_sim[col_ROBUST_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "Robust"])
  ROBUST_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_ROBUST_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_ROBUST_GSD_theta_se]])
  ROBUST_GSD_type1error_calc[i] <- type1error(est_sim[[col_ROBUST_GSD_mr_pval]])  
  ## 8/ PRESSO
  col_PRESSO_GSD_theta <- paste0("PRESSO_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_PRESSO_GSD_theta_se <- paste0("PRESSO_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_PRESSO_GSD_mr_pval <- paste0("PRESSO_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  PRESSO_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_PRESSO_GSD_theta], 
                                       thetaobs_se = est_sim[col_PRESSO_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "PRESSO"])
  PRESSO_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_PRESSO_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_PRESSO_GSD_theta_se]])
  PRESSO_GSD_type1error_calc[i] <- type1error(est_sim[[col_PRESSO_GSD_mr_pval]])
  ## 9/ RAPS
  col_RAPS_GSD_theta <- paste0("RAPS_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_RAPS_GSD_theta_se <- paste0("RAPS_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_RAPS_GSD_mr_pval <- paste0("RAPS_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  RAPS_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_RAPS_GSD_theta], 
                                       thetaobs_se = est_sim[col_RAPS_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "RAPS"])
  RAPS_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_RAPS_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_RAPS_GSD_theta_se]])
  RAPS_GSD_type1error_calc[i] <- type1error(est_sim[[col_RAPS_GSD_mr_pval]])  
  ## 10/ LASSO
  col_LASSO_GSD_theta <- paste0("LASSO_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_LASSO_GSD_theta_se <- paste0("LASSO_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_LASSO_GSD_mr_pval <- paste0("LASSO_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  LASSO_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_LASSO_GSD_theta], 
                                       thetaobs_se = est_sim[col_LASSO_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "Lasso"])
  LASSO_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_LASSO_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_LASSO_GSD_theta_se]])
  LASSO_GSD_type1error_calc[i] <- type1error(est_sim[[col_LASSO_GSD_mr_pval]])  
}
```


### GSD Performance Metrics

```{r add-additional-metrics-to-df-and-print-GSD}
## GSD --------------------------
#1
IVW_GSD_performance_metrics <- IVW_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(IVW_GSD_coverage_calc, IVW_GSD_coverage_bias_eliminated_calc, IVW_GSD_type1error_calc)) %>% 
  rename(coverage = IVW_GSD_coverage_calc, 
         coverage_bias_eliminated = IVW_GSD_coverage_bias_eliminated_calc, 
         type1error = IVW_GSD_type1error_calc) %>% 
  as_tibble()
IVW_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "IVW method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#2
MODE_GSD_performance_metrics <- MODE_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(MODE_GSD_coverage_calc, 
                   MODE_GSD_coverage_bias_eliminated_calc, MODE_GSD_type1error_calc)) %>% 
  rename(coverage = MODE_GSD_coverage_calc, 
         coverage_bias_eliminated = MODE_GSD_coverage_bias_eliminated_calc, 
         type1error = MODE_GSD_type1error_calc) %>% 
  as_tibble()
MODE_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "MODE method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#3
CONMIX_GSD_performance_metrics <- CONMIX_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(CONMIX_GSD_coverage_calc, 
                   CONMIX_GSD_coverage_bias_eliminated_calc, CONMIX_GSD_type1error_calc)) %>% 
  rename(coverage = CONMIX_GSD_coverage_calc, 
         coverage_bias_eliminated = CONMIX_GSD_coverage_bias_eliminated_calc, 
         type1error = CONMIX_GSD_type1error_calc) %>% 
  as_tibble()
CONMIX_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "CONMIX method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#4
MRMIX_GSD_performance_metrics <- MRMIX_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(MRMIX_GSD_coverage_calc, 
                   MRMIX_GSD_coverage_bias_eliminated_calc, MRMIX_GSD_type1error_calc)) %>% 
  rename(coverage = MRMIX_GSD_coverage_calc, 
         coverage_bias_eliminated = MRMIX_GSD_coverage_bias_eliminated_calc, 
         type1error = MRMIX_GSD_type1error_calc) %>% 
  as_tibble()
MRMIX_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "MRMIX method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#5
MEDIAN_GSD_performance_metrics <- MEDIAN_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(MEDIAN_GSD_coverage_calc, 
                   MEDIAN_GSD_coverage_bias_eliminated_calc, MEDIAN_GSD_type1error_calc)) %>% 
  rename(coverage = MEDIAN_GSD_coverage_calc, 
         coverage_bias_eliminated = MEDIAN_GSD_coverage_bias_eliminated_calc, 
         type1error = MEDIAN_GSD_type1error_calc) %>% 
  as_tibble()
MEDIAN_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "MEDIAN method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#6
EGGER_GSD_performance_metrics <- EGGER_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(EGGER_GSD_coverage_calc, 
                   EGGER_GSD_coverage_bias_eliminated_calc, EGGER_GSD_type1error_calc)) %>% 
  rename(coverage = EGGER_GSD_coverage_calc, 
         coverage_bias_eliminated = EGGER_GSD_coverage_bias_eliminated_calc, 
         type1error = EGGER_GSD_type1error_calc) %>% 
  as_tibble()
EGGER_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "EGGER method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#7
ROBUST_GSD_performance_metrics <- ROBUST_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(ROBUST_GSD_coverage_calc, 
                   ROBUST_GSD_coverage_bias_eliminated_calc, ROBUST_GSD_type1error_calc)) %>% 
  rename(coverage = ROBUST_GSD_coverage_calc, 
         coverage_bias_eliminated = ROBUST_GSD_coverage_bias_eliminated_calc, 
         type1error = ROBUST_GSD_type1error_calc) %>% 
  as_tibble()
ROBUST_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "ROBUST method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#8
PRESSO_GSD_performance_metrics <- PRESSO_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(PRESSO_GSD_coverage_calc, 
                   PRESSO_GSD_coverage_bias_eliminated_calc, PRESSO_GSD_type1error_calc)) %>% 
  rename(coverage = PRESSO_GSD_coverage_calc, 
         coverage_bias_eliminated = PRESSO_GSD_coverage_bias_eliminated_calc, 
         type1error = PRESSO_GSD_type1error_calc) %>% 
  as_tibble()
PRESSO_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "PRESSO method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#9
RAPS_GSD_performance_metrics <- RAPS_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(RAPS_GSD_coverage_calc, 
                   RAPS_GSD_coverage_bias_eliminated_calc, RAPS_GSD_type1error_calc)) %>% 
  rename(coverage = RAPS_GSD_coverage_calc, 
         coverage_bias_eliminated = RAPS_GSD_coverage_bias_eliminated_calc, 
         type1error = RAPS_GSD_type1error_calc) %>% 
  as_tibble()
RAPS_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "RAPS method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
#10
LASSO_GSD_performance_metrics <- LASSO_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(LASSO_GSD_coverage_calc, 
                   LASSO_GSD_coverage_bias_eliminated_calc, LASSO_GSD_type1error_calc)) %>% 
  rename(coverage = LASSO_GSD_coverage_calc, 
         coverage_bias_eliminated = LASSO_GSD_coverage_bias_eliminated_calc, 
         type1error = LASSO_GSD_type1error_calc) %>% 
  as_tibble()
LASSO_GSD_performance_metrics %>% 
  mutate(across(where(is.numeric), round, 3)) %>% 
  kable(caption = "LASSO method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
```



```{r}
library(rsimsum)
s1 <- simsum(data = est_sim, estvarname = "IVW_theta_sim_BMI_GSD_150000", se = "IVW_theta_se_sim_BMI_GSD_150000", true = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "IVW"] )

## stats param
# • nsim,the number of replications with non-missing point estimates and standard error.
# • thetamean, average point estimate.
# • thetamedian, median point estimate.
# • se2mean, average variance.
# • se2median, median variance.
# • bias, bias in point estimate.
# • empse, empirical standard error.
# • mse, mean squared error.
# • relprec, percentage gain in precision relative to the reference method.
# • modelse, model-based standard error.
# • relerror, relative percentage error in standard error.
# • cover, coverage of a nominal level\
# • becover, bias corrected coverage of a nominal level\
# • power, power of a (1 - level)\
statsparam <- c("nsim", "thetamean", "se2mean", "bias", "empse", "mse", "cover", "becover", "power")
simsum_summary <- summary(s1, stats = statsparam)
```

---

## Performance Plots

### GSD simulated theta distribution

```{r theta-sim-distribution-GSD, fig.height=6, fig.width=10, warning=FALSE}
## IVW
IVW_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(IVW_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.ivw$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.ivw$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = IVW_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "IVW estimate: BMI -> GSD")+
  theme_few() -> IVW_GSD_simplot_thetas
IVW_GSD_simplot_thetas

## MODE
MODE_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(MODE_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.mode$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.mode$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = MODE_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "MODE estimate: BMI -> GSD")+
  theme_few() -> MODE_GSD_simplot_thetas
MODE_GSD_simplot_thetas

## CONMIX
CONMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.conmix$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.conmix$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = CONMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "CONMIX estimate: BMI -> GSD")+
  theme_few() -> CONMIX_GSD_simplot_thetas
CONMIX_GSD_simplot_thetas

## MRMIX
MRMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.mrmix$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.mrmix$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = MRMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "MRMIX estimate: BMI -> GSD")+
  theme_few() -> MRMIX_GSD_simplot_thetas
MRMIX_GSD_simplot_thetas

# # SAVE   # we loop in the next section and save the ggarrange plots for inspection
# filename_IVW_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_IVW_GSD_simplot_thetas_NSIM_", nsim,".png")
# ggsave(filename = filename_IVW_GSD_simplot_thetas, plot = IVW_GSD_simplot_thetas, 
#        width = 15, height = 6, 
#        units = "in"  # default
#        )
# filename_MODE_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_MODE_GSD_simplot_thetas_NSIM_", nsim,".png")
# ggsave(filename = filename_MODE_GSD_simplot_thetas, plot = MODE_GSD_simplot_thetas, 
#        width = 15, height = 6, 
#        units = "in"  # default
#        )
# filename_CONMIX_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_CONMIX_GSD_simplot_thetas_NSIM_", nsim,".png")
# ggsave(filename = filename_CONMIX_GSD_simplot_thetas, plot = CONMIX_GSD_simplot_thetas, 
#        width = 15, height = 6, 
#        units = "in"  # default
#        )
# filename_MRMIX_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_CONMIX_GSD_simplot_thetas_NSIM_", nsim,".png")
# ggsave(filename = filename_MRMIX_GSD_simplot_thetas, plot = MRMIX_GSD_simplot_thetas, 
#        width = 15, height = 6, 
#        units = "in"  # default
#        )
```

```{r theta-sim-distribution-GSD-arrange, fig.height=18, fig.width=10, warning=FALSE}
IVW_GSD_simplot_thetas$labels$title <- ""
MODE_GSD_simplot_thetas$labels$title <- ""
CONMIX_GSD_simplot_thetas$labels$title <- ""
MRMIX_GSD_simplot_thetas$labels$title <- ""
ggarrange(IVW_GSD_simplot_thetas,MODE_GSD_simplot_thetas,CONMIX_GSD_simplot_thetas, MRMIX_GSD_simplot_thetas,
          ncol = 1, nrow = 4) -> GSD_simplot_thetas_arrange


# if (SAVE.files ==TRUE ){
#   filename_GSD_simplot_thetas_arrange <- paste0("../output/plots/", Sys.Date(), "_GSD_simplot_thetas_NSIM_", nsim,".png")
#   ggsave(filename = filename_GSD_simplot_thetas_arrange, plot = GSD_simplot_thetas_arrange, 
#          width = 10, height = 22, 
#          units = "in"  # default
#          )
# }
```


### GSD Bias & MSE

```{r plot-performance-metrics-bias-mse-GSD, fig.height=10, fig.width=8}
## IVW
#BIAS
ggplot(IVW_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "IVW estimate: BMI -> GSD")+
  theme_few() -> biasplot_IVW_GSD
#MSE
ggplot(IVW_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "IVW estimate: BMI -> GSD")+
  theme_few() -> mseplot_IVW_GSD
ggarrange(biasplot_IVW_GSD, mseplot_IVW_GSD) -> bias_mse_plot_IVW_GSD

## MODE
ggplot(MODE_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "MODE estimate: BMI -> GSD")+
  theme_few() -> biasplot_MODE_GSD
ggplot(MODE_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "MODE estimate: BMI -> GSD")+
  theme_few() -> mseplot_MODE_GSD
ggarrange(biasplot_MODE_GSD, mseplot_MODE_GSD) -> bias_mse_plot_MODE_GSD

## CONMIX
ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "CONMIX estimate: BMI -> GSD")+
  theme_few() -> biasplot_CONMIX_GSD
ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "CONMIX estimate: BMI -> GSD")+
  theme_few() -> mseplot_CONMIX_GSD
ggarrange(biasplot_CONMIX_GSD, mseplot_CONMIX_GSD) -> bias_mse_plot_CONMIX_GSD

## MRMIX
ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "MRMIX estimate: BMI -> GSD")+
  theme_few() -> biasplot_MRMIX_GSD
ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "MRMIX estimate: BMI -> GSD")+
  theme_few() -> mseplot_MRMIX_GSD
ggarrange(biasplot_MRMIX_GSD, mseplot_MRMIX_GSD) -> bias_mse_plot_MRMIX_GSD

ggarrange(bias_mse_plot_IVW_GSD, bias_mse_plot_MODE_GSD, bias_mse_plot_CONMIX_GSD, bias_mse_plot_MRMIX_GSD,
          ncol = 1, nrow = 4, common.legend = TRUE) -> bias_mse_plot_GSD
bias_mse_plot_GSD

# if (SAVE.files ==TRUE ){
#   filename_simplot_bias_mse <- paste0("../output/plots/", Sys.Date(), "_GSD_simplot_bias_mse_NSIM_", nsim, ".png")
#   ggsave(filename = filename_simplot_bias_mse, plot = bias_mse_plot_GSD, 
#          width = 7, height = 12, 
#          units = "in"  # default
#          )
# }
```


**Deleted all code for GBC simulation downstream analysis since no longer needed for thesis.**

---

# Parallelize performance estimates for different "TRUE" estimates

We use 5 BMI -> GSD reference estimates: IVW random model, MRMix, weighted median, Wald estimate MC4R variant, Wald estimate FTO variant

## Calculate performance measures

```{r}
selected_true_methods <- c("IVW", "MRMix", "median", "MC4R_Wald", "FTO_Wald")
true_estimates <- est.bmi.gsd %>% 
  filter(mr_methods %in% selected_true_methods)
```

```{r parallel-performance}
for (true_method in 1:length(selected_true_methods)) {
  # define reference estimate for ALL 5 chosen methods in loop 
  args.gsd.method <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]], 
                          nsim=nrow(est_sim)
  )
  #1
  IVW_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, IVW_GSD_mean=mean, IVW_GSD_emp_se=emp_se, IVW_GSD_bias=bias, IVW_GSD_mse=mse)
  #2
  MODE_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, MODE_GSD_mean=mean, MODE_GSD_emp_se=emp_se, MODE_GSD_bias=bias, MODE_GSD_mse=mse)
  #3
  CONMIX_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, CONMIX_GSD_mean=mean, CONMIX_GSD_emp_se=emp_se, CONMIX_GSD_bias=bias, CONMIX_GSD_mse=mse)
  #4
  MRMIX_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, MRMIX_GSD_mean=mean, MRMIX_GSD_emp_se=emp_se, MRMIX_GSD_bias=bias, MRMIX_GSD_mse=mse)
  #5
  MEDIAN_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_MEDIAN_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, MEDIAN_GSD_mean=mean, MEDIAN_GSD_emp_se=emp_se, MEDIAN_GSD_bias=bias, MEDIAN_GSD_mse=mse)
  #6
  EGGER_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_EGGER_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, EGGER_GSD_mean=mean, EGGER_GSD_emp_se=emp_se, EGGER_GSD_bias=bias, EGGER_GSD_mse=mse)
  #7
  ROBUST_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_ROBUST_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, ROBUST_GSD_mean=mean, ROBUST_GSD_emp_se=emp_se, ROBUST_GSD_bias=bias, ROBUST_GSD_mse=mse)
  #8
  PRESSO_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_PRESSO_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, PRESSO_GSD_mean=mean, PRESSO_GSD_emp_se=emp_se, PRESSO_GSD_bias=bias, PRESSO_GSD_mse=mse)
  #9
  RAPS_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_RAPS_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, RAPS_GSD_mean=mean, RAPS_GSD_emp_se=emp_se, RAPS_GSD_bias=bias, RAPS_GSD_mse=mse)
  #10
  LASSO_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_LASSO_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, LASSO_GSD_mean=mean, LASSO_GSD_emp_se=emp_se, LASSO_GSD_bias=bias, LASSO_GSD_mse=mse)
  
  ## further metrics
  IVW_GSD_coverage_calc <- c()
  IVW_GSD_coverage_bias_eliminated_calc <- c()
  IVW_GSD_type1error_calc <- c()
  MODE_GSD_coverage_calc <- c()
  MODE_GSD_coverage_bias_eliminated_calc <- c()
  MODE_GSD_type1error_calc <- c()
  CONMIX_GSD_coverage_calc <- c()
  CONMIX_GSD_coverage_bias_eliminated_calc <- c()
  CONMIX_GSD_type1error_calc <- c()
  MRMIX_GSD_coverage_calc <- c()
  MRMIX_GSD_coverage_bias_eliminated_calc <- c()
  MRMIX_GSD_type1error_calc <- c()
  #5
  MEDIAN_GSD_coverage_calc <- c()
  MEDIAN_GSD_coverage_bias_eliminated_calc <- c()
  MEDIAN_GSD_type1error_calc <- c()
  #6
  EGGER_GSD_coverage_calc <- c()
  EGGER_GSD_coverage_bias_eliminated_calc <- c()
  EGGER_GSD_type1error_calc <- c()
  #7
  ROBUST_GSD_coverage_calc <- c()
  ROBUST_GSD_coverage_bias_eliminated_calc <- c()
  ROBUST_GSD_type1error_calc <- c()
  #8
  PRESSO_GSD_coverage_calc <- c()
  PRESSO_GSD_coverage_bias_eliminated_calc <- c()
  PRESSO_GSD_type1error_calc <- c()
  #9
  RAPS_GSD_coverage_calc <- c()
  RAPS_GSD_coverage_bias_eliminated_calc <- c()
  RAPS_GSD_type1error_calc <- c()
  #10
  LASSO_GSD_coverage_calc <- c()
  LASSO_GSD_coverage_bias_eliminated_calc <- c()
  LASSO_GSD_type1error_calc <- c()

  for (i in 1:length(sample_size_filtered)) {
    sample_size <- sample_size_filtered[i]
    # GSD additional metrics ----------------------------------------------------
    ## 1/ IVW
    col_IVW_GSD_theta <- paste0("IVW_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_IVW_GSD_theta_se <- paste0("IVW_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_IVW_GSD_mr_pval <- paste0("IVW_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    IVW_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_IVW_GSD_theta], thetaobs_se = est_sim[col_IVW_GSD_theta_se], theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    IVW_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_IVW_GSD_theta]], thetaobs_se = est_sim[[col_IVW_GSD_theta_se]])
    IVW_GSD_type1error_calc[i] <- type1error(est_sim[[col_IVW_GSD_mr_pval]])
    ## 2/ MODE
    col_MODE_GSD_theta <- paste0("MODE_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MODE_GSD_theta_se <- paste0("MODE_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MODE_GSD_mr_pval <- paste0("MODE_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    MODE_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MODE_GSD_theta], 
                                         thetaobs_se = est_sim[col_MODE_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    MODE_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MODE_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_MODE_GSD_theta_se]])
    MODE_GSD_type1error_calc[i] <- type1error(est_sim[[col_MODE_GSD_mr_pval]])
    ## 3/ CONMIX
    col_CONMIX_GSD_theta <- paste0("CONMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_CONMIX_GSD_theta_se <- paste0("CONMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_CONMIX_GSD_mr_pval <- paste0("CONMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    CONMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_CONMIX_GSD_theta], 
                                         thetaobs_se = est_sim[col_CONMIX_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    CONMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_CONMIX_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_CONMIX_GSD_theta_se]])
    CONMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_CONMIX_GSD_mr_pval]])  
    ## 4/ MRMIX
    col_MRMIX_GSD_theta <- paste0("MRMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MRMIX_GSD_theta_se <- paste0("MRMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MRMIX_GSD_mr_pval <- paste0("MRMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    MRMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MRMIX_GSD_theta], 
                                         thetaobs_se = est_sim[col_MRMIX_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    MRMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MRMIX_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_MRMIX_GSD_theta_se]])
    MRMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_MRMIX_GSD_mr_pval]])  
    ## 5/ MEDIAN
    col_MEDIAN_GSD_theta <- paste0("MEDIAN_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MEDIAN_GSD_theta_se <- paste0("MEDIAN_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MEDIAN_GSD_mr_pval <- paste0("MEDIAN_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    MEDIAN_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MEDIAN_GSD_theta], 
                                         thetaobs_se = est_sim[col_MEDIAN_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    MEDIAN_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MEDIAN_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_MEDIAN_GSD_theta_se]])
    MEDIAN_GSD_type1error_calc[i] <- type1error(est_sim[[col_MEDIAN_GSD_mr_pval]])  
    ## 6/ EGGER
    col_EGGER_GSD_theta <- paste0("EGGER_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_EGGER_GSD_theta_se <- paste0("EGGER_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_EGGER_GSD_mr_pval <- paste0("EGGER_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    EGGER_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_EGGER_GSD_theta], 
                                         thetaobs_se = est_sim[col_EGGER_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    EGGER_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_EGGER_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_EGGER_GSD_theta_se]])
    EGGER_GSD_type1error_calc[i] <- type1error(est_sim[[col_EGGER_GSD_mr_pval]])
    ## 7/ ROBUST
    col_ROBUST_GSD_theta <- paste0("ROBUST_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_ROBUST_GSD_theta_se <- paste0("ROBUST_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_ROBUST_GSD_mr_pval <- paste0("ROBUST_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    ROBUST_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_ROBUST_GSD_theta], 
                                         thetaobs_se = est_sim[col_ROBUST_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    ROBUST_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_ROBUST_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_ROBUST_GSD_theta_se]])
    ROBUST_GSD_type1error_calc[i] <- type1error(est_sim[[col_ROBUST_GSD_mr_pval]])
    ## 8/ PRESSO
    col_PRESSO_GSD_theta <- paste0("PRESSO_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_PRESSO_GSD_theta_se <- paste0("PRESSO_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_PRESSO_GSD_mr_pval <- paste0("PRESSO_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    PRESSO_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_PRESSO_GSD_theta], 
                                         thetaobs_se = est_sim[col_PRESSO_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    PRESSO_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_PRESSO_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_PRESSO_GSD_theta_se]])
    PRESSO_GSD_type1error_calc[i] <- type1error(est_sim[[col_PRESSO_GSD_mr_pval]])
    ## 9/ RAPS
    col_RAPS_GSD_theta <- paste0("RAPS_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_RAPS_GSD_theta_se <- paste0("RAPS_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_RAPS_GSD_mr_pval <- paste0("RAPS_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    RAPS_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_RAPS_GSD_theta], 
                                         thetaobs_se = est_sim[col_RAPS_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    RAPS_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_RAPS_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_RAPS_GSD_theta_se]])
    RAPS_GSD_type1error_calc[i] <- type1error(est_sim[[col_RAPS_GSD_mr_pval]])
    ## 10/LASSO
    col_LASSO_GSD_theta <- paste0("LASSO_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_LASSO_GSD_theta_se <- paste0("LASSO_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_LASSO_GSD_mr_pval <- paste0("LASSO_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    LASSO_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_LASSO_GSD_theta], 
                                         thetaobs_se = est_sim[col_LASSO_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    LASSO_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_LASSO_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_LASSO_GSD_theta_se]])
    LASSO_GSD_type1error_calc[i] <- type1error(est_sim[[col_LASSO_GSD_mr_pval]])
    #-----------
  }
  
  ## combine metrics
  IVW_GSD_performance_metrics <- IVW_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(IVW_GSD_coverage_calc, IVW_GSD_coverage_bias_eliminated_calc, IVW_GSD_type1error_calc)) %>% 
    as_tibble()
  MODE_GSD_performance_metrics <- MODE_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(MODE_GSD_coverage_calc, 
                     MODE_GSD_coverage_bias_eliminated_calc, MODE_GSD_type1error_calc)) %>% 
    as_tibble()
  CONMIX_GSD_performance_metrics <- CONMIX_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(CONMIX_GSD_coverage_calc, 
                     CONMIX_GSD_coverage_bias_eliminated_calc, CONMIX_GSD_type1error_calc)) %>% 
    as_tibble()
  MRMIX_GSD_performance_metrics <- MRMIX_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(MRMIX_GSD_coverage_calc, 
                     MRMIX_GSD_coverage_bias_eliminated_calc, MRMIX_GSD_type1error_calc)) %>% 
    as_tibble()
  #5
  MEDIAN_GSD_performance_metrics <- MEDIAN_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(MEDIAN_GSD_coverage_calc, 
                     MEDIAN_GSD_coverage_bias_eliminated_calc, MEDIAN_GSD_type1error_calc)) %>% 
    as_tibble()
  #6
  EGGER_GSD_performance_metrics <- EGGER_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(EGGER_GSD_coverage_calc, 
                     EGGER_GSD_coverage_bias_eliminated_calc, EGGER_GSD_type1error_calc)) %>% 
    as_tibble()
  #7
  ROBUST_GSD_performance_metrics <- ROBUST_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(ROBUST_GSD_coverage_calc, 
                     ROBUST_GSD_coverage_bias_eliminated_calc, ROBUST_GSD_type1error_calc)) %>% 
    as_tibble()
  #8
  PRESSO_GSD_performance_metrics <- PRESSO_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(PRESSO_GSD_coverage_calc, 
                     PRESSO_GSD_coverage_bias_eliminated_calc, PRESSO_GSD_type1error_calc)) %>% 
    as_tibble()
  #9
  RAPS_GSD_performance_metrics <- RAPS_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(RAPS_GSD_coverage_calc, 
                     RAPS_GSD_coverage_bias_eliminated_calc, RAPS_GSD_type1error_calc)) %>% 
    as_tibble()
  #10
  LASSO_GSD_performance_metrics <- LASSO_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(LASSO_GSD_coverage_calc, 
                     LASSO_GSD_coverage_bias_eliminated_calc, LASSO_GSD_type1error_calc)) %>% 
    as_tibble()
  
  # #----
  # COMBINE TO PRESENT RESULTS ----------------------------------------------------
  # #----
  df_name <- paste0("df_combined_method_metrics_TRUE_EST_", selected_true_methods[true_method])
  assign(x = df_name, value = tibble(N=n))
  # scenarios <- c("all_instruments", "common_strong", "common_weak", "rare_strong", "rare_weak")
  assign(x = df_name, value = eval(as.symbol(df_name)) %>% 
    left_join(IVW_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(MODE_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(CONMIX_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(MRMIX_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(MEDIAN_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(EGGER_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(ROBUST_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(PRESSO_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(RAPS_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
    left_join(LASSO_GSD_performance_metrics %>% select(-var), by = c("N"="N"))
  )
  #eval(as.symbol(df_name))

  ###
  ## PLOT  -------------------------------------------------------------------
  ###
  IVW_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(IVW_GSD_performance_metrics, aes(x=N, y=IVW_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    # geom_errorbar(aes(ymin=IVW_GSD_mean-1.96*IVW_GSD_emp_se, ymax=IVW_GSD_mean+1.96*IVW_GSD_emp_se), 
    #               width=3000, # x-axis dimension
    #               position = "dodge")+
    ## ADD rainbow plots
    geom_boxplot(data = IVW_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("IVW estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> IVW_GSD_simplot_thetas
  
  ## 2 MODE
  MODE_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(MODE_GSD_performance_metrics, aes(x=N, y=MODE_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    # geom_errorbar(aes(ymin=MODE_GSD_mean-1.96*MODE_GSD_emp_se, ymax=MODE_GSD_mean+1.96*MODE_GSD_emp_se), 
    #               width=3000, # x-axis dimension
    #               position = "dodge")+
    ## ADD rainbow plots
    geom_boxplot(data = MODE_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("MODE estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> MODE_GSD_simplot_thetas
  
  ## 3 CONMIX
  CONMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=CONMIX_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = CONMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("CONMIX estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> CONMIX_GSD_simplot_thetas
  
  ## 4 MRMIX
  MRMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=MRMIX_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = MRMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("MRMIX estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> MRMIX_GSD_simplot_thetas
  
  ## 5 MEDIAN
  MEDIAN_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_MEDIAN_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(MEDIAN_GSD_performance_metrics, aes(x=N, y=MEDIAN_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = MEDIAN_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = MEDIAN_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = MEDIAN_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("MEDIAN estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> MEDIAN_GSD_simplot_thetas
  
  ## 6 EGGER
  EGGER_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_EGGER_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(EGGER_GSD_performance_metrics, aes(x=N, y=EGGER_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = EGGER_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = EGGER_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = EGGER_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("EGGER estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> EGGER_GSD_simplot_thetas
  ## 7 ROBUST
  ROBUST_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_ROBUST_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(ROBUST_GSD_performance_metrics, aes(x=N, y=ROBUST_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = ROBUST_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = ROBUST_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = ROBUST_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("ROBUST estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> ROBUST_GSD_simplot_thetas
  ## 8 PRESSO
  PRESSO_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_PRESSO_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(PRESSO_GSD_performance_metrics, aes(x=N, y=PRESSO_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = PRESSO_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = PRESSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = PRESSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("PRESSO estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> PRESSO_GSD_simplot_thetas
  ## 9 RAPS
  RAPS_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_RAPS_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(RAPS_GSD_performance_metrics, aes(x=N, y=RAPS_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = RAPS_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = RAPS_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = RAPS_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("RAPS estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> RAPS_GSD_simplot_thetas
  ## 10 LASSO
  LASSO_BMI_GSD_geom_point_dat <- est_sim %>% 
    select(!!!(colname_LASSO_theta_sim_BMI_GSD)) %>% 
    pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
    mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
  
  ggplot(LASSO_GSD_performance_metrics, aes(x=N, y=LASSO_GSD_mean))+ 
    geom_point()+
    geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
           label= expression( paste("ref: ", theta[X] ))  , 
           color= ggthemes_data$few$colors$Dark[9,2][[1]]
             )+
    ## ADD rainbow plots
    geom_boxplot(data = LASSO_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                 col=ggthemes_data$few$colors$Dark[3,2][[1]],
                 alpha=0.2, size=0.2, outlier.shape = NA)+
    ggdist::stat_halfeye(data = LASSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                         inherit.aes = FALSE, 
                         slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                         geom = "slabinterval",
                         na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
    gghalves::geom_half_point(
      data = LASSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
      inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
      side = "l", range_scale = .4, alpha = .2)+
    ## FINISH rainbow
    scale_x_continuous(labels = comma)+
    xlab("Size of Sample I")+
    ylab("Mean thetas")+
    scale_y_continuous(limits = c(0.25,0.8))+
    labs(title = "Simulated Thetas with Empirical Standard Error", 
         subtitle = paste0("LASSO estimate: BMI -> GSD\nSelected Reference Method: ", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> LASSO_GSD_simplot_thetas

  IVW_GSD_simplot_thetas$labels$title <- ""
  MODE_GSD_simplot_thetas$labels$title <- ""
  CONMIX_GSD_simplot_thetas$labels$title <- ""
  MRMIX_GSD_simplot_thetas$labels$title <- ""
  #5
  MEDIAN_GSD_simplot_thetas$labels$title <- ""
  #6
  EGGER_GSD_simplot_thetas$labels$title <- ""
  #7
  ROBUST_GSD_simplot_thetas$labels$title <- ""
  #8
  PRESSO_GSD_simplot_thetas$labels$title <- ""
  #9
  RAPS_GSD_simplot_thetas$labels$title <- ""
  #10
  LASSO_GSD_simplot_thetas$labels$title <- ""
  ggarrange(IVW_GSD_simplot_thetas,MODE_GSD_simplot_thetas,CONMIX_GSD_simplot_thetas, MRMIX_GSD_simplot_thetas, MEDIAN_GSD_simplot_thetas,
            EGGER_GSD_simplot_thetas, ROBUST_GSD_simplot_thetas, PRESSO_GSD_simplot_thetas, RAPS_GSD_simplot_thetas, LASSO_GSD_simplot_thetas,
            ncol = 1, nrow = 10) -> GSD_simplot_thetas_arrange
  GSD_simplot_thetas_arrange
  
  if (SAVE.files ==TRUE ){
    filename_GSD_simplot_thetas_arrange <- paste0("./output/plots/theta_sims/", Sys.Date(), "_GSD_simplot_thetas_REF_METHOD_", selected_true_methods[true_method], "_NSIM_", nsim,".svg")
    ggsave(filename = filename_GSD_simplot_thetas_arrange, plot = GSD_simplot_thetas_arrange, 
           width = 10, height = 38, device = "svg",
           units = "in"  # default
           )
  }
  
  ### GSD Bias & MSE
  #BIAS
  ggplot(IVW_GSD_performance_metrics, aes(x=N, y=IVW_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "IVW estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_IVW_GSD
  #MSE
  ggplot(IVW_GSD_performance_metrics, aes(x=N, y=IVW_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "IVW estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_IVW_GSD
  ggarrange(biasplot_IVW_GSD, mseplot_IVW_GSD) -> bias_mse_plot_IVW_GSD
  
  ## MODE
  ggplot(MODE_GSD_performance_metrics, aes(x=N, y=MODE_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "MODE estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_MODE_GSD
  ggplot(MODE_GSD_performance_metrics, aes(x=N, y=MODE_GSD_mse))+ 
    geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "MODE estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_MODE_GSD
  ggarrange(biasplot_MODE_GSD, mseplot_MODE_GSD) -> bias_mse_plot_MODE_GSD
  
  ## CONMIX
  ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=CONMIX_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "CONMIX estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_CONMIX_GSD
  ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=CONMIX_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "CONMIX estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_CONMIX_GSD
  ggarrange(biasplot_CONMIX_GSD, mseplot_CONMIX_GSD) -> bias_mse_plot_CONMIX_GSD
  
  ## MRMIX
  ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=MRMIX_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "MRMIX estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_MRMIX_GSD
  ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=MRMIX_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "MRMIX estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_MRMIX_GSD
  ggarrange(biasplot_MRMIX_GSD, mseplot_MRMIX_GSD) -> bias_mse_plot_MRMIX_GSD
  
  ## 5 MEDIAN
  ggplot(MEDIAN_GSD_performance_metrics, aes(x=N, y=MEDIAN_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "MEDIAN estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_MEDIAN_GSD
  ggplot(MEDIAN_GSD_performance_metrics, aes(x=N, y=MEDIAN_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "MEDIAN estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_MEDIAN_GSD
  ggarrange(biasplot_MEDIAN_GSD, mseplot_MEDIAN_GSD) -> bias_mse_plot_MEDIAN_GSD
  ## 6 EGGER
  ggplot(EGGER_GSD_performance_metrics, aes(x=N, y=EGGER_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "EGGER estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_EGGER_GSD
  ggplot(EGGER_GSD_performance_metrics, aes(x=N, y=EGGER_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "EGGER estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_EGGER_GSD
  ggarrange(biasplot_EGGER_GSD, mseplot_EGGER_GSD) -> bias_mse_plot_EGGER_GSD
  ## 7 ROBUST
  ggplot(ROBUST_GSD_performance_metrics, aes(x=N, y=ROBUST_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "ROBUST estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_ROBUST_GSD
  ggplot(ROBUST_GSD_performance_metrics, aes(x=N, y=ROBUST_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "ROBUST estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_ROBUST_GSD
  ggarrange(biasplot_ROBUST_GSD, mseplot_ROBUST_GSD) -> bias_mse_plot_ROBUST_GSD
  ## 8 PRESSO
  ggplot(PRESSO_GSD_performance_metrics, aes(x=N, y=PRESSO_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "PRESSO estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_PRESSO_GSD
  ggplot(PRESSO_GSD_performance_metrics, aes(x=N, y=PRESSO_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "PRESSO estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_PRESSO_GSD
  ggarrange(biasplot_PRESSO_GSD, mseplot_PRESSO_GSD) -> bias_mse_plot_PRESSO_GSD
  ## 9 RAPS
  ggplot(RAPS_GSD_performance_metrics, aes(x=N, y=RAPS_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "RAPS estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_RAPS_GSD
  ggplot(RAPS_GSD_performance_metrics, aes(x=N, y=RAPS_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "RAPS estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_RAPS_GSD
  ggarrange(biasplot_RAPS_GSD, mseplot_RAPS_GSD) -> bias_mse_plot_RAPS_GSD
  ## LASSO
  ggplot(LASSO_GSD_performance_metrics, aes(x=N, y=LASSO_GSD_bias))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(-1.5,1.5))+
    xlab("Size of Sample I")+
    ylab("Bias")+
    labs(title = "LASSO estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> biasplot_LASSO_GSD
  ggplot(LASSO_GSD_performance_metrics, aes(x=N, y=LASSO_GSD_mse))+ 
    geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
    scale_x_continuous(labels = comma)+
    scale_y_continuous(limits = c(0,1.8))+
    xlab("Size of Sample I")+
    ylab("MSE")+
    labs(title = "LASSO estimate: BMI -> GSD",
         caption = paste0("Selected Reference Method:\n", 
                           selected_true_methods[true_method], 
                           " estimate calculated with total sample size"))+
    theme_few() -> mseplot_LASSO_GSD
  ggarrange(biasplot_LASSO_GSD, mseplot_LASSO_GSD) -> bias_mse_plot_LASSO_GSD
  
  
  ggarrange(bias_mse_plot_IVW_GSD, bias_mse_plot_MODE_GSD, bias_mse_plot_CONMIX_GSD, bias_mse_plot_MRMIX_GSD, bias_mse_plot_MEDIAN_GSD,
            bias_mse_plot_EGGER_GSD, bias_mse_plot_ROBUST_GSD, bias_mse_plot_PRESSO_GSD, bias_mse_plot_RAPS_GSD, bias_mse_plot_LASSO_GSD,
            ncol = 1, nrow = 10, common.legend = TRUE) -> bias_mse_plot_GSD
  bias_mse_plot_GSD
  
  if (SAVE.files ==TRUE ){
    filename_simplot_bias_mse <- paste0("./output/plots/bias-mse/", Sys.Date(), "_GSD_simplot_bias_mse_REF_METHOD_", selected_true_methods[true_method], "_NSIM_",nsim, ".svg")
    ggsave(filename = filename_simplot_bias_mse, plot = bias_mse_plot_GSD, 
           width = 7, height = 24, 
           units = "in"  # default
           )
  }
}

```


## Now loop over all genetic architecture simulations


```{r parallel-performance-genetic-architectures}
effect_size <- c("strong", "weak")
maf <- c("rare", "common")
all_quadrants <- tidyr::crossing(effect_size, maf)

# create copy of est_sim so the genetic architectures can be assigned to this variable in the loop 
# without losing the simulations on all variants not considering genetic architecture
est_sim_all_variants <- est_sim

# now we do not need to rewrite the code from above but can assign each genetic architecture df to  est_sim
# and only need to change the names of the files we store for later usage
for (my_architecture in 1:length(est_sim_architectures)) {
  
  est_sim <- est_sim_architectures[[my_architecture]]
  name_GenArch <- gsub(pattern = "_est_sim", replacement = "", names(est_sim_architectures)[my_architecture])
  
  # fewer variants -> no results for smaller sample sizes -> reassign sample sizes and colnames
  
  #1
  colname_IVW_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "IVW_theta_sim_BMI_GSD_")]
  colname_IVW_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "IVW_theta_se_sim_BMI_GSD_")]
  #2
  colname_MODE_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MODE_theta_sim_BMI_GSD_")]
  colname_MODE_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MODE_theta_se_sim_BMI_GSD_")]
  #3
  colname_CONMIX_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "CONMIX_theta_sim_BMI_GSD_")]
  colname_CONMIX_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "CONMIX_theta_se_sim_BMI_GSD_")]
  #4
  colname_MRMIX_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MRMIX_theta_sim_BMI_GSD_")]
  colname_MRMIX_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MRMIX_theta_se_sim_BMI_GSD_")]
  #5
  colname_MEDIAN_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MEDIAN_theta_sim_BMI_GSD_")]
  colname_MEDIAN_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MEDIAN_theta_se_sim_BMI_GSD_")]
  #6
  colname_EGGER_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "EGGER_theta_sim_BMI_GSD_")]
  colname_EGGER_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "EGGER_theta_se_sim_BMI_GSD_")]
  #7
  colname_ROBUST_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "ROBUST_theta_sim_BMI_GSD_")]
  colname_ROBUST_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "ROBUST_theta_se_sim_BMI_GSD_")]
  #8
  colname_PRESSO_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "PRESSO_theta_sim_BMI_GSD_")]
  colname_PRESSO_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "PRESSO_theta_se_sim_BMI_GSD_")]
  #9
  colname_RAPS_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "RAPS_theta_sim_BMI_GSD_")]
  colname_RAPS_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "RAPS_theta_se_sim_BMI_GSD_")]
  #10
  colname_LASSO_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "LASSO_theta_sim_BMI_GSD_")]
  colname_LASSO_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "LASSO_theta_se_sim_BMI_GSD_")]
  
  # filter out sample sizes (IVW only used as example, could be any other col)
  sample_size_filtered <- as.numeric(str_extract(colname_IVW_theta_sim_BMI_GSD, pattern = "(\\d)+"))
  
  ### NOW START LOOP --------------------------------------------------------
  for (true_method in 1:length(selected_true_methods)) {
    # define reference estimate for ALL 5 chosen methods in loop 
    args.gsd.method <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]], 
                            nsim=nrow(est_sim)
    )
    #1
    IVW_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, IVW_GSD_mean=mean, IVW_GSD_emp_se=emp_se, IVW_GSD_bias=bias, IVW_GSD_mse=mse)
    #2
    MODE_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, MODE_GSD_mean=mean, MODE_GSD_emp_se=emp_se, MODE_GSD_bias=bias, MODE_GSD_mse=mse)
    #3
    CONMIX_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, CONMIX_GSD_mean=mean, CONMIX_GSD_emp_se=emp_se, CONMIX_GSD_bias=bias, CONMIX_GSD_mse=mse)
    #4
    MRMIX_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, MRMIX_GSD_mean=mean, MRMIX_GSD_emp_se=emp_se, MRMIX_GSD_bias=bias, MRMIX_GSD_mse=mse)
    #5
    MEDIAN_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_MEDIAN_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, MEDIAN_GSD_mean=mean, MEDIAN_GSD_emp_se=emp_se, MEDIAN_GSD_bias=bias, MEDIAN_GSD_mse=mse)
    #6
    EGGER_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_EGGER_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, EGGER_GSD_mean=mean, EGGER_GSD_emp_se=emp_se, EGGER_GSD_bias=bias, EGGER_GSD_mse=mse)
    #7
    ROBUST_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_ROBUST_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, ROBUST_GSD_mean=mean, ROBUST_GSD_emp_se=emp_se, ROBUST_GSD_bias=bias, ROBUST_GSD_mse=mse)
    #8
    PRESSO_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_PRESSO_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, PRESSO_GSD_mean=mean, PRESSO_GSD_emp_se=emp_se, PRESSO_GSD_bias=bias, PRESSO_GSD_mse=mse)
    #9
    RAPS_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_RAPS_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, RAPS_GSD_mean=mean, RAPS_GSD_emp_se=emp_se, RAPS_GSD_bias=bias, RAPS_GSD_mse=mse)
    #10
    LASSO_GSD_performance_metrics <- est_sim %>% 
      select(!!!(colname_LASSO_theta_sim_BMI_GSD)) %>% 
      map_df(.f = ~ funs %>% 
               map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
      cbind(sample_size_filtered) %>% 
      rename(N = sample_size_filtered, LASSO_GSD_mean=mean, LASSO_GSD_emp_se=emp_se, LASSO_GSD_bias=bias, LASSO_GSD_mse=mse)
    
    ## further metrics
    IVW_GSD_coverage_calc <- c()
    IVW_GSD_coverage_bias_eliminated_calc <- c()
    IVW_GSD_type1error_calc <- c()
    MODE_GSD_coverage_calc <- c()
    MODE_GSD_coverage_bias_eliminated_calc <- c()
    MODE_GSD_type1error_calc <- c()
    CONMIX_GSD_coverage_calc <- c()
    CONMIX_GSD_coverage_bias_eliminated_calc <- c()
    CONMIX_GSD_type1error_calc <- c()
    MRMIX_GSD_coverage_calc <- c()
    MRMIX_GSD_coverage_bias_eliminated_calc <- c()
    MRMIX_GSD_type1error_calc <- c()
    #5
    MEDIAN_GSD_coverage_calc <- c()
    MEDIAN_GSD_coverage_bias_eliminated_calc <- c()
    MEDIAN_GSD_type1error_calc <- c()
    #6
    EGGER_GSD_coverage_calc <- c()
    EGGER_GSD_coverage_bias_eliminated_calc <- c()
    EGGER_GSD_type1error_calc <- c()
    #7
    ROBUST_GSD_coverage_calc <- c()
    ROBUST_GSD_coverage_bias_eliminated_calc <- c()
    ROBUST_GSD_type1error_calc <- c()
    #8
    PRESSO_GSD_coverage_calc <- c()
    PRESSO_GSD_coverage_bias_eliminated_calc <- c()
    PRESSO_GSD_type1error_calc <- c()
    #9
    RAPS_GSD_coverage_calc <- c()
    RAPS_GSD_coverage_bias_eliminated_calc <- c()
    RAPS_GSD_type1error_calc <- c()
    #10
    LASSO_GSD_coverage_calc <- c()
    LASSO_GSD_coverage_bias_eliminated_calc <- c()
    LASSO_GSD_type1error_calc <- c()
  
    for (i in 1:length(sample_size_filtered)) {
      sample_size <- sample_size_filtered[i]
      # GSD additional metrics ----------------------------------------------------
      ## 1/ IVW
      col_IVW_GSD_theta <- paste0("IVW_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_IVW_GSD_theta_se <- paste0("IVW_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_IVW_GSD_mr_pval <- paste0("IVW_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      IVW_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_IVW_GSD_theta], thetaobs_se = est_sim[col_IVW_GSD_theta_se], theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      IVW_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_IVW_GSD_theta]], thetaobs_se = est_sim[[col_IVW_GSD_theta_se]])
      IVW_GSD_type1error_calc[i] <- type1error(est_sim[[col_IVW_GSD_mr_pval]])
      ## 2/ MODE
      col_MODE_GSD_theta <- paste0("MODE_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_MODE_GSD_theta_se <- paste0("MODE_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_MODE_GSD_mr_pval <- paste0("MODE_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      MODE_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MODE_GSD_theta], 
                                           thetaobs_se = est_sim[col_MODE_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      MODE_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MODE_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_MODE_GSD_theta_se]])
      MODE_GSD_type1error_calc[i] <- type1error(est_sim[[col_MODE_GSD_mr_pval]])
      ## 3/ CONMIX
      col_CONMIX_GSD_theta <- paste0("CONMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_CONMIX_GSD_theta_se <- paste0("CONMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_CONMIX_GSD_mr_pval <- paste0("CONMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      CONMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_CONMIX_GSD_theta], 
                                           thetaobs_se = est_sim[col_CONMIX_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      CONMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_CONMIX_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_CONMIX_GSD_theta_se]])
      CONMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_CONMIX_GSD_mr_pval]])  
      ## 4/ MRMIX
      col_MRMIX_GSD_theta <- paste0("MRMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_MRMIX_GSD_theta_se <- paste0("MRMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_MRMIX_GSD_mr_pval <- paste0("MRMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      MRMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MRMIX_GSD_theta], 
                                           thetaobs_se = est_sim[col_MRMIX_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      MRMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MRMIX_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_MRMIX_GSD_theta_se]])
      MRMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_MRMIX_GSD_mr_pval]])  
      ## 5/ MEDIAN
      col_MEDIAN_GSD_theta <- paste0("MEDIAN_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_MEDIAN_GSD_theta_se <- paste0("MEDIAN_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_MEDIAN_GSD_mr_pval <- paste0("MEDIAN_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      MEDIAN_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MEDIAN_GSD_theta], 
                                           thetaobs_se = est_sim[col_MEDIAN_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      MEDIAN_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MEDIAN_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_MEDIAN_GSD_theta_se]])
      MEDIAN_GSD_type1error_calc[i] <- type1error(est_sim[[col_MEDIAN_GSD_mr_pval]])  
      ## 6/ EGGER
      col_EGGER_GSD_theta <- paste0("EGGER_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_EGGER_GSD_theta_se <- paste0("EGGER_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_EGGER_GSD_mr_pval <- paste0("EGGER_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      EGGER_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_EGGER_GSD_theta], 
                                           thetaobs_se = est_sim[col_EGGER_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      EGGER_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_EGGER_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_EGGER_GSD_theta_se]])
      EGGER_GSD_type1error_calc[i] <- type1error(est_sim[[col_EGGER_GSD_mr_pval]])
      ## 7/ ROBUST
      col_ROBUST_GSD_theta <- paste0("ROBUST_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_ROBUST_GSD_theta_se <- paste0("ROBUST_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_ROBUST_GSD_mr_pval <- paste0("ROBUST_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      ROBUST_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_ROBUST_GSD_theta], 
                                           thetaobs_se = est_sim[col_ROBUST_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      ROBUST_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_ROBUST_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_ROBUST_GSD_theta_se]])
      ROBUST_GSD_type1error_calc[i] <- type1error(est_sim[[col_ROBUST_GSD_mr_pval]])
      ## 8/ PRESSO
      col_PRESSO_GSD_theta <- paste0("PRESSO_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_PRESSO_GSD_theta_se <- paste0("PRESSO_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_PRESSO_GSD_mr_pval <- paste0("PRESSO_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      PRESSO_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_PRESSO_GSD_theta], 
                                           thetaobs_se = est_sim[col_PRESSO_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      PRESSO_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_PRESSO_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_PRESSO_GSD_theta_se]])
      PRESSO_GSD_type1error_calc[i] <- type1error(est_sim[[col_PRESSO_GSD_mr_pval]])
      ## 9/ RAPS
      col_RAPS_GSD_theta <- paste0("RAPS_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_RAPS_GSD_theta_se <- paste0("RAPS_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_RAPS_GSD_mr_pval <- paste0("RAPS_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      RAPS_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_RAPS_GSD_theta], 
                                           thetaobs_se = est_sim[col_RAPS_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      RAPS_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_RAPS_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_RAPS_GSD_theta_se]])
      RAPS_GSD_type1error_calc[i] <- type1error(est_sim[[col_RAPS_GSD_mr_pval]])
      ## 10/LASSO
      col_LASSO_GSD_theta <- paste0("LASSO_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_LASSO_GSD_theta_se <- paste0("LASSO_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
      col_LASSO_GSD_mr_pval <- paste0("LASSO_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
      # calc
      LASSO_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_LASSO_GSD_theta], 
                                           thetaobs_se = est_sim[col_LASSO_GSD_theta_se], 
                                           theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
      LASSO_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_LASSO_GSD_theta]], 
                                                                           thetaobs_se = est_sim[[col_LASSO_GSD_theta_se]])
      LASSO_GSD_type1error_calc[i] <- type1error(est_sim[[col_LASSO_GSD_mr_pval]])
      #-----------
    }
    
    ## combine metrics
    IVW_GSD_performance_metrics <- IVW_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(IVW_GSD_coverage_calc, IVW_GSD_coverage_bias_eliminated_calc, IVW_GSD_type1error_calc)) %>% 
      as_tibble()
    MODE_GSD_performance_metrics <- MODE_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(MODE_GSD_coverage_calc, 
                       MODE_GSD_coverage_bias_eliminated_calc, MODE_GSD_type1error_calc)) %>% 
      as_tibble()
    CONMIX_GSD_performance_metrics <- CONMIX_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(CONMIX_GSD_coverage_calc, 
                       CONMIX_GSD_coverage_bias_eliminated_calc, CONMIX_GSD_type1error_calc)) %>% 
      as_tibble()
    MRMIX_GSD_performance_metrics <- MRMIX_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(MRMIX_GSD_coverage_calc, 
                       MRMIX_GSD_coverage_bias_eliminated_calc, MRMIX_GSD_type1error_calc)) %>% 
      as_tibble()
    #5
    MEDIAN_GSD_performance_metrics <- MEDIAN_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(MEDIAN_GSD_coverage_calc, 
                       MEDIAN_GSD_coverage_bias_eliminated_calc, MEDIAN_GSD_type1error_calc)) %>% 
      as_tibble()
    #6
    EGGER_GSD_performance_metrics <- EGGER_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(EGGER_GSD_coverage_calc, 
                       EGGER_GSD_coverage_bias_eliminated_calc, EGGER_GSD_type1error_calc)) %>% 
      as_tibble()
    #7
    ROBUST_GSD_performance_metrics <- ROBUST_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(ROBUST_GSD_coverage_calc, 
                       ROBUST_GSD_coverage_bias_eliminated_calc, ROBUST_GSD_type1error_calc)) %>% 
      as_tibble()
    #8
    PRESSO_GSD_performance_metrics <- PRESSO_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(PRESSO_GSD_coverage_calc, 
                       PRESSO_GSD_coverage_bias_eliminated_calc, PRESSO_GSD_type1error_calc)) %>% 
      as_tibble()
    #9
    RAPS_GSD_performance_metrics <- RAPS_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(RAPS_GSD_coverage_calc, 
                       RAPS_GSD_coverage_bias_eliminated_calc, RAPS_GSD_type1error_calc)) %>% 
      as_tibble()
    #10
    LASSO_GSD_performance_metrics <- LASSO_GSD_performance_metrics %>% 
      select(N, everything()) %>% 
      cbind(data.frame(LASSO_GSD_coverage_calc, 
                       LASSO_GSD_coverage_bias_eliminated_calc, LASSO_GSD_type1error_calc)) %>% 
      as_tibble()

    # #----
    # COMBINE TO PRESENT RESULTS Scenarios 2-5 ----------------------------------------------------
    # #----
    df_name <- paste0("df_combined_method_metrics_TRUE_EST_", selected_true_methods[true_method], "_GenArch_", name_GenArch)
    assign(x = df_name, value = tibble(N=n))
    
    # scenarios <- c("all_instruments", "common_strong", "common_weak", "rare_strong", "rare_weak")
    assign(x = df_name, value = eval(as.symbol(df_name)) %>% 
      left_join(IVW_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(MODE_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(CONMIX_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(MRMIX_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(MEDIAN_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(EGGER_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(ROBUST_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(PRESSO_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(RAPS_GSD_performance_metrics %>% select(-var), by = c("N"="N")) %>% 
      left_join(LASSO_GSD_performance_metrics %>% select(-var), by = c("N"="N"))
    )
    #eval(as.symbol(df_name))

    ###
    ## PLOT  -------------------------------------------------------------------
    ###
    IVW_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(IVW_GSD_performance_metrics, aes(x=N, y=IVW_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      # geom_errorbar(aes(ymin=IVW_GSD_mean-1.96*IVW_GSD_emp_se, ymax=IVW_GSD_mean+1.96*IVW_GSD_emp_se), 
      #               width=3000, # x-axis dimension
      #               position = "dodge")+
      ## ADD rainbow plots
      geom_boxplot(data = IVW_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("IVW estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> IVW_GSD_simplot_thetas
    
    ## 2 MODE
    MODE_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(MODE_GSD_performance_metrics, aes(x=N, y=MODE_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      # geom_errorbar(aes(ymin=MODE_GSD_mean-1.96*MODE_GSD_emp_se, ymax=MODE_GSD_mean+1.96*MODE_GSD_emp_se), 
      #               width=3000, # x-axis dimension
      #               position = "dodge")+
      ## ADD rainbow plots
      geom_boxplot(data = MODE_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("MODE estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> MODE_GSD_simplot_thetas
    
    ## 3 CONMIX
    CONMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=CONMIX_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = CONMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("CONMIX estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> CONMIX_GSD_simplot_thetas
    
    ## 4 MRMIX
    MRMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=MRMIX_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = MRMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("MRMIX estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> MRMIX_GSD_simplot_thetas
    
    ## 5 MEDIAN
    MEDIAN_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_MEDIAN_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(MEDIAN_GSD_performance_metrics, aes(x=N, y=MEDIAN_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = MEDIAN_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = MEDIAN_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = MEDIAN_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("MEDIAN estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> MEDIAN_GSD_simplot_thetas
    
    ## 6 EGGER
    EGGER_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_EGGER_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(EGGER_GSD_performance_metrics, aes(x=N, y=EGGER_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = EGGER_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = EGGER_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = EGGER_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("EGGER estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> EGGER_GSD_simplot_thetas
    ## 7 ROBUST
    ROBUST_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_ROBUST_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(ROBUST_GSD_performance_metrics, aes(x=N, y=ROBUST_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = ROBUST_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = ROBUST_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = ROBUST_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("ROBUST estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> ROBUST_GSD_simplot_thetas
    ## 8 PRESSO
    PRESSO_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_PRESSO_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(PRESSO_GSD_performance_metrics, aes(x=N, y=PRESSO_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = PRESSO_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = PRESSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = PRESSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("PRESSO estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> PRESSO_GSD_simplot_thetas
    ## 9 RAPS
    RAPS_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_RAPS_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(RAPS_GSD_performance_metrics, aes(x=N, y=RAPS_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = RAPS_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = RAPS_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = RAPS_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("RAPS estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> RAPS_GSD_simplot_thetas
    ## 10 LASSO
    LASSO_BMI_GSD_geom_point_dat <- est_sim %>% 
      select(!!!(colname_LASSO_theta_sim_BMI_GSD)) %>% 
      pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
      mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))
    
    ggplot(LASSO_GSD_performance_metrics, aes(x=N, y=LASSO_GSD_mean))+ 
      geom_point()+
      geom_hline(yintercept=  args.gsd.method$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      annotate(geom = "text", x = 50000, y = args.gsd.method$theta+0.05,
             label= expression( paste("ref: ", theta[X] ))  , 
             color= ggthemes_data$few$colors$Dark[9,2][[1]]
               )+
      ## ADD rainbow plots
      geom_boxplot(data = LASSO_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                   col=ggthemes_data$few$colors$Dark[3,2][[1]],
                   alpha=0.2, size=0.2, outlier.shape = NA)+
      ggdist::stat_halfeye(data = LASSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                           inherit.aes = FALSE, 
                           slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                           geom = "slabinterval",
                           na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
      gghalves::geom_half_point(
        data = LASSO_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
        inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
        side = "l", range_scale = .4, alpha = .2)+
      ## FINISH rainbow
      scale_x_continuous(labels = comma)+
      xlab("Size of Sample I")+
      ylab("Mean thetas")+
      scale_y_continuous(limits = c(0.25,0.8))+
      labs(title = "Simulated Thetas with Empirical Standard Error", 
           subtitle = paste0("LASSO estimate: BMI -> GSD\nSelected Reference Method: ", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> LASSO_GSD_simplot_thetas
    
  
    IVW_GSD_simplot_thetas$labels$title <- ""
    MODE_GSD_simplot_thetas$labels$title <- ""
    CONMIX_GSD_simplot_thetas$labels$title <- ""
    MRMIX_GSD_simplot_thetas$labels$title <- ""
    #5
    MEDIAN_GSD_simplot_thetas$labels$title <- ""
    #6
    EGGER_GSD_simplot_thetas$labels$title <- ""
    #7
    ROBUST_GSD_simplot_thetas$labels$title <- ""
    #8
    PRESSO_GSD_simplot_thetas$labels$title <- ""
    #9
    RAPS_GSD_simplot_thetas$labels$title <- ""
    #10
    LASSO_GSD_simplot_thetas$labels$title <- ""
    ggarrange(IVW_GSD_simplot_thetas,MODE_GSD_simplot_thetas,CONMIX_GSD_simplot_thetas, MRMIX_GSD_simplot_thetas, MEDIAN_GSD_simplot_thetas,
              EGGER_GSD_simplot_thetas, ROBUST_GSD_simplot_thetas, PRESSO_GSD_simplot_thetas, RAPS_GSD_simplot_thetas, LASSO_GSD_simplot_thetas,
              ncol = 1, nrow = 10) -> GSD_simplot_thetas_arrange
    GSD_simplot_thetas_arrange
    
    if (SAVE.files ==TRUE ){
      filename_GSD_simplot_thetas_arrange <- paste0("./output/plots/theta-sims/", Sys.Date(), "_GSD_simplot_thetas_REF_METHOD_", selected_true_methods[true_method],
                                                    "_GenArch_", name_GenArch, "_NSIM_", nsim,".svg")
      ggsave(filename = filename_GSD_simplot_thetas_arrange, plot = GSD_simplot_thetas_arrange, 
             width = 10, height = 38, 
             units = "in"  # default
             )
    }
    
    ### GSD Bias & MSE
    #BIAS
    ggplot(IVW_GSD_performance_metrics, aes(x=N, y=IVW_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "IVW estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_IVW_GSD
    #MSE
    ggplot(IVW_GSD_performance_metrics, aes(x=N, y=IVW_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "IVW estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_IVW_GSD
    ggarrange(biasplot_IVW_GSD, mseplot_IVW_GSD) -> bias_mse_plot_IVW_GSD
    
    ## MODE
    ggplot(MODE_GSD_performance_metrics, aes(x=N, y=MODE_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "MODE estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_MODE_GSD
    ggplot(MODE_GSD_performance_metrics, aes(x=N, y=MODE_GSD_mse))+ 
      geom_point()+
        geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "MODE estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_MODE_GSD
    ggarrange(biasplot_MODE_GSD, mseplot_MODE_GSD) -> bias_mse_plot_MODE_GSD
    
    ## CONMIX
    ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=CONMIX_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "CONMIX estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_CONMIX_GSD
    ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=CONMIX_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "CONMIX estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_CONMIX_GSD
    ggarrange(biasplot_CONMIX_GSD, mseplot_CONMIX_GSD) -> bias_mse_plot_CONMIX_GSD
    
    ## MRMIX
    ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=MRMIX_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "MRMIX estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_MRMIX_GSD
    ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=MRMIX_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "MRMIX estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_MRMIX_GSD
    ggarrange(biasplot_MRMIX_GSD, mseplot_MRMIX_GSD) -> bias_mse_plot_MRMIX_GSD
    
    ## 5 MEDIAN
    ggplot(MEDIAN_GSD_performance_metrics, aes(x=N, y=MEDIAN_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "MEDIAN estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_MEDIAN_GSD
    ggplot(MEDIAN_GSD_performance_metrics, aes(x=N, y=MEDIAN_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "MEDIAN estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_MEDIAN_GSD
    ggarrange(biasplot_MEDIAN_GSD, mseplot_MEDIAN_GSD) -> bias_mse_plot_MEDIAN_GSD
    ## 6 EGGER
    ggplot(EGGER_GSD_performance_metrics, aes(x=N, y=EGGER_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "EGGER estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_EGGER_GSD
    ggplot(EGGER_GSD_performance_metrics, aes(x=N, y=EGGER_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "EGGER estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_EGGER_GSD
    ggarrange(biasplot_EGGER_GSD, mseplot_EGGER_GSD) -> bias_mse_plot_EGGER_GSD
    ## 7 ROBUST
    ggplot(ROBUST_GSD_performance_metrics, aes(x=N, y=ROBUST_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "ROBUST estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_ROBUST_GSD
    ggplot(ROBUST_GSD_performance_metrics, aes(x=N, y=ROBUST_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "ROBUST estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_ROBUST_GSD
    ggarrange(biasplot_ROBUST_GSD, mseplot_ROBUST_GSD) -> bias_mse_plot_ROBUST_GSD
    ## 8 PRESSO
    ggplot(PRESSO_GSD_performance_metrics, aes(x=N, y=PRESSO_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "PRESSO estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_PRESSO_GSD
    ggplot(PRESSO_GSD_performance_metrics, aes(x=N, y=PRESSO_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "PRESSO estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_PRESSO_GSD
    ggarrange(biasplot_PRESSO_GSD, mseplot_PRESSO_GSD) -> bias_mse_plot_PRESSO_GSD
    ## 9 RAPS
    ggplot(RAPS_GSD_performance_metrics, aes(x=N, y=RAPS_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "RAPS estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_RAPS_GSD
    ggplot(RAPS_GSD_performance_metrics, aes(x=N, y=RAPS_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "RAPS estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_RAPS_GSD
    ggarrange(biasplot_RAPS_GSD, mseplot_RAPS_GSD) -> bias_mse_plot_RAPS_GSD
    ## LASSO
    ggplot(LASSO_GSD_performance_metrics, aes(x=N, y=LASSO_GSD_bias))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(-1.5,1.5))+
      xlab("Size of Sample I")+
      ylab("Bias")+
      labs(title = "LASSO estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> biasplot_LASSO_GSD
    ggplot(LASSO_GSD_performance_metrics, aes(x=N, y=LASSO_GSD_mse))+ 
      geom_point()+
      geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
      scale_x_continuous(labels = comma)+
      scale_y_continuous(limits = c(0,1.8))+
      xlab("Size of Sample I")+
      ylab("MSE")+
      labs(title = "LASSO estimate: BMI -> GSD",
           caption = paste0("Selected Reference Method:\n", 
                             selected_true_methods[true_method], 
                             " estimate calculated with total sample size\nGenetic Architecture: ", name_GenArch))+
      theme_few() -> mseplot_LASSO_GSD
    ggarrange(biasplot_LASSO_GSD, mseplot_LASSO_GSD) -> bias_mse_plot_LASSO_GSD
    
    
    ggarrange(bias_mse_plot_IVW_GSD, bias_mse_plot_MODE_GSD, bias_mse_plot_CONMIX_GSD, bias_mse_plot_MRMIX_GSD, bias_mse_plot_MEDIAN_GSD,
              bias_mse_plot_EGGER_GSD, bias_mse_plot_ROBUST_GSD, bias_mse_plot_PRESSO_GSD, bias_mse_plot_RAPS_GSD, bias_mse_plot_LASSO_GSD,
              ncol = 1, nrow = 10, common.legend = TRUE) -> bias_mse_plot_GSD
    bias_mse_plot_GSD
    
    if (SAVE.files ==TRUE ){
      filename_simplot_bias_mse <- paste0("./output/plots/bias-mse/", Sys.Date(), "_GSD_simplot_bias_mse_REF_METHOD_", selected_true_methods[true_method], 
                                          "_GenArch_", name_GenArch, "_NSIM_", nsim,".svg")
      ggsave(filename = filename_simplot_bias_mse, plot = bias_mse_plot_GSD, 
             width = 7, height = 24, 
             units = "in"  # default
             )
    }
  }
}
```

\clearpage

# Create final tables and figures

## Tables

For all scenarios (a. all instruments, b. common+strong, c. common+weak, d. rare+strong, e. rare+weak) we need to describe our results for a smaller set of sample sizes. We create a list that holds a dataframe for each reference method (1:5). Each dataframe contains results from all scenarios.

```{r}
# we have assigned dataframes in the loop above!
df_name <- paste0("df_combined_method_metrics_TRUE_EST_", selected_true_methods[true_method])
df_name <- paste0("df_combined_method_metrics_TRUE_EST_", selected_true_methods[true_method], "_GenArch_", name_GenArch)

df_metrics_all_scenarios <- list()
for(true_method in seq_along(selected_true_methods)) {
  # scenario 1) all IVs
  df_name <- paste0("df_combined_method_metrics_TRUE_EST_", selected_true_methods[true_method])
  df_metrics_all_scenarios[[true_method]] <- eval(as.symbol(df_name)) %>% 
    rename_with(.fn = ~ paste0(.x, "_ALL_INSTR"), .cols = -"N")
  
  # scenarios 2)-5) GenArchs
  for (my_architecture in seq_along(est_sim_architectures)) {
    name_GenArch <- gsub(pattern = "_est_sim", replacement = "", names(est_sim_architectures)[my_architecture])
    df_name <- paste0("df_combined_method_metrics_TRUE_EST_", selected_true_methods[true_method], "_GenArch_", name_GenArch)
    
    df_metrics_all_scenarios[[true_method]] <- df_metrics_all_scenarios[[true_method]] %>% 
      left_join(eval(as.symbol(df_name)) %>% 
                  rename_with(.fn = ~ paste0(.x, "_", name_GenArch), .cols = -"N"), by=c("N"="N"))
  }
}

# name list
names(df_metrics_all_scenarios) <- paste0("ref_", selected_true_methods)
```


```{r gt-summary-tabs}
# https://rfortherestofus.com/2019/11/how-to-make-beautiful-tables-in-r/
library(gt)
df.temp.long <- list()
df.temp.wide <- list()

sample_size_subset <- c(50000, 100000, 200000, 300000)

for(ref.method in seq_along(df_metrics_all_scenarios)) {
  
  ## MEAN ## ---------------------------------------------------------------
  # later plot N=50,000 all methods and different scenarios
  
  ## CREATE LONG FORMAT
  df_metrics_all_scenarios[[ref.method]] %>% 
    filter(N %in% sample_size_subset) %>% 
    select(N, contains("_mean_") & 
             (contains("_ALL_INSTR") | contains("_STRONG_COMMON") | 
                contains("STRONG_RARE") | contains("WEAK_COMMON") | contains("WEAK_RARE"))
           ) %>% 
    #mutate(N = format(N, scientific=FALSE)) %>% 
    pivot_longer(cols=!N, names_to="Method", values_to="Mean")  %>% 
    separate(
      col="Method", into = c("Method", "Scenario"), sep = "_GSD_mean_"
    ) -> df.temp.long[[ref.method]]
  
  ## EMP SE ## ---------------------------------------------------------------
  df_metrics_all_scenarios[[ref.method]] %>% 
    filter(N %in% sample_size_subset) %>% 
    select(N, contains("_emp_se_") & 
             (contains("_ALL_INSTR") | contains("_STRONG_COMMON") | 
                contains("STRONG_RARE") | contains("WEAK_COMMON") | contains("WEAK_RARE"))
           ) %>% 
    #mutate(N = format(N, scientific=FALSE)) %>% 
    pivot_longer(cols=!N, names_to="Method", values_to="emp_se")  %>% 
    separate(
      col="Method", into = c("Method", "Scenario"), sep = "_GSD_emp_se_"
    )  -> temp.dat
  
  df.temp.long[[ref.method]] <- df.temp.long[[ref.method]] %>% 
    left_join(temp.dat, by=c("Method"="Method", "Scenario"="Scenario", "N"="N"))
  
  ## BIAS ## ---------------------------------------------------------------
  df_metrics_all_scenarios[[ref.method]] %>% 
    filter(N %in% sample_size_subset) %>% 
    select(N, contains("GSD_bias_") & 
             (contains("_ALL_INSTR") | contains("_STRONG_COMMON") | 
                contains("STRONG_RARE") | contains("WEAK_COMMON") | contains("WEAK_RARE"))
           ) %>% 
    #mutate(N = format(N, scientific=FALSE)) %>% 
    pivot_longer(cols=!N, names_to="Method", values_to="bias")  %>% 
    separate(
      col="Method", into = c("Method", "Scenario"), sep = "_GSD_bias_"
    )  -> temp.dat
  
  df.temp.long[[ref.method]] <- df.temp.long[[ref.method]] %>% 
    left_join(temp.dat, by=c("Method"="Method", "Scenario"="Scenario", "N"="N"))
  
  ## MSE ## ---------------------------------------------------------------
  df_metrics_all_scenarios[[ref.method]] %>% 
    filter(N %in% sample_size_subset) %>% 
    select(N, contains("_mse_") & 
             (contains("_ALL_INSTR") | contains("_STRONG_COMMON") | 
                contains("STRONG_RARE") | contains("WEAK_COMMON") | contains("WEAK_RARE"))
           ) %>% 
    #mutate(N = format(N, scientific=FALSE)) %>% 
    pivot_longer(cols=!N, names_to="Method", values_to="mse")  %>% 
    separate(
      col="Method", into = c("Method", "Scenario"), sep = "_GSD_mse_"
    )  -> temp.dat
  
  df.temp.long[[ref.method]] <- df.temp.long[[ref.method]] %>% 
    left_join(temp.dat, by=c("Method"="Method", "Scenario"="Scenario", "N"="N"))
  
  ## COVERAGE ## ---------------------------------------------------------------
  df_metrics_all_scenarios[[ref.method]] %>% 
    filter(N %in% sample_size_subset) %>% 
    select(N, contains("GSD_coverage_calc_") & 
             (contains("_ALL_INSTR") | contains("_STRONG_COMMON") | 
                contains("STRONG_RARE") | contains("WEAK_COMMON") | contains("WEAK_RARE"))
           ) %>% 
    #mutate(N = format(N, scientific=FALSE)) %>% 
    pivot_longer(cols=!N, names_to="Method", values_to="coverage")  %>% 
    separate(
      col="Method", into = c("Method", "Scenario"), sep = "_GSD_coverage_calc_"
    )  -> temp.dat
  
  df.temp.long[[ref.method]] <- df.temp.long[[ref.method]] %>% 
    left_join(temp.dat, by=c("Method"="Method", "Scenario"="Scenario", "N"="N"))
  
  ## TYPE1ERROR ## ---------------------------------------------------------------
  df_metrics_all_scenarios[[ref.method]] %>% 
    filter(N %in% sample_size_subset) %>% 
    select(N, contains("GSD_type1error_calc_") & 
             (contains("_ALL_INSTR") | contains("_STRONG_COMMON") | 
                contains("STRONG_RARE") | contains("WEAK_COMMON") | contains("WEAK_RARE"))
           ) %>% 
    #mutate(N = format(N, scientific=FALSE)) %>% 
    pivot_longer(cols=!N, names_to="Method", values_to="type1error")  %>% 
    separate(
      col="Method", into = c("Method", "Scenario"), sep = "_GSD_type1error_calc_"
    )  -> temp.dat
  
  df.temp.long[[ref.method]] <- df.temp.long[[ref.method]] %>% 
    left_join(temp.dat, by=c("Method"="Method", "Scenario"="Scenario", "N"="N"))
  
  #--
  ### CREATE WIDE FORMAT -------------------------------------------------------
  #--
  
  ## only one variable 
  #df.temp.long[[ref.method]] %>% 
  #  select(N, Method, Scenario, Mean) %>% 
  #  pivot_wider(names_from = "Scenario", values_from = "Mean", names_glue = "Mean_{Scenario}") 
  
  ## several variables
  df.temp.long[[ref.method]] %>% 
    select(N, Method, Scenario, Mean, emp_se) %>% 
    pivot_wider(names_from = Scenario, values_from = c(Mean, emp_se),  
                #names_vary = "slowest",  # naming scheme
                names_glue = "{.value}_{Scenario}") -> df.temp.wide[[ref.method]]
  
}

# name list
names(df.temp.long) <- paste0("ref_", selected_true_methods)



gt.tables.mean.se <- list()
gt.tables.performance2 <- list()

## loop to create tables
for (ref.method in seq_along(selected_true_methods)) {
  
  ## TABLE 01: MEAN +- SE -----------------------------------------------------
  df.temp.long[[ref.method]] %>% 
    select(N, Method, Scenario, Mean, emp_se) %>% 
    # first pivot wider
    pivot_wider(names_from = Scenario, values_from = c(Mean, emp_se),  
                  #names_vary = "slowest",  # naming scheme
                  names_glue = "{.value}_{Scenario}") %>% 
    mutate(N = paste0("Sample Size = ", format(N, scientific=FALSE))) %>% 
    gt(groupname_col="N") %>% 
    tab_header(
      title = md("**Mean estimate with corresponding empirical standard errors**"),
      subtitle = md(paste0("Reference estimate: ", selected_true_methods[ref.method]))
    ) %>% 
    tab_spanner(
      label = "Mean [\u00b1 Empirical Standard Error]", 
      columns = 3:7
    ) %>% 
    cols_label(
      Mean_ALL_INSTR = md("**Scenario 1**"),
      Mean_STRONG_COMMON = md("**Scenario 2**"),
      Mean_STRONG_RARE = md("**Scenario 3**"),
      Mean_WEAK_COMMON = md("**Scenario 4**"),
      Mean_WEAK_RARE = md("**Scenario 5**"),
    ) %>% 
    cols_merge(
      columns = c(Mean_ALL_INSTR, emp_se_ALL_INSTR),
      pattern = "{1} [\u00b1 {2}]"  # +- sign : "\u00b1"
    ) %>% 
    cols_merge(
      columns = c(Mean_STRONG_COMMON, emp_se_STRONG_COMMON),
      pattern = "{1} [\u00b1 {2}]"  # +- sign : "\u00b1"
    ) %>% 
    cols_merge(
      columns = c(Mean_STRONG_RARE, emp_se_STRONG_RARE),
      pattern = "{1} [\u00b1 {2}]"  # +- sign : "\u00b1"
    ) %>% 
    cols_merge(
      columns = c(Mean_WEAK_COMMON, emp_se_WEAK_COMMON),
      pattern = "{1} [\u00b1 {2}]"  # +- sign : "\u00b1"
    ) %>% 
    cols_merge(
      columns = c(Mean_WEAK_RARE, emp_se_WEAK_RARE),
      pattern = "{1} [\u00b1 {2}]"  # +- sign : "\u00b1"
    ) %>% 
    fmt_number(columns = 3:12,decimals = 3, use_seps=TRUE) %>% 
    tab_source_note(
      source_note = md("*Scenario 1* corresponds to all genetic instruments being used; 
                       *Scenario 2* uses relatively strong and common variants;
                       *Scenario 3* uses relatively strong and rare variants; 
                       *Scenario 4* uses relatively weak and common variants;
                       *Scenario 5* uses relatively weak and rare variants")
    ) %>% 
    opt_stylize(style=1, color="gray") -> gt.tables.mean.se[[ref.method]]
  
  if (SAVE.files ==TRUE ){
    filename.gt.mean.se <- paste0("./output/gt-tables/sim-results/mean-empse/", Sys.Date(), 
                                          "_mean_emp_se_performance_metrics_TRUE_EST_", 
                                          selected_true_methods[ref.method], 
                                          "_NSIM_", nsim)
    
    gt.tables.mean.se[[ref.method]] %>% 
      gtsave(paste0(filename.gt.mean.se, ".docx"))   # alternative .rtf, .tex, .png, .html
    gt.tables.mean.se[[ref.method]] %>% 
      gtsave(paste0(filename.gt.mean.se, ".html")) 
    gt.tables.mean.se[[ref.method]] %>% 
      gtsave(paste0(filename.gt.mean.se, ".tex")) 
  }
  
  # TABLE 02: BIAS, MSE, COVERAGE, TYPE1ERROR ---------------------------------
  df.temp.long[[ref.method]] %>%
    select(N, Method, Scenario, bias, mse, coverage, type1error) %>% 
    # first pivot wider
    pivot_wider(names_from = Scenario, values_from = c(bias, mse, coverage, type1error),  
                  #names_vary = "slowest",  # naming scheme
                  names_glue = "{.value}_{Scenario}") %>% 
    mutate(N = paste0("Sample Size = ", format(N, scientific=FALSE))) %>% 
    gt(groupname_col="N") %>% 
    tab_header(
      title = md("**Performance measures**"),
      subtitle = md(paste0("Reference estimate: ", selected_true_methods[ref.method]))
    ) %>% 
    tab_spanner(
      label = "Bias", 
      columns = 3:7
    ) %>% 
    tab_spanner(
      label = "MSE", 
      columns = 8:12
    ) %>% 
    tab_spanner(
      label = "Coverage", 
      columns = 13:17
    ) %>% 
    tab_spanner(
      label = "Type 1 Error", 
      columns = 18:22
    ) %>% 
    cols_label(
      bias_ALL_INSTR = md("**Scenario 1**"),
      bias_STRONG_COMMON = md("**Scenario 2**"),
      bias_STRONG_RARE = md("**Scenario 3**"),
      bias_WEAK_COMMON = md("**Scenario 4**"),
      bias_WEAK_RARE = md("**Scenario 5**"),
    ) %>% 
    cols_label(
      mse_ALL_INSTR = md("**Scenario 1**"),
      mse_STRONG_COMMON = md("**Scenario 2**"),
      mse_STRONG_RARE = md("**Scenario 3**"),
      mse_WEAK_COMMON = md("**Scenario 4**"),
      mse_WEAK_RARE = md("**Scenario 5**"),
    ) %>% 
    cols_label(
      coverage_ALL_INSTR = md("**Scenario 1**"),
      coverage_STRONG_COMMON = md("**Scenario 2**"),
      coverage_STRONG_RARE = md("**Scenario 3**"),
      coverage_WEAK_COMMON = md("**Scenario 4**"),
      coverage_WEAK_RARE = md("**Scenario 5**"),
    ) %>% 
    cols_label(
      type1error_ALL_INSTR = md("**Scenario 1**"),
      type1error_STRONG_COMMON = md("**Scenario 2**"),
      type1error_STRONG_RARE = md("**Scenario 3**"),
      type1error_WEAK_COMMON = md("**Scenario 4**"),
      type1error_WEAK_RARE = md("**Scenario 5**"),
    ) %>% 
    fmt_number(columns = 3:22,decimals = 3, use_seps=TRUE) %>% 
    tab_source_note(
      source_note = md("*Scenario 1* corresponds to all genetic instruments being used; 
                       *Scenario 2* uses relatively strong and common variants;
                       *Scenario 3* uses relatively strong and rare variants; 
                       *Scenario 4* uses relatively weak and common variants;
                       *Scenario 5* uses relatively weak and rare variants")
    ) %>% 
    opt_stylize(style=1, color="gray") -> gt.tables.performance2[[ref.method]]
  
    if (SAVE.files ==TRUE ){
      filename.gt.performance2 <- paste0("./output/gt-tables/sim-results/", Sys.Date(), 
                                            "_performance2_metrics_TRUE_EST_", 
                                            selected_true_methods[ref.method], 
                                            "_NSIM_", nsim)
      
      gt.tables.performance2[[ref.method]] %>% 
        gtsave(paste0(filename.gt.performance2, ".docx"))   # alternative .rtf, .tex, .png, .html
      gt.tables.performance2[[ref.method]] %>% 
        gtsave(paste0(filename.gt.performance2, ".html")) 
      gt.tables.performance2[[ref.method]] %>% 
        gtsave(paste0(filename.gt.performance2, ".tex")) 
    }
}

```

\clearpage

## Summary Plots


```{r}
dodged.barplot.mse <- list()
dodged.barplot.mse.scenario <- list()  # store plots for 1 ref-estimate, all Scenarios
dodged.barplot.mse.references <- list()  # store plots for ALL_INSTR for all different ref-estimates
dodged.barplot.bias <- list()
dodged.barplot.bias.scenario <- list()
dodged.barplot.bias.references <- list()
dodged.barplot.coverage <- list()
dodged.barplot.power<- list()

scenarios <- c("ALL_INSTR", "STRONG_COMMON", "STRONG_RARE", "WEAK_COMMON", "WEAK_RARE")

for(ref.method in seq_along(df_metrics_all_scenarios)) {
  ## --
  # CREATE PLOTS -------------------------------------------------------------
  ## --
  
  ## MSE PLOT FACET WRAP ------------------------------------------------------
  
  # plot per scenario
  for(scenario in seq_along(scenarios)) {
    
    # customize axis 
    if (scenarios[scenario] != "WEAK_RARE") {
      my.y.lim <- 0.15
    } else {
      my.y.lim <- 0.3
    }
    
    df.temp.long[[ref.method]] %>%
      mutate(N=paste0("Sample Size: ", format(N, scientific=FALSE))) %>% 
      mutate(label.too.large = ifelse(mse>my.y.lim, paste0("MSE \u2265 ", my.y.lim) , NA)) %>% 
      filter(Scenario == scenarios[scenario]) %>% 
      ## DROP NA (e.g. no sample sizes 50.000)
      drop_na(Mean) %>%  
      ggplot(aes(x=Method, y=mse)) +
      geom_bar(stat="identity", position=position_dodge2(reverse=TRUE),
               #color=ggthemes_data$few$colors$Dark[2,2][[1]],
               ) +   # reverse dodge
      geom_text(aes(x= Method, y= my.y.lim, label = label.too.large,
                    colour=label.too.large), #position = position_dodge(),
                nudge_y = -0.1, #size = 2.25, angle = 90
                color=ggthemes_data$few$colors$Dark[4,2][[1]],
                fontface="bold"
                )+
      facet_wrap(~N, nrow = 4) + 
      xlab("Method")+
      ylab(paste0("Mean Squared Error with nsim=", nsim))+
      labs(title = paste0("Selected Reference Method: ", 
                              selected_true_methods[ref.method]),
           subtitle = paste0("Genetic Architecture: ", scenarios[scenario])
           )+
      #ylim(0, 0.5)+
      coord_flip(ylim=c(0, my.y.lim))+
      ggthemes::theme_few()+  # remove legend
      ggthemes::scale_fill_few() -> dodged.barplot.mse.scenario[[scenario]]
    
      if (SAVE.files ==TRUE ){
        ## arranged plots are toooo small 
        filename.dodged.barplot.mse.scenario <- paste0("./output/plots/summary-plots/scenario_",scenarios[scenario],"/",
                                              Sys.Date(), "_mse_GSD_dodged_barplot_REF_METHOD_",
                                              selected_true_methods[ref.method], "_NSIM_", nsim)
        ggsave(filename = paste0(filename.dodged.barplot.mse.scenario, ".svg"), 
               plot = dodged.barplot.mse.scenario[[scenario]], 
               width = 5, height = 8, 
               units = "in"  # default
               )
      }
  }
  dodged.barplot.mse.references[ref.method] <- dodged.barplot.mse.scenario[1]
  
  # scenarios without all instruments in arranged plot
  arrangedplot.mse.scenarios <- 
    ggarrange(plotlist = dodged.barplot.mse.scenario[-1], labels = c("A", "B", "C", "D", "E", "F"))
  if (SAVE.files ==TRUE ){
        ## arranged plots are toooo small 
        filename.arrangedplot.mse.scenarios <- paste0("./output/plots/summary-plots/arranged_scenarios/",
                                              Sys.Date(), "_mse_GSD_dodged_barplot_REF_METHOD_",
                                              selected_true_methods[ref.method], "_NSIM_", nsim)
        ggsave(filename = paste0(filename.arrangedplot.mse.scenarios, ".svg"), 
               plot = arrangedplot.mse.scenarios, 
               width = 10, height = 12, 
               units = "in"  # default
               )
  }
  
  # with all scenarios in 1 plot stacked
  df.temp.long[[ref.method]] %>%
    mutate(N=paste0("Sample Size: ", format(N, scientific=FALSE))) %>% 
    mutate(label.too.large = ifelse(mse>0.5, "MSE \u2265 0.5", NA)) %>% 
    ggplot(aes(x=Method, y=mse, fill=Scenario )) +
    geom_bar(stat="identity", position=position_dodge2(reverse=TRUE)) +   # reverse dodge
    geom_text(aes(x= Method, y= 0.4, label = label.too.large),# position = position_dodge(),
              #size = 2.25, angle = 90
              )+
    facet_wrap(~N) + 
    xlab("Method")+
    ylab(paste0("Mean Squared Error with nsim=", nsim))+
    labs(title = paste0("Selected Reference Method: ", 
                            selected_true_methods[ref.method])
         )+
    #ylim(0, 0.5)+
    coord_flip(ylim=c(0,0.5))+
    ggthemes::theme_few()+
    ggthemes::scale_fill_few() -> dodged.barplot.mse[[ref.method]]
  
  if (SAVE.files ==TRUE ){
    ## arranged plots are toooo small 
    filename.dodged.barplot.mse <- paste0("./output/plots/summary-plots/MSE_stacked/", Sys.Date(),
                                          "_mse_GSD_dodged_barplot_REF_METHOD_",
                                          selected_true_methods[ref.method], "_NSIM_", nsim)
    ggsave(filename = paste0(filename.dodged.barplot.mse, ".svg"), plot =  dodged.barplot.mse[[ref.method]], 
             width = 12, height = 8, 
             units = "in"  # default
             )
  }
  
  ## BIAS PLOT FACET WRAP ------------------------------------------------------
  
  # plot per scenario
  for(scenario in seq_along(scenarios)) {
    
    # customize axis 
    if (scenarios[scenario] != "WEAK_RARE") {
      my.y.lim <- 0.15
    } else {
      my.y.lim <- 0.3
    }
    
    df.temp.long[[ref.method]] %>%
      mutate(N=paste0("Sample Size: ", format(N, scientific=FALSE))) %>% 
      filter(Scenario == scenarios[scenario]) %>% 
      ## DROP NA (e.g. no sample sizes 50.000)
      drop_na(Mean) %>%  
      ggplot(aes(x=Method, y=bias)) +
      geom_bar(stat="identity", position=position_dodge2(reverse=TRUE),
               #color=ggthemes_data$few$colors$Dark[2,2][[1]],
               ) +   # reverse dodge
      geom_hline(yintercept = 0,  col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
      facet_wrap(~N, nrow = 4) + 
      xlab("Method")+
      ylab(paste0("Bias with nsim=", nsim))+
      labs(title = paste0("Selected Reference Method: ", 
                              selected_true_methods[ref.method]),
           subtitle = paste0("Genetic Architecture: ", scenarios[scenario])
           )+
      #ylim(0, 0.5)+
      coord_flip()+
      ggthemes::theme_few()+  # remove legend
      ggthemes::scale_fill_few() -> dodged.barplot.bias.scenario[[scenario]]
    
      if (SAVE.files ==TRUE ){
        ## arranged plots are toooo small 
        filename.dodged.barplot.bias.scenario <- paste0("./output/plots/summary-plots/scenario_",scenarios[scenario],"/",
                                              Sys.Date(), "_bias__GSD_dodged_barplot_REF_METHOD_",
                                              selected_true_methods[ref.method], "_NSIM_", nsim)
        ggsave(filename = paste0(filename.dodged.barplot.bias.scenario, ".svg"), 
               plot = dodged.barplot.bias.scenario[[scenario]], 
               width = 5, height = 8, 
               units = "in"  # default
               )
      }
  }
  dodged.barplot.bias.references[ref.method] <- dodged.barplot.bias.scenario[1]
  
   # scenarios without all instruments in arranged plot
  arrangedplot.bias.scenarios <- 
    ggarrange(plotlist = dodged.barplot.bias.scenario[-1], labels = c("A", "B", "C", "D", "E", "F"))
  if (SAVE.files ==TRUE ){
        ## arranged plots are toooo small 
        filename.arrangedplot.bias.scenarios <- paste0("./output/plots/summary-plots/arranged_scenarios/",
                                              Sys.Date(), "_bias_GSD_dodged_barplot_REF_METHOD_",
                                              selected_true_methods[ref.method], "_NSIM_", nsim)
        ggsave(filename = paste0(filename.arrangedplot.bias.scenarios, ".svg"), 
               plot = arrangedplot.bias.scenarios, 
               width = 10, height = 12, 
               units = "in"  # default
               )
  }
  
  # stacked
  df.temp.long[[ref.method]] %>%
    mutate(N=paste0("Sample Size: ", format(N, scientific=FALSE))) %>% 
    ggplot(aes(x=Method, y=bias, fill=Scenario )) +
    geom_bar(stat="identity", position=position_dodge2(reverse=TRUE)) +   # reverse dodge
    geom_hline(yintercept = 0,  col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
    facet_wrap(~N) + 
    xlab("Method")+
    ylab(paste0("Bias with nsim=", nsim))+
    labs(title = paste0("Selected Reference Method: ", 
                            selected_true_methods[ref.method])
         )+
    coord_flip()+
    ggthemes::theme_few()+
    ggthemes::scale_fill_few() -> dodged.barplot.bias[[ref.method]]
  
  if (SAVE.files ==TRUE ){
    ## arranged plots are toooo small 
    filename.dodged.barplot.bias <- paste0("./output/plots/summary-plots/BIAS_stacked/", Sys.Date(),
                                           "_bias_GSD_dodged_barplot_REF_METHOD_",
                                          selected_true_methods[ref.method], "_NSIM_", nsim)
    ggsave(filename = paste0(filename.dodged.barplot.bias, ".svg"), plot =  dodged.barplot.bias[[ref.method]], 
             width = 12, height = 8, 
             units = "in"  # default
             )
  }
  
  ## COVERAGE PLOT FACET WRAP ------------------------------------------------------
  # stacked
  df.temp.long[[ref.method]] %>%
    mutate(N=paste0("Sample Size: ", format(N, scientific=FALSE))) %>% 
    ggplot(aes(x=Method, y=coverage, fill=Scenario )) +
    geom_bar(stat="identity", position=position_dodge2(reverse=TRUE)) +   # reverse dodge
    geom_hline(yintercept = 0.95,  col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
    facet_wrap(~N) + 
    xlab("Method")+
    ylab(paste0("Coverage with nsim=", nsim))+
    labs(title = paste0("Selected Reference Method: ", 
                            selected_true_methods[ref.method])
         )+
    coord_flip()+
    ggthemes::theme_few()+
    ggthemes::scale_fill_few() -> dodged.barplot.coverage[[ref.method]]
  
  if (SAVE.files ==TRUE ){
    ## arranged plots are toooo small 
    filename.dodged.barplot.coverage <- paste0("./output/plots/summary-plots/COVERAGE_stacked/", Sys.Date(),
                                               "_coverage_GSD_dodged_barplot_REF_METHOD_",
                                          selected_true_methods[ref.method], "_NSIM_", nsim)
    ggsave(filename = paste0(filename.dodged.barplot.coverage, ".svg"), plot =  dodged.barplot.coverage[[ref.method]], 
             width = 12, height = 8, 
             units = "in"  # default
             )
  }
  
  ## Power of 5% level test (not TYPE1ERROR) PLOT FACET WRAP ------------------------------------------------------
  # stacked
  df.temp.long[[ref.method]] %>%
    mutate(N=paste0("Sample Size: ", format(N, scientific=FALSE))) %>% 
    ggplot(aes(x=Method, y=type1error, fill=Scenario )) +
    geom_bar(stat="identity", position=position_dodge2(reverse=TRUE)) +   # reverse dodge
    geom_hline(yintercept = 0.95,  col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
    facet_wrap(~N) + 
    xlab("Method")+
    ylab(paste0("Power of 5% Level Test with nsim=", nsim))+
    labs(title = paste0("Selected Reference Method: ", 
                            selected_true_methods[ref.method])
         )+
    coord_flip()+
    ggthemes::theme_few()+
    ggthemes::scale_fill_few() -> dodged.barplot.power[[ref.method]]
  
  if (SAVE.files ==TRUE ){
    ## arranged plots are toooo small 
    filename.dodged.barplot.power <- paste0("./output/plots/summary-plots/POWER_stacked/", Sys.Date(), "_power_GSD_dodged_barplot_REF_METHOD_",
                                          selected_true_methods[ref.method], "_NSIM_", nsim)
    ggsave(filename = paste0(filename.dodged.barplot.power, ".svg"), plot =  dodged.barplot.power[[ref.method]], 
             width = 12, height = 8, 
             units = "in"  # default
             )
  }
}
  
  # scenarios without all instruments in arranged plot

dodged.barplot.mse.references.update <- map(seq_along(dodged.barplot.mse.references), ~ {
  dodged.barplot.mse.references[[.x]][["labels"]][["subtitle"]] <- NULL
  return(dodged.barplot.mse.references[[.x]])
} )

arrangedplot.mse.references <- 
    ggarrange(plotlist = dodged.barplot.mse.references.update, labels = c("A", "B", "C", "D", "E", "F"), ncol=3, nrow=2)

if (SAVE.files ==TRUE ){
    ## arranged plots are toooo small 
    filename.dodged.barplot.mse <- paste0("./output/plots/summary-plots/arranged_references/", Sys.Date(), "_mse_GSD_dodged_barplot_REF_METHOD_",
                                          selected_true_methods[ref.method], "_NSIM_", nsim)
    ggsave(filename = paste0(filename.dodged.barplot.mse, ".svg"), plot =  arrangedplot.mse.references, 
             width = 14, height = 14, 
             units = "in"  # default
             )
}


dodged.barplot.bias.references.update <- map(seq_along(dodged.barplot.bias.references), ~ {
  dodged.barplot.bias.references[[.x]][["labels"]][["subtitle"]] <- NULL
  return(dodged.barplot.bias.references[[.x]])
} )

arrangedplot.bias.references <- 
    ggarrange(plotlist = dodged.barplot.bias.references.update, labels = c("A", "B", "C", "D", "E", "F"), ncol=3, nrow=2)

if (SAVE.files ==TRUE ){
    ## arranged plots are toooo small 
    filename.dodged.barplot.bias <- paste0("./output/plots/summary-plots/arranged_references/", Sys.Date(), "_bias_GSD_dodged_barplot_REF_METHOD_",
                                          selected_true_methods[ref.method], "_NSIM_", nsim)
    ggsave(filename = paste0(filename.dodged.barplot.bias, ".svg"), plot =  arrangedplot.bias.references, 
             width = 14, height = 14, 
             units = "in"  # default
             )
}

```

