---
title: "`r params$doc_title`"
author: | 
        | christoph.reich@med.uni-heidelberg.de 
        | 
date: "`r format(Sys.time(), '%d %B, %Y, %H:%M')`"
params:
  date: !r Sys.Date()
  doc_title: "MSc Default Title"
output: 
  html_document:
    code_folding: hide
    theme: paper
    toc: true
    number_sections: true
    toc_float: true
    toc_depth: 4
    df_print: kable
    latex_engine: xelatex
bibliography: ../references/references.bib
editor_options: 
  chunk_output_type: inline
---


```{r setup, include=FALSE}
#knitr::opts_chunk$set(echo = FALSE)
figurespath <- paste0("../reports/", format(Sys.time(), "%Y%m%d"), '_figures/')
knitr::opts_chunk$set(fig.path = figurespath)  # reference is script! also creates new dirs

library(dplyr)
library(purrr)
library(tidyr)
library(stringr)
library(tibble)
library(ggplot2)
library(ggdist)
library(gghalves)
library(ggthemes)
library(ggpubr)
library(scales)
#library(biomaRt)  # different script
library(kableExtra)
library(MendelianRandomization)
library(TwoSampleMR)
library(mr.raps)
library(MRMix)
library(MRPRESSO)
library(penalized)
library(conflicted)
source("helper/MR_lasso.R")  # Slob&Burgess 2020
source("helper/PVE_calculation_fns.R") 
source("helper/plotfigure01fn.R") 
conflict_prefer("select", "dplyr")
conflict_prefer("filter", "dplyr")
```

# Discussion

-   NA

# Version Update

The new report version includes: 

-   All genetic associations with BMI > 0 (for association statistics < 0, sign of the association statistic with GSD changed respectively)


```{r load-data}
my_data <- as_tibble(read.delim("../data-boekstegers_felix/GST_BMI_association_results.txt"))
# GBC: new data 2022-11-04
GBC_dat <- as_tibble(read.delim("../data-boekstegers_felix/C23C24_BMI_association_results.txt"))

# FTO data
FTO_dat <- as_tibble(read.delim("../data-boekstegers_felix/2023-01-19_FTO/FTO_association_results.txt"))
```

# Data Harmonization

We start with harmonizing our data sets for the exposure 

```{r data-harmonization, message=FALSE}
##exposure_dat:
#SNP
#beta.exposure
#se.exposure
#effect_allele.exposure
#other_allele.exposure
#eaf.exposure
exposure_dat <- my_data %>% 
  select(dummyID, SNP, beta_exp, se_exp, a1, a2, eaf_exp, exposure) %>% 
  rename(
    id.exposure = dummyID,
    beta.exposure = beta_exp,
    se.exposure = se_exp,
    effect_allele.exposure = a1,
    other_allele.exposure = a2, 
    eaf.exposure = eaf_exp, 
    exposure = exposure
  )
##outcome_dat:
#SNP
#beta.outcome
#se.outcome
#effect_allele.outcome
#other_allele.outcome
#eaf.outcome
#outcome
outcome_dat <- GBC_dat %>% 
  select(dummyID, SNP, beta_out, se_out, a1, a2, eaf_out, outcome) %>% 
  rename(
    id.outcome = dummyID, 
    beta.outcome = beta_out,
    se.outcome = se_out,
    effect_allele.outcome = a1,
    other_allele.outcome = a2, 
    eaf.outcome = eaf_out,
    outcome = outcome
  )

harmonised_dat <- TwoSampleMR::harmonise_data(
  exposure_dat = exposure_dat, 
  outcome_dat = outcome_dat, 
  action = 2  # default
  ) # # used later for analysis with the {TwoSampleMR} package

sum_SNPs_remove <- sum(harmonised_dat$remove)
sum_SNPs_palindromic <- sum(harmonised_dat$palindromic)
sum_SNPs_ambiguous <- sum(harmonised_dat$ambiguous)
sum_SNPs_remove <- nrow(harmonised_dat)-sum(harmonised_dat$mr_keep)
```


```{r combine-dat}
my_data_harm <- 
  harmonised_dat %>% 
  rename(
    exposureBMI = exposure, 
    outcomeGBC = outcome,
    # estimates exposure BMI
    eaf.bmi=eaf.exposure,
    beta.bmi=beta.exposure,
    se.bmi=se.exposure, 
    
    # estimates outcome GBC
    eaf.gbc=eaf.outcome,
    beta.gbc=beta.outcome,  
    se.gbc=se.outcome,
  ) %>% 
  left_join(my_data %>% select(SNP, p_value_exp, outcome, p_value_out, eaf_out, beta_out, se_out),by = c("SNP"="SNP")) %>% 
  rename(
    # BMI only need to add p-val
    p.value.bmi = p_value_exp,
    
    # GSD rename to "confounder"
    confounderGSD = outcome, 
    eaf.gsd = eaf_out,
    beta.gsd = beta_out, 
    se.gsd = se_out, 
    p.value.gsd = p_value_out,
  ) %>% 
  select(
    SNP, 
    # DROP:   "effect_allele.exposure" "other_allele.exposure"  "effect_allele.outcome"  "other_allele.outcome"
    
    # EXPOSURE
    eaf.bmi,
    beta.bmi,
    se.bmi,
    p.value.bmi,
    
    # CONFOUNDER
    eaf.gsd,
    beta.gsd,
    se.gsd,
    p.value.gsd,
    # OUTCOME
    eaf.gbc,
    beta.gbc,
    se.gbc,
    # pvalue missing and needs to be added from GBC_dat
    # harmonize decision
    palindromic, ambiguous, mr_keep
  ) %>% 
  left_join(GBC_dat %>% select(SNP, p_value_out), by=c("SNP"="SNP")) %>% 
  rename(
    p.value.gbc = p_value_out
  ) %>% 
  relocate(
    p.value.gbc, .before = palindromic
  ) %>% 
  mutate(
    p.value.gsd = ifelse(p.value.gsd == "<.0001",  # convert from string
                         yes = "0.0001", 
                         no=p.value.gsd), 
    p.value.gsd = as.numeric(p.value.gsd)) %>% 
  as_tibble()
```


After harmonizing our data sets we remove `r sum_SNPs_remove` SNPs since they are ambiguous. 

```{r rm-ambigous-SNPs}
dropped_data <- my_data_harm %>% 
  filter(mr_keep==FALSE)

my_data_harm <- my_data_harm %>% 
  filter(mr_keep==TRUE | SNP == "rs1558902") %>%  # also include FTO SNP 
  # change sign of beta after checking consistency of data (see mail felix 2023-01-19)
  mutate(beta.bmi = ifelse(SNP == "rs1558902", -beta.bmi, beta.bmi)) 

my_data_harm <- my_data_harm %>% 
  mutate(
    beta.gsd = ifelse(beta.bmi<0, -beta.gsd, beta.gsd),
    beta.gbc = ifelse(beta.bmi<0, -beta.gbc, beta.gbc),
    # change last beta.bmi, otherwise not negative any more ;)
    beta.bmi = ifelse(beta.bmi<0, -beta.bmi, beta.bmi)
  )
```


## Get gene names from rs-ids

```{r highlight-fto-df}
# see script biomart_search.R
rsIds_GeneNames <- as_tibble(readRDS("../data/biomartBMIsearch.rds"))

filter_FTO <- rsIds_GeneNames %>% 
  filter(str_detect(associated_gene, pattern = "FTO") | str_detect(associated_gene, pattern = "MC4R")) %>% 
  distinct(refsnp_id, .keep_all = TRUE)

t(filter_FTO) %>% kable() %>% 
  kable_styling(font_size=10)
```

After data harmonization we drop `r sum_SNPs_remove` IVs one of which is the SNP associated with the FTO gene. 

\clearpage

# Simulating on empirical data

**Calculate approximate SEs for decreasing sample sizes and corresponding p-vals**

-   se, pval and study size depend on each other, within each study
-   selection of IVs based on exposure study (BMI in our case)
-   Summary statistics on genetic associations with BMI based on results of the GIANT + GERA studies [@Hoffmann_2018] with $N=100,418 + 234,069 =334487$ and $SNPs=289$

**Steps**

1.  Calculate more accurate SEs using the betas and Pvals
2.  Calculate the approximate SEs for decreasing samples sizes
3.  Calculate the probability values considering the estimated betas and the corresponding SEs
4.  Exclude the IV if Pval $\ge$ genome-wide significance threshold (5 × 10−6)
5.  Calculate cumulative explained variances
6.  Represent figure similar to Fig1 in Chatterjee's paper @QiChatterjee_2021


```{r z-statistic}
## LOOP simulation ----------------------------------------------
n.orig <- 334487
n <- c(1000, 2000, 5000, seq(10000, 330000, by=10000), 334487)

# create vectors to store pvals_n in
colnames_pval <- c()
for (i in 1:length(n)) {
  colnames_pval[i] <- paste0("p_value_exp_sample_", format(n[i], scientific = FALSE))
}
## initialize
se_update <- c()
pvals_df <- data.frame(matrix(ncol = length(colnames_pval), nrow =  nrow(my_data_harm)))
colnames(pvals_df) <- colnames_pval
for (i in 1:nrow(my_data_harm)) {
  # 1) Calculate more accurates SEs using the provided betas and Pvals 
  se_update[i] <- my_data_harm[["beta.bmi"]][i]/-qnorm(p=(my_data_harm[["p.value.bmi"]][i]/2),mean=0)
  # 2) Calculate the approximate SEs for decreasing samples sizes
  ### approximation of sd (single value)
  sd <- se_update[i] * sqrt(n.orig)
  #### approximation of standard error simulating different sample size
  se_n <- sd/(sqrt(n))
  # 3) Calculate the probability values considering the estimated betas and the corresponding SEs
  for (j in 1:length(n)) {
    pvals_n <- 2*(1-pnorm(abs(my_data_harm[["beta.bmi"]][i])/abs(se_n[j]), mean=0, 
                          lower.tail = TRUE))
    pvals_df[i, j] <- pvals_n
  }
}

# 4. Exclude the IV if Pval > genome-wide significance threshold  (5 × 10−8) --> updated to 5 × 10−6
pvals_df %>% 
  summarize(across(.cols = everything(), function(x) sum(x < 5*10^(-6)))) %>% 
  t() %>% 
  cbind(n) %>% 
  base::as.data.frame() %>% 
  tibble::rownames_to_column() %>% 
  rename(sample_size = n, No_of_IVs=V1) %>% 
  select(sample_size, No_of_IVs) %>% 
  kable(caption = "Count of significant SNPs depending on sample size") %>% 
  kable_styling(font_size=10)

# pvals_df %>%
#   filter(p_value_exp_sample_100000< 5*10^(-6)) %>% 
#   count()
my_data_harm <- as_tibble(cbind(my_data_harm, pvals_df))
```


## Calculate proportion of variance in phenotype explained by a given SNP (PVE)

-   The variance in BMI explained by the IVs is calculated with the following formula: $explained\ variance = 2f(1-f)\beta^2$, where $f$ denotes the allele frequency and $\beta$ is the additive genetic effect.
-   The explained variance in liability to GSD and GBC is calculated as described by @So_2011
    -   Prevalence of disease $K$ must be set:
        -   [Prevalence GSD](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3343155/#:~:text=Gallstones%20are%20common.,develop%20stones%20at%20some%20time.&text=The%20majority%20will%20not%20develop,cholecystitis,%20cholangitis,%20or%20pancreatitis)
            - @Aerts_2003
        -   [Prevalence GBC](https://www.wcrf.org/cancer-trends/gallbladder-cancer-statistics/) 
            - Update prevalence $K=0.05$ from [Cancer Today](https://gco.iarc.fr/today/home)

We added a second formula to calculate the explained variance for numeric traits [@Teslovich_2010]. 

```{r calculate-explained-variance}
# outsourced to "./scripts/PVE_calculation_fns.R"

## CALCULATE explained variance for total sample size
my_data_harm <- my_data_harm %>%
  mutate(
    explained_variance_BMI = explained_variance_numeric(eaf = eaf.bmi, beta = beta.bmi),
    # USE THIS !
    explained_variance_BMI2 = explained_variance_numeric2(
      maf = eaf.bmi, beta = beta.bmi,
      se_beta = se.bmi,
      samplesize = n[length(n)]
    ),
    explained_variance_GSD = explained_variance_binary(
      PA = eaf.gsd,
      RR1 = exp(beta.gsd),
      RR2 = 1 + (2 * (exp(beta.gsd) - 1)),
      K = 0.1 # assumed prevalence in population
    )$Vg,
    explained_variance_GBC = explained_variance_binary(
      PA = eaf.gbc,
      RR1 = exp(beta.gbc),
      RR2 = 1 + (2 * (exp(beta.gbc) - 1)), # exp(beta.gbc)^2,
      K = 1.2 / 100000 # assumed prevalence in population
    )$Vg
  )

# sum(my_data_harm$explained_variance_GBC)
```


### %var(X) BMI formula comparison

```{r compare-formulas-numeric-pve}
ggplot(data=my_data_harm, aes(x=explained_variance_BMI, y=explained_variance_BMI2)) +
  geom_point(alpha=0.6)+
  geom_abline()+
  theme_few()
```


## Plot explained variance in BMI vs -log10 P-value from Hoffmann et al.

```{r PVE_BMI_vs_pval, warning=FALSE}
# labeling (only FTO label)
my_data_harm$gene_label <- NA
my_data_harm$gene_label[my_data_harm$SNP %in% filter_FTO$refsnp_id] <- "FTO/MC4R"
my_data_harm$gene_label[my_data_harm$SNP == "rs1558902"] <- "FTO"
my_data_harm$gene_label[my_data_harm$SNP == "rs6567160"] <- "MC4R"

# color labeling
my_data_harm$fto_col_annot <- "No FTO/MC4R variant"
my_data_harm$fto_col_annot[my_data_harm$SNP %in% filter_FTO$refsnp_id] <- "FTO/MC4R variant"

ggplot(data = my_data_harm, 
       mapping = aes(x=explained_variance_BMI*100, y= -log10(p.value.bmi),  
                    col=fto_col_annot, label=gene_label)
       )+
  geom_point(alpha=0.6)+
  ggrepel::geom_text_repel(show.legend = FALSE)+
  xlab("Explained variance in BMI in %")+
  ylab("-log10 P-val")+
  labs(title = "Explained variance exposure",
       col='SNP'
       ) +
  theme_few()+
  scale_colour_few() -> fig00
# fig00   # we use the sample size dependent formula

# second formula variance explained2
ggplot(data = my_data_harm, 
       mapping = aes(x=explained_variance_BMI2*100, y= -log10(p.value.bmi),  
                    col=fto_col_annot, label=gene_label)
       )+
  geom_point(alpha=0.6)+
  ggrepel::geom_text_repel(show.legend = FALSE)+
  xlab("Explained variance in BMI in %")+
  ylab("-log10 P-val")+
  labs(title = "Explained variance exposure",
       col='SNP'
       ) +
  theme_few()+
  scale_colour_few() -> fig00_b
fig00_b

# filenamefig00_a <- paste0("../output/plots/", Sys.Date(), "_explained_var_X_old_formula.png")
# ggsave(filename = filenamefig00_a, plot = fig00, 
#        width = 10, height = 6, 
#        units = "in"  # default
#        )

filenamefig00_b <- paste0("../output/plots/", Sys.Date(), "_explained_var_X_new_sample_size_dep.png")
ggsave(filename = filenamefig00_b, plot = fig00_b, 
       width = 10, height = 6, 
       units = "in"  # default
       )
```

## Plot Figure 1 from Qi&Chatterjee 2021

```{r get-data-for-fig1, warning=FALSE}
# 4.       Exclude the IV if Pval > genome-wide significance threshold  (5 × 10−8) **NEW 5x10^(-6))**
# 5.       Calculate cumulative explained variances
# initialize df
df_cum_expl_var <- tibble(n=n, 
                          cum_expl_varX = rep(NA, length(n)), 
                          cum_expl_varX2 = rep(NA, length(n)), 
                          cum_expl_varU = rep(NA, length(n)), 
                          cum_expl_varY = rep(NA, length(n)),
                          count_IVs = rep(NA, length(n))
)

threshold <- 5*10^(-6)
for (i in 1:length(colnames_pval) ) {
  select_col <- colnames_pval[i]
  # calculate cum explained variance and count of IVs
  summarydat <- my_data_harm %>% 
    filter(!!sym(select_col) < threshold) %>% 
    select(explained_variance_BMI, explained_variance_BMI2, explained_variance_GSD, explained_variance_GBC) %>% 
    # cumulative PVE
    summarise(expl_variances =across(.cols = everything(), function(x) sum(x, na.rm = TRUE)),
              count_IVs = n())
  # fill up df
  df_cum_expl_var[i, 2] <- summarydat$expl_variances[1]
  df_cum_expl_var[i, 3] <- summarydat$expl_variances[2]
  df_cum_expl_var[i, 4] <- summarydat$expl_variances[3]
  df_cum_expl_var[i, 5] <- summarydat$expl_variances[4]
  df_cum_expl_var[i, 6] <- summarydat$count_IVs
}

# for the numeric traits we have to map over sample size and recalculate se_updated
explained_variance_BMI2_sample_size <- map(n, ~explained_variance_numeric2(maf = my_data_harm$eaf.bmi, beta = my_data_harm$beta.bmi, 
                                                                           se_beta = abs(my_data_harm$beta.bmi / -qnorm(p=(my_data_harm$p.value.bmi / 2), mean = 0) * sqrt(n.orig) / sqrt(.x)), 
                                                                           samplesize = .x))

cum_expl_varX2_sample_size <- c()
for (mysamples in 1:length(colnames_pval)) {
  temp_dat <- my_data_harm %>% 
    select(all_of(colnames_pval[mysamples])) %>% 
    filter(!!sym(colnames_pval[mysamples]) < threshold) %>% 
    mutate(explained_variance_BMI2_sample_size = explained_variance_BMI2_sample_size[[mysamples]])
  #print(temp_dat)
  cum_expl_varX2_sample_size <- append(cum_expl_varX2_sample_size, sum(temp_dat$explained_variance_BMI2_sample_size))
}

df_cum_expl_var <- df_cum_expl_var %>% 
  cbind(cum_expl_varX2_sample_size) %>% 
  relocate(cum_expl_varX2_sample_size, .before = cum_expl_varU)

ggplot(df_cum_expl_var, aes(x=n))+
  geom_line(aes(y=cum_expl_varX2_sample_size*100), col="red")+
  geom_line(aes(y=cum_expl_varX2*100))
```


```{r updated-fig01, fig.height=6, fig.width=10, warning=FALSE}
# 6.       Represent figure similar to Fig1 in Chatterjee’s paper

# Dual Y axis: we have 284 IVs, scale them to 5% (y-axis [0;5%]): 3/284 
## https://r-graph-gallery.com/line-chart-dual-Y-axis-ggplot2.html
scale_factor <- 4.5/284

df_cum_expl_var <- df_cum_expl_var %>% 
  mutate(
    cum_expl_varX = cum_expl_varX *100, # in percent 
    cum_expl_varX2 = cum_expl_varX2_sample_size *100,  #  CAVE CHANGED HERE TO INCLUDE SAMPLE SIZE DEPENDENT SE and PVE CALC!
    cum_expl_varU = cum_expl_varU *100,
    cum_expl_varY = cum_expl_varY *100,
    count_IVs_scaled = count_IVs*scale_factor)

plot_data1b <- df_cum_expl_var %>% 
  pivot_longer(cols = c(cum_expl_varX2, cum_expl_varU, cum_expl_varY)) %>% 
  mutate(name = factor(name, labels = c("% var(U)", "% var(X)", "% var(Y)")),
         name = factor(name, levels = c("% var(X)", "% var(Y)", "% var(U)") )
         )
## plot
ggplot(plot_data1b, aes(x=n, y=value, col=name))+
  geom_point(alpha=0.6)+
  geom_line(alpha=0.6)+
  geom_line(aes(x=n, y=count_IVs_scaled, col="Number of IVs"), linetype="dotdash")+
  scale_x_continuous(name = paste0("Size of sample ", as.roman(1)),  # 2-sample-MR or label "Study Size"
                     labels = comma, 
                     breaks = N_annotation, #c( 10000, n[7:length(n)]),
                     limits = c(10000, 300000))+
  #ylab("% variance explained by IVs")+
  labs(title = "Figure 1",
       col='', 
       caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  scale_y_continuous(
    # Features of the first axis
    name = "% variance explained by IVs",
    # Add a second axis and specify its features
    sec.axis = sec_axis(~./scale_factor, name="Number of IVs"),
    limits = c(0,5)
  ) + 
  theme_few()+
  #scale_colour_few() +
  #scale_color_manual(name = "", values = c("% var(U)" = "red", 
  #                                         "% var(X)" = "blue", 
  #                                         "% var(Y)" = "darkblue", 
  #                                         "Number of IVs" = "black"
  #                                         ))
  #
  scale_color_manual(name = "", values = c("% var(U)" = ggthemes_data$few$colors$Dark[2,2][[1]], 
                                           "% var(X)" = ggthemes_data$few$colors$Dark[3,2][[1]], 
                                           "% var(Y)" = ggthemes_data$few$colors$Dark[4,2][[1]], 
                                           "Number of IVs" = "black"),
                     # order
                     breaks = c( "Number of IVs", "% var(X)", "% var(Y)", "% var(U)")
  ) -> fig01b.update
fig01b.update  # USE THIS

filenamefig01_b <- paste0("../output/plots/", Sys.Date(), "_figure01_new_PVE_X_formula.png")
ggsave(filename = filenamefig01_b, plot = fig01b.update, 
       width = 10, height = 6, 
       units = "in"  # default
       )
```

Discussion: *Pleiotropy* with increasing sample size. INSIDE assumption. 


# Different genetic architectures of the exposure

## MAF vs P-val

For the ~200 IVs, please plot the minor allele frequency (x-axis) versus -log10Pval (y-axis). Calculate the median MAF and the median –log10Pval, and draw the corresponding vertical and horizontal lines. The four areas will define four genetic architectures for the simulations (e.g. relatively rare variants with relatively large exposure effects).

```{r MAF-Pval, warning=FALSE}
ggplot(data = my_data_harm, aes(x=eaf.bmi, y=-log10(p.value.bmi)))+
  geom_point(alpha=0.6)+
  geom_hline(yintercept=-log10(median(my_data$p_value_exp)), col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
  geom_vline(xintercept=median(my_data$eaf_exp), col=ggthemes_data$few$colors$Dark[3,2][[1]], alpha=0.6)+
  xlab("Effect allele frequency for exposure")+
  ylab("-log10(P-value) ")+
  labs(title = "MAF vs Pval",
       subtitle = "Define different genetic architectures",
       #col='', 
       #caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  theme_few()+
  scale_colour_few() ->fig02
fig02

# ggsave(filename = "../output/plots/figure02.png", plot = fig02)
```

## MAF vs betas

```{r mafVSbetas}
plot_data_2 <- my_data_harm %>% 
  mutate(maf.bmi = ifelse(eaf.bmi >0.5, 1-eaf.bmi, eaf.bmi)) %>% 
  select(maf.bmi, beta.bmi, gene_label, fto_col_annot)


ggplot(data = plot_data_2, aes(x=maf.bmi, y=abs(beta.bmi), 
                               col=fto_col_annot, label=gene_label))+
  geom_point(alpha=0.6)+
  ggrepel::geom_text_repel(show.legend = FALSE)+
  geom_hline(yintercept=median(abs(plot_data_2$beta.bmi)), 
             col=ggthemes_data$few$colors$Dark[6,2][[1]], alpha=0.6, linetype=2)+
  geom_vline(xintercept=median(plot_data_2$maf.bmi), 
             col=ggthemes_data$few$colors$Dark[6,2][[1]], alpha=0.6, linetype=2)+
  annotate(geom = "text", x = 0.15, y = 0.045, 
           label="relatively rare\nrelatively strong", 
           color= ggthemes_data$few$colors$Dark[2,2][[1]]
             )+
  annotate(geom = "text", x = 0.42, y = 0.05, 
           label="relatively common\nrelatively strong", 
           color=ggthemes_data$few$colors$Dark[3,2][[1]])+
  annotate(geom = "text", x = 0.1, y = 0.015, 
           label="relatively rare\nrelatively weak", color=ggthemes_data$few$colors$Dark[4,2][[1]])+
  annotate("segment", x=0.44,xend=0.42,y=0.025,yend=0.013,arrow=arrow(), 
           color=ggthemes_data$few$colors$Dark[5,2][[1]])+
  annotate(geom = "text", x = 0.44, y = 0.03, 
           label="relatively common\nrelatively weak", 
           color=ggthemes_data$few$colors$Dark[5,2][[1]])+
  #geom_text(x=0.1, y=0.05, label="hello", family = "Times New Roman",
  #          size=6, col="red")+
  xlab("Minor allele frequency")+
  ylab("Exposure effect size")+  # abs(beta_BMI)
  labs(title = "",
       subtitle = "Define different genetic architectures",
       col='SNP', 
       #caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  theme_few()+
  scale_colour_few() -> fig02.update
fig02.update

# SAVE
filenamefig02 <- paste0("../output/plots/", Sys.Date(), "_figure02.png")
ggsave(filename = filenamefig02, plot = fig02.update)
```


We want to repeat Figure 01 for each of the 4 defined quadrants in the MAF vs beta plot above. 

```{r figure1quadrants, fig.show="hold", out.width="50%", warning=FALSE}
## plotfigure sourced from helper functions ----------------------------------
# create all combis
effect_size <- c("strong", "weak")
maf <- c("rare", "common")
all_quadrants <- tidyr::crossing(effect_size, maf)

####### PLOT NOW  ---
for (i in 1:nrow(all_quadrants)) {
  plotfigure01(data = my_data_harm, 
               effect_size = all_quadrants$effect_size[i], 
               maf = all_quadrants$maf[i], 
               n=n
               )
}

```

## FINAL PLOT

```{r FINAL-PLOT-1-huray, fig.height=10, fig.width=14, message=FALSE, warning=FALSE}
fig01b.update + 
  labs(title= "", caption = "") -> fig01b.update.arrange

fig02.update+
  labs(title= "", subtitle="", caption = "") -> fig02.update.arrange

figure01_EFFECT_SIZE_strong_VARIANTS_common+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[3,2][[1]])) -> figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange

figure01_EFFECT_SIZE_strong_VARIANTS_rare+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[2,2][[1]])) -> 
  figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_common+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[5,2][[1]])) -> figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_rare+
  labs(title= "", caption = "") +
  theme(plot.subtitle = element_text(color = ggthemes_data$few$colors$Dark[4,2][[1]])) -> figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange

ggpubr::ggarrange(fig02.update.arrange, fig01b.update.arrange,
                  figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange, figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange,
                  figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange, figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange, 
                  ncol=2, nrow=3, 
                  heights = c(2, 1.5, 1.5),
                  legend = "bottom", common.legend = TRUE, align = "hv"
                  ) -> final_plot1.arrange


figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange+
  # delete y-labels
  scale_y_continuous(name="% variance explained by IVs", sec.axis = sec_axis(~./scale_factor, name=""), limits = c(0,2)) -> figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange

figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange+
  # delete y-labels
  scale_y_continuous(name="", sec.axis = sec_axis(~./scale_factor, name=""), limits = c(0,2)) -> figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange+
  # delete y-labels
  scale_y_continuous(name="",sec.axis = sec_axis(~./scale_factor, name=""), limits = c(0,2)) -> figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange

figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange+
  # delete y-labels
  scale_y_continuous(name="", sec.axis = sec_axis(~./scale_factor, name="Number of IVs"), limits = c(0,2)) -> figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange

ggpubr::ggarrange(ggarrange(fig02.update.arrange, fig01b.update.arrange, ncol=2, labels = c("A", "B"), legend = "none", align = "hv"),
                  ggarrange(figure01_EFFECT_SIZE_strong_VARIANTS_rare.arrange,
                            figure01_EFFECT_SIZE_strong_VARIANTS_common.arrange, 
                            figure01_EFFECT_SIZE_weak_VARIANTS_rare.arrange,
                            figure01_EFFECT_SIZE_weak_VARIANTS_common.arrange, 
                            ncol=4, labels = c("C", "D", "E", "F"), 
                            legend="bottom", common.legend = TRUE),
                  nrow=2
                  #heights = c(2, 1.5, 1.5),
                  #legend = "bottom", common.legend = TRUE, align = "hv"
                  ) -> final_plot1.arrange2

# final_plot1.arrange   # OLD
final_plot1.arrange2   # use this!

filename_final_plot1 <- paste0("../output/plots/", Sys.Date(), "_final_plot1.arrange_OLD.png")
filename_final_plot1b <- paste0("../output/plots/", Sys.Date(), "_final_plot1b.arrange_NEW.png")

ggsave(filename = filename_final_plot1, plot = final_plot1.arrange, 
         width = 14, height = 10, 
         units = "in"  # default
         )
ggsave(filename = filename_final_plot1b, plot = final_plot1.arrange2, 
         width = 14, height = 10, 
         units = "in"  # default
         )
```


\clearpage

# A *Gallstone Disease* Simulation Framework

## Get MR-estimates first

### Using {MendelianRandomization} package

```{r MendelianRandomization-analysis-GSD, warning=FALSE}
# outsourced calculation to `perform_mr.R`
# data FOR estimates in `perform_mr.R` script and simulation in `simstudy.R` script
saveRDS(my_data_harm, paste0("../output/Rdata/", Sys.Date(), "_my_data_harm.rds"))

## load results from script `perform_mr.R`
est.bmi.gsd <- readRDS(file = dplyr::last(list.files("../output/Rdata/", pattern = "_est_BMI_GSD.rds", full.names = TRUE))) 

# biology plausibility FTO & MC4R
# ## The mr_ivw function implements the inverse-variance method, informally 
# known as the "Toby Johnson" method. With a single genetic variant, this 
# is simply the ratio method.
fto_wald <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$beta.gsd[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.gsd[my_data_harm$SNP == "rs1558902"]
)

fto_wald_ukbiobank <- mr_wald_ratio(
  b_exp = exp(FTO_dat$beta[3])-1, # BMI UK Biobank log-transformed  (before Hoffmann GERA + GIANT)
  b_out = FTO_dat$beta[2], # GSD UKB as before
  se_exp = exp(FTO_dat$se[3])-1, 
  se_out = FTO_dat$se[2]
  )

fto_wald_ukbiobank_noexp <- mr_wald_ratio(
  b_exp = (FTO_dat$beta[3]), # BMI UK Biobank log-transformed  (before Hoffmann GERA + GIANT)
  b_out = FTO_dat$beta[2], # GSD UKB as before
  se_exp = (FTO_dat$se[3]), 
  se_out = FTO_dat$se[2]
  )

mc4r_wald <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$beta.gsd[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.gsd[my_data_harm$SNP == "rs6567160"]
)
# add rows
est.bmi.gsd <- est.bmi.gsd %>% 
  add_row(mr_methods = "FTO_Wald", estimate = fto_wald$b, se.estimate=fto_wald$se) %>% 
  add_row(mr_methods = "MC4R_Wald", estimate = mc4r_wald$b, se.estimate=mc4r_wald$se) %>% 
  arrange(desc(estimate))

est.bmi.gsd %>% 
  mutate(OR=exp(estimate), CI.lower=exp(estimate-se.estimate),CI.upper=exp(estimate+se.estimate)) %>%  
  kable(digits=3, caption = "MR Results for GSD (outcome) using genetic variants as instrumental variables for BMI") %>% 
  kable_classic(full_width = FALSE, html_font = "Cambria")
```


### Scatter Plot

```{r scatter-MendelianRandomization-GSD, fig.width=10, fig.height=6, warning=FALSE, message=FALSE}
# create mr.object
mr.obj = MendelianRandomization::mr_input(
    # exposure
    bx = my_data_harm$beta.bmi, 
    bxse = my_data_harm$se.bmi, 
    # outcome
    by = my_data_harm$beta.gsd, 
    byse = my_data_harm$se.gsd, 
    snps = my_data_harm$SNP,
    exposure = "Body mass index",
    outcome = "Gallstone disease"
  )

p1.gsd <- mr_plot(mr.obj, error = TRUE, 
                  line = "ivw", 
                  orientate = FALSE, 
                  interactive = FALSE, 
                  labels = FALSE)

p1.gsd$layers[[1]]$aes_params$size <- 1
p1.gsd$layers[[1]]$aes_params$alpha <- 0.5

p1.gsd$layers[[4]]$aes_params$colour <- ggthemes_data$few$colors$Dark[1,2][[1]]
p1.gsd$layers[[5]]$aes_params$colour <- ggthemes_data$few$colors$Dark[1,2][[1]]
p1.gsd$layers[[4]]$aes_params$alpha <- 0.3
p1.gsd$layers[[5]]$aes_params$alpha <- 0.3

p1.gsd$layers[[6]]$aes_params$colour <- ggthemes_data$few$colors$Dark[2,2][[1]]

p1.gsd+
  ggthemes::theme_few() -> p1.gsd

p1.gsd

# SAVE
filename.mr.scatter.gsd <- paste0("../output/plots/", Sys.Date(), "_figure_mrscatter_gsd.png")
ggsave(filename = filename.mr.scatter.gsd, plot = p1.gsd, 
         width = 10, height = 6, 
         units = "in"  # default
         )
```


### Forest Plot

```{r forestplot-MendelianRandomization-GSD, fig.width=5, fig.height=27, warning=FALSE}
p2.gsd <- mr_forest(
  mr.obj,
  alpha = 0.05,
  snp_estimates = TRUE,
  methods = "ivw",
  ordered = TRUE
) +
  ggthemes::theme_few()

#SAVE
filename.mr.forest.gsd <- paste0("../output/plots/", Sys.Date(), "_figure_mrforest_GSD.png")
ggsave(filename = filename.mr.forest.gsd, plot = p2.gsd, 
       width = 5, 
       height = 24, 
         #units = "in"  # default
         )
```

### Funnel Plot

```{r funnel-plot-GSD, fig.height=3, fig.width=5}
p4.gsd <- mr_funnel(mr.obj, CI=TRUE)

p4.gsd$layers[[1]]$aes_params$size <- 1
p4.gsd$layers[[1]]$aes_params$alpha <- 0.8
# linerange layer
p4.gsd$layers[[2]]$aes_params$alpha <- 0.2
# xintercept layer - estimate
p4.gsd$layers[[4]]$aes_params$colour <- ggthemes_data$few$colors$Dark[9,2][[1]]
# delete linerange layer
p4.gsd$layers <- p4.gsd$layers[-2]  

# plot
p4.gsd <- p4.gsd+
  ggthemes::theme_few()
p4.gsd

#SAVE
filename.mr.funnel.gsd <- paste0("../output/plots/", Sys.Date(), "_figure_mrfunnel_gsd.png")
ggsave(filename = filename.mr.funnel.gsd, plot = p4.gsd, 
       width = 8, 
       height = 4, 
       units = "in"  # default
         )

```


\clearpage


# B *Gallbladder Cancer* Simulation Framework

## Get MR-estimates first

### Using {TwoSampleMR} package

```{r TwoSampleMR-analysis}
harmonised_dat4TwoSampleMR <- harmonised_dat %>% 
  mutate(id.exposure = "BMI", id.outcome = "GBC")  # https://mrcieu.github.io/TwoSampleMR/articles/perform_mr.html

res <- TwoSampleMR::mr(harmonised_dat4TwoSampleMR)
res <- res %>% mutate(OR = exp(b), CI.lower= exp(b-se), CI.upper=exp(b+se)) %>% 
  arrange(desc(b))

res %>% 
  kable(digits = 3, 
        caption = "MR Results for GBC (outcome) using genetic variants as instrumental variables for BMI"
        ) %>% 
  kable_classic(full_width = FALSE, html_font = "Cambria")
```

### Scatter Plot

```{r scatter-TwoSampleMR, warning=FALSE, message=FALSE}
# mr_method_list()
res.selected_methods <- TwoSampleMR::mr(harmonised_dat4TwoSampleMR, 
                                        method_list = c("mr_egger_regression", "mr_ivw"))
p1 <- mr_scatter_plot(res.selected_methods, harmonised_dat4TwoSampleMR)
p1[[1]] #+
  #ggthemes::theme_few()

# SAVE
filename.mr.scatter <- paste0("../output/plots/", Sys.Date(), "_figure_mrscatter.png")
ggsave(filename = filename.mr.scatter, plot = p1[[1]], 
         #width = 10, height = 6, 
         #units = "in"  # default
         )
```

### Forest Plot

```{r forestplot-TwoSampleMR-GBC, fig.width=5, fig.height=24, warning=FALSE}
res_single <- TwoSampleMR::mr_singlesnp(harmonised_dat4TwoSampleMR, all_method = c("mr_egger_regression", "mr_ivw"))
p2 <- mr_forest_plot(res_single,  exponentiate = FALSE)
p2[[1]]

#SAVE
filename.mr.forest <- paste0("../output/plots/", Sys.Date(), "_figure_mrforest_GBC.png")
ggsave(filename = filename.mr.forest, plot = p2[[1]], 
       width = 5, 
       height = 24, 
         #units = "in"  # default
         )
```

### Funnel Plot

```{r funnel-plot, fig.height=3, fig.width=5}
p4 <- mr_funnel_plot(res_single)
p4[[1]] +
  ggthemes::theme_few() -> p4

p4

#SAVE
filename.mr.funnel.gbc <- paste0("../output/plots/", Sys.Date(), "_figure_mrfunnel_gbc.png")
ggsave(filename = filename.mr.funnel.gbc, plot = p4, 
       width = 8, 
       height = 4, 
       units = "in"  # default
         )
```


### Using {MendelianRandomization} package

```{r MendelianRandomization-analysis, warning=FALSE}
## load results
est <- readRDS(file = dplyr::last(list.files("../output/Rdata/", pattern = "_est.rds", full.names = TRUE)))

# biology plausibility FTO & MC4R
# ## The mr_ivw function implements the inverse-variance method, informally 
# known as the "Toby Johnson" method. With a single genetic variant, this 
# is simply the ratio method.
fto_wald.gbc <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$beta.gbc[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs1558902"], 
  my_data_harm$se.gbc[my_data_harm$SNP == "rs1558902"]
)
mc4r_wald.gbc <- mr_wald_ratio(
  my_data_harm$beta.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$beta.gbc[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.bmi[my_data_harm$SNP == "rs6567160"], 
  my_data_harm$se.gbc[my_data_harm$SNP == "rs6567160"]
)
# add rows
est <- est %>% 
  add_row(mr_methods = "FTO_Wald", estimate = fto_wald.gbc$b, se.estimate=fto_wald.gbc$se) %>% 
  add_row(mr_methods = "MC4R_Wald", estimate = mc4r_wald.gbc$b, se.estimate=mc4r_wald.gbc$se)


est %>% 
  mutate(OR=exp(estimate), CI.lower=exp(estimate-se.estimate),CI.upper=exp(estimate+se.estimate)) %>%  
  arrange(desc(estimate)) %>% 
  kable(digits=3, caption = "MR Results for GBC (outcome) using genetic variants as instrumental variables for BMI") %>% 
  kable_classic(full_width = FALSE, html_font = "Cambria")
```



# Simulation

We simulate betas from a normal distribution. The script is outsourced to `simstudy.R`. 

```{r simulatesimulatesimulate, warning=FALSE}
# save simulation-vars for simulation `simstudy.R` script
sim_vars <- list(n=n, colnames_pval=colnames_pval)
saveRDS(sim_vars, paste0("../output/Rdata/", Sys.Date(), "_sim_vars.rds"))
## SIMULATION OUTSOURCED TO "./scripts/simstudy.R"

# read dat from "simstudy.R"  #------------------------------------------------
est_sim <- readRDS(dplyr::last(list.files(path = "../output/Rdata", pattern = "_est_sim_NSIM_", full.names = TRUE)))
nsim <- nrow(est_sim)
```


## PVE with simulated data Plot

```{r simulated-pves, fig.height=6, fig.width=10, warning=FALSE}
# retrieve PVE colnames
colname_pve_X <-  colnames(est_sim)[str_detect(colnames(est_sim), pattern = "pve_X_sim")]
# PVE_Y, PVE_U removed on 2023-01-24 (see github)

## for performance measures later
# retrieve IVW/ MODE/ CONMIX cols for BMI -> GSD 
colname_IVW_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "IVW_theta_sim_BMI_GSD_")]
colname_IVW_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "IVW_theta_se_sim_BMI_GSD_")]
colname_MODE_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MODE_theta_sim_BMI_GSD_")]
colname_MODE_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MODE_theta_se_sim_BMI_GSD_")]
colname_CONMIX_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "CONMIX_theta_sim_BMI_GSD_")]
colname_CONMIX_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "CONMIX_theta_se_sim_BMI_GSD_")]
colname_MRMIX_theta_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MRMIX_theta_sim_BMI_GSD_")]
colname_MRMIX_theta_se_sim_BMI_GSD <- colnames(est_sim)[str_detect(colnames(est_sim), pattern = "MRMIX_theta_se_sim_BMI_GSD_")]
# retrieve IVW/ MODE/ CONMIX cols for BMI -> GBC
# removed on 2023-01-24 (see github)

# filter out sample sizes (IVW only used as example, could be any other col)
sample_size_filtered <- as.numeric(str_extract(colname_IVW_theta_sim_BMI_GSD, pattern = "(\\d)+"))

# EXPOSURE PVE
est_sim %>% 
  select(!!!(colname_pve_X)) %>% 
  pivot_longer(everything(), names_to = "PVE_X", values_to = "value_X") %>% 
  cbind(sample_size_filtered) %>% as_tibble() -> pve_X_sim
# # CONFOUNDER PVE
# est_sim %>% 
#   select(!!!(colname_pve_U)) %>% 
#   pivot_longer(everything(), names_to = "PVE_U", values_to = "value_U") %>% 
#   cbind(sample_size_filtered) %>% as_tibble() -> pve_U_sim
# # OUTCOME PVE
# est_sim %>% 
#   select(!!!(colname_pve_Y)) %>% 
#   pivot_longer(everything(), names_to = "PVE_Y", values_to = "value_Y") %>% 
#   cbind(sample_size_filtered) %>% as_tibble() -> pve_Y_sim

boxplot_overlay_dat <- pve_X_sim %>% 
  # cbind(pve_U_sim %>% select(-sample_size_filtered)) %>% 
  # cbind(pve_Y_sim %>% select(-sample_size_filtered)) %>% 
  as_tibble()

## plot
ggplot(plot_data1b, aes(x=n, y=value, col=name))+    # plot_data1b from section Simlating on empirical data (PVE)
  geom_point(alpha=0.6)+
  geom_line(alpha=0.6)+
  geom_line(aes(x=n, y=count_IVs_scaled, col="Number of IVs"), linetype="dotdash")+
  scale_x_continuous(name = paste0("Size of sample ", as.roman(1)),  # 2-sample-MR or label "Study Size"
                     labels = comma, 
                     breaks = N_annotation, #c( 10000, n[7:length(n)]),
                     limits = c(10000, 300000))+
  #ylab("% variance explained by IVs")+
  labs(title = "PVE with simulation added",
       col='', 
       caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
       ) +
  scale_y_continuous(
    # Features of the first axis
    name = "% variance explained by IVs",
    # Add a second axis and specify its features
    sec.axis = sec_axis(~./scale_factor, name="Number of IVs"),
    limits = c(0,7.5)
  ) + 
  #geom_boxplot(data=boxplot_overlay_dat, aes(x=sample_size_filtered, y=value_X*100, group=sample_size_filtered))
  geom_boxplot(data = boxplot_overlay_dat, aes(x=sample_size_filtered, y=value_X*100, # *100 to get percent values
                                group=sample_size_filtered), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2)+
  # U/Y removed on 2023-01-24 (see github)
  theme_few()+
  #scale_colour_few() +
  #scale_color_manual(name = "", values = c("% var(U)" = "red", 
  #                                         "% var(X)" = "blue", 
  #                                         "% var(Y)" = "darkblue", 
  #                                         "Number of IVs" = "black"
  #                                         ))
  #
  scale_color_manual(name = "", values = c("% var(U)" = ggthemes_data$few$colors$Dark[2,2][[1]], 
                                           "% var(X)" = ggthemes_data$few$colors$Dark[3,2][[1]], 
                                           "% var(Y)" = ggthemes_data$few$colors$Dark[4,2][[1]], 
                                           "Number of IVs" = "black"),
                     # order
                     breaks = c( "Number of IVs", "% var(X)", "% var(Y)", "% var(U)")
  ) -> plot_pve_simulated
#plot_pve_simulated

filename_pveplot_sim <- paste0("../output/plots/", Sys.Date(), "_figure_simplot_pve.png")
ggsave(filename = filename_pveplot_sim, plot = plot_pve_simulated, 
       width = 10, height = 6, 
       units = "in"  # default
       )
```

```{r simulated-pves-raincloud-plot, fig.height=6, fig.width=10, warning=FALSE}
ggplot(plot_data1b, aes(x=n, y=value, col=name))+    # plot_data1b from section Simlating on empirical data (PVE)
  geom_point(alpha=0.6)+
  geom_line(alpha=0.6)+
  geom_line(aes(x=n, y=count_IVs_scaled, col="Number of IVs"), linetype="dotdash")+
  scale_x_continuous(name = paste0("Size of sample ", as.roman(1)),  # 2-sample-MR or label "Study Size"
                     labels = comma, 
                     breaks = N_annotation, #c( 10000, n[7:length(n)]),
                     #limits = c(10000, 300000)  # throws an error when used with ggdist -> use coord_cartesian() instead!
                     )+
  #ylab("% variance explained by IVs")+
  labs(title = "PVE with simulation added",
       col='', 
       caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
  ) +
  scale_y_continuous(
    # Features of the first axis
    name = "% variance explained by IVs",
    # Add a second axis and specify its features
    sec.axis = sec_axis(~./scale_factor, name="Number of IVs"),
    limits = c(0,7.5)
  ) + 
  ## Rainclouds
  # https://z3tt.github.io/Rainclouds/
  # exposure simulation
  geom_boxplot(data = boxplot_overlay_dat, aes(x=sample_size_filtered, y=value_X*100, # *100 to get percent values
                                               group=sample_size_filtered), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = boxplot_overlay_dat, mapping=aes(x=sample_size_filtered, y=value_X*100, group=sample_size_filtered), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = boxplot_overlay_dat, mapping=aes(x=sample_size_filtered-2000, y=value_X*100, group=sample_size_filtered), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  coord_cartesian(xlim=c(10000, 300000))+
  theme_few()+
  #scale_colour_few() +
  #scale_color_manual(name = "", values = c("% var(U)" = "red", 
  #                                         "% var(X)" = "blue", 
  #                                         "% var(Y)" = "darkblue", 
  #                                         "Number of IVs" = "black"
  #                                         ))
  #
  scale_color_manual(name = "", values = c("% var(U)" = ggthemes_data$few$colors$Dark[2,2][[1]], 
                                           "% var(X)" = ggthemes_data$few$colors$Dark[3,2][[1]], 
                                           "% var(Y)" = ggthemes_data$few$colors$Dark[4,2][[1]], 
                                           "Number of IVs" = "black"),
                     # order
                     breaks = c( "Number of IVs", "% var(X)", "% var(Y)", "% var(U)")
  ) -> plot_pve_simulated.raincloud
plot_pve_simulated.raincloud

filename_pveplot_sim.raincloud <- paste0("../output/plots/", Sys.Date(), "_figure_simplot_pve_raincloud.png")
ggsave(filename = filename_pveplot_sim.raincloud, plot = plot_pve_simulated.raincloud, 
       width = 10, height = 6, 
       units = "in"  # default
       )
```

## Performance Metrics

### Calculate BIAS, MSE, MEAN, EMP_SE

```{r calculate-performance-metrics}
# fns
bias <- function(thetaobs,theta, ...) mean(thetaobs)- theta
mse <- function(thetaobs, theta, ...) mean((thetaobs-theta)^2)
emp_se <- function(nsim, thetaobs, theta, ...) sqrt( 1/(nsim-1) * sum((thetaobs-mean(thetaobs))^2) )  # == sd (with ...)


## get performance metrics    # https://stackoverflow.com/questions/47415072/pass-multiple-functions-to-purrrmap
funs <- c(mean = mean, emp_se=emp_se, bias = bias, mse = mse)
## args gsd
args.gsd.ivw <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "IVW"], 
                     nsim=nrow(est_sim)
                     ) # trim argument non-matching and ignored for median
args.gsd.mode <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "mode"], 
                     nsim=nrow(est_sim)
                     )
args.gsd.conmix <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "conmix"], 
                     nsim=nrow(est_sim)
                     )
args.gsd.mrmix <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "MRMix"], 
                     nsim=nrow(est_sim)
                     )
## args gbc
#deleted 2023-01-29

## Standard Deviation
# est_sim %>%    # equivalent to `emp_se()` function, but `sd()` misses dotdotdot argument 
#   select(!!!(colname_theta)) %>% 
#   map_dbl(.f = ~sd(.))
```


```{r CALC-perf-metrics-GSD}
IVW_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.ivw), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)

MODE_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.mode), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)

CONMIX_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.conmix), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)

MRMIX_GSD_performance_metrics <- est_sim %>% 
  select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
  map_df(.f = ~ funs %>% 
           map(exec, .x, !!!args.gsd.conmix), .id = "var") %>% 
  cbind(sample_size_filtered) %>% 
  rename(N = sample_size_filtered)
```

```{r CALC-perf-metrics-GBC}
# deleted 2023-01-29
```

### Calculate COVERAGE, COVERAGE-BIAS-ELIMINATED, TYPE1ERROR

```{r further-metrics}
# coverage
coverage <- function(thetaobs, thetaobs_se, theta, ...) {
  lower_ci <- thetaobs - thetaobs_se
  upper_ci <- thetaobs + thetaobs_se
  coverage <- mean(lower_ci <= theta & upper_ci >= theta)
  return(coverage)
}
# coverage bias eliminated
coverage_bias_eliminated <- function(thetaobs, thetaobs_se, ...) {
  lower_ci <- thetaobs - thetaobs_se
  upper_ci <- thetaobs + thetaobs_se
  mean_thetaobs <- mean(thetaobs)
  coverage_bias_eliminated <- mean(lower_ci <= mean_thetaobs & upper_ci >=mean_thetaobs)
  return(coverage_bias_eliminated)
}
# type1error
type1error <- function(pvals) sum(pvals<=0.05)/length(pvals)  # no causal_effect

# GSD
IVW_GSD_coverage_calc <- c()
IVW_GSD_coverage_bias_eliminated_calc <- c()
IVW_GSD_type1error_calc <- c()
MODE_GSD_coverage_calc <- c()
MODE_GSD_coverage_bias_eliminated_calc <- c()
MODE_GSD_type1error_calc <- c()
CONMIX_GSD_coverage_calc <- c()
CONMIX_GSD_coverage_bias_eliminated_calc <- c()
CONMIX_GSD_type1error_calc <- c()
MRMIX_GSD_coverage_calc <- c()
MRMIX_GSD_coverage_bias_eliminated_calc <- c()
MRMIX_GSD_type1error_calc <- c()
# GBC
## deleted 2023-01-29
for (i in 1:length(sample_size_filtered)) {
  sample_size <- sample_size_filtered[i]
  # GSD additional metrics ----------------------------------------------------
  ## 1/ IVW
  col_IVW_GSD_theta <- paste0("IVW_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_IVW_GSD_theta_se <- paste0("IVW_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_IVW_GSD_mr_pval <- paste0("IVW_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  IVW_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_IVW_GSD_theta], thetaobs_se = est_sim[col_IVW_GSD_theta_se], theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "IVW"])
  IVW_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_IVW_GSD_theta]], thetaobs_se = est_sim[[col_IVW_GSD_theta_se]])
  IVW_GSD_type1error_calc[i] <- type1error(est_sim[[col_IVW_GSD_mr_pval]])
  ## 2/ MODE
  col_MODE_GSD_theta <- paste0("MODE_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MODE_GSD_theta_se <- paste0("MODE_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MODE_GSD_mr_pval <- paste0("MODE_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  MODE_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MODE_GSD_theta], 
                                       thetaobs_se = est_sim[col_MODE_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "mode"])
  MODE_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MODE_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_MODE_GSD_theta_se]])
  MODE_GSD_type1error_calc[i] <- type1error(est_sim[[col_MODE_GSD_mr_pval]])
  ## 3/ CONMIX
  col_CONMIX_GSD_theta <- paste0("CONMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_CONMIX_GSD_theta_se <- paste0("CONMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_CONMIX_GSD_mr_pval <- paste0("CONMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  CONMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_CONMIX_GSD_theta], 
                                       thetaobs_se = est_sim[col_CONMIX_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "conmix"])
  CONMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_CONMIX_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_CONMIX_GSD_theta_se]])
  CONMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_CONMIX_GSD_mr_pval]])  
    ## 4/ MRMIX
  col_MRMIX_GSD_theta <- paste0("MRMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MRMIX_GSD_theta_se <- paste0("MRMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
  col_MRMIX_GSD_mr_pval <- paste0("MRMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
  # calc
  MRMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MRMIX_GSD_theta], 
                                       thetaobs_se = est_sim[col_MRMIX_GSD_theta_se], 
                                       theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == "MRMix"])
  MRMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MRMIX_GSD_theta]], 
                                                                       thetaobs_se = est_sim[[col_MRMIX_GSD_theta_se]])
  MRMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_MRMIX_GSD_mr_pval]])  
  
  ## GBC additional metrics ----------------------------------------------------
  # deleted 2023-01-29
}
```


### GSD Performance Metrics

```{r add-additional-metrics-to-df-and-print-GSD}
## GSD --------------------------
IVW_GSD_performance_metrics <- IVW_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(IVW_GSD_coverage_calc, IVW_GSD_coverage_bias_eliminated_calc, IVW_GSD_type1error_calc)) %>% 
  rename(coverage = IVW_GSD_coverage_calc, 
         coverage_bias_eliminated = IVW_GSD_coverage_bias_eliminated_calc, 
         type1error = IVW_GSD_type1error_calc) %>% 
  as_tibble()
IVW_GSD_performance_metrics %>% 
  mutate(across(cols = where(is.numeric), round(3))) %>% 
  kable(caption = "IVW method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)

MODE_GSD_performance_metrics <- MODE_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(MODE_GSD_coverage_calc, 
                   MODE_GSD_coverage_bias_eliminated_calc, MODE_GSD_type1error_calc)) %>% 
  rename(coverage = MODE_GSD_coverage_calc, 
         coverage_bias_eliminated = MODE_GSD_coverage_bias_eliminated_calc, 
         type1error = MODE_GSD_type1error_calc) %>% 
  as_tibble()
MODE_GSD_performance_metrics %>% 
  mutate(across(cols = where(is.numeric), round(3))) %>% 
  kable(caption = "MODE method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)

CONMIX_GSD_performance_metrics <- CONMIX_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(CONMIX_GSD_coverage_calc, 
                   CONMIX_GSD_coverage_bias_eliminated_calc, CONMIX_GSD_type1error_calc)) %>% 
  rename(coverage = CONMIX_GSD_coverage_calc, 
         coverage_bias_eliminated = CONMIX_GSD_coverage_bias_eliminated_calc, 
         type1error = CONMIX_GSD_type1error_calc) %>% 
  as_tibble()
CONMIX_GSD_performance_metrics %>% 
  mutate(across(cols = where(is.numeric), round(3))) %>% 
  kable(caption = "CONMIX method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)

MRMIX_GSD_performance_metrics <- MRMIX_GSD_performance_metrics %>% 
  select(N, everything()) %>% 
  cbind(data.frame(MRMIX_GSD_coverage_calc, 
                   MRMIX_GSD_coverage_bias_eliminated_calc, MRMIX_GSD_type1error_calc)) %>% 
  rename(coverage = MRMIX_GSD_coverage_calc, 
         coverage_bias_eliminated = MRMIX_GSD_coverage_bias_eliminated_calc, 
         type1error = MRMIX_GSD_type1error_calc) %>% 
  as_tibble()
MRMIX_GSD_performance_metrics %>% 
  mutate(across(cols = where(is.numeric), round(3))) %>% 
  kable(caption = "MRMIX method on BMI -> GSD performance metrics") %>% 
  kable_styling(font_size=10)
```



```{r store-performance-metrics-csv}
## GSD
filename_IVW_GSD_performance_metrics <- paste0("../output/performance_metrics/", 
                                               Sys.Date(), "_IVW_GSD_performance_metrics_NSIM_", nsim, ".csv")
filename_MODE_GSD_performance_metrics <- paste0("../output/performance_metrics/", 
                                               Sys.Date(), "_MODE_GSD_performance_metrics_NSIM_", nsim, ".csv")
filename_CONMIX_GSD_performance_metrics <- paste0("../output/performance_metrics/", 
                                               Sys.Date(), "_CONMIX_GSD_performance_metrics_NSIM_", nsim, ".csv")
filename_MRMIX_GSD_performance_metrics <- paste0("../output/performance_metrics/", 
                                               Sys.Date(), "_MRMIX_GSD_performance_metrics_NSIM_", nsim, ".csv")
write.csv2(x = IVW_GSD_performance_metrics, file = filename_IVW_GSD_performance_metrics)
write.csv2(x = MODE_GSD_performance_metrics, file = filename_MODE_GSD_performance_metrics)
write.csv2(x = CONMIX_GSD_performance_metrics, file = filename_CONMIX_GSD_performance_metrics)
write.csv2(x = MRMIX_GSD_performance_metrics, file = filename_MRMIX_GSD_performance_metrics)
```


## Performance Plots

### GSD simulated theta distribution

```{r theta-sim-distribution-GSD-old, eval=FALSE, fig.height=6, fig.width=10, include=FALSE}
### OLD
# # compare distribution of simulated thetas against theta_324000
# boxplot_dat <- est_sim %>% 
#   select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
#   pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
#   mutate(theta_sim_numeric = format(as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")), scientific=FALSE))
# 
# ggplot(boxplot_dat, aes(x=forcats::fct_rev( as.factor(theta_sim_numeric)), y=value)) +
#   geom_boxplot()+
#   geom_jitter(width=0.1,alpha=0.2)+
#   geom_hline(yintercept=  args.gsd.ivw$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
#   annotate(geom = "text", x = "300000", y = args$theta+0.5,
#            label= expression( paste("ref: ", theta[X] ))  , 
#            color= ggthemes_data$few$colors$Dark[9,2][[1]]
#              )+
#   scale_x_discrete(name = paste0("Size of sample ", as.roman(1)),  breaks = c(" 50000", "100000", "150000", "200000", "250000", "300000"), )+ 
#   ylab(expression(paste(theta[sim], " with ", n[sim]== 25 ) )  ) +
#   labs(title = "Estimates from beta simulation",
#        caption = "Relationship between sample size, average number of instrumental variables (IVs) and variance of traits explained by the IVs"
#        )+
#   coord_flip()+
#   theme_few() -> simplot_thetas
# simplot_thetas
```


```{r theta-sim-distribution-GSD, fig.height=6, fig.width=10, warning=FALSE}
## IVW
IVW_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(IVW_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.ivw$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.ivw$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = IVW_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = IVW_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "IVW estimate: BMI -> GSD")+
  theme_few() -> IVW_GSD_simplot_thetas
IVW_GSD_simplot_thetas

## MODE
MODE_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(MODE_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.mode$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.mode$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = MODE_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = MODE_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "MODE estimate: BMI -> GSD")+
  theme_few() -> MODE_GSD_simplot_thetas
MODE_GSD_simplot_thetas

## CONMIX
CONMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.conmix$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.conmix$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = CONMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = CONMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "CONMIX estimate: BMI -> GSD")+
  theme_few() -> CONMIX_GSD_simplot_thetas
CONMIX_GSD_simplot_thetas

## MRMIX
MRMIX_BMI_GSD_geom_point_dat <- est_sim %>% 
  select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
  pivot_longer(cols = everything(), names_to = "theta_sim") %>% 
  mutate(theta_sim_numeric = as.numeric(stringr::str_extract(string =theta_sim, pattern = "\\d+")))

ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=mean))+ 
  geom_point()+
  geom_hline(yintercept=  args.gsd.mrmix$theta, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  annotate(geom = "text", x = 50000, y = args.gsd.mrmix$theta+0.05,
         label= expression( paste("ref: ", theta[X] ))  , 
         color= ggthemes_data$few$colors$Dark[9,2][[1]]
           )+
  geom_errorbar(aes(ymin=mean-emp_se, ymax=mean+emp_se), 
                width=3000, # x-axis dimension
                position = "dodge")+
  ## ADD rainbow plots
  geom_boxplot(data = MRMIX_BMI_GSD_geom_point_dat, aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
               col=ggthemes_data$few$colors$Dark[3,2][[1]],
               alpha=0.2, size=0.2, outlier.shape = NA)+
  ggdist::stat_halfeye(data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric, y=value, group=theta_sim_numeric), 
                       inherit.aes = FALSE, 
                       slab_fill=ggthemes_data$few$colors$Dark[3,2][[1]], slab_alpha = 0.2,
                       geom = "slabinterval",
                       na.rm=TRUE, trim=TRUE, adjust = .5, width = 10000, .width = 0, justification = -.3, point_colour = NA)+
  gghalves::geom_half_point(
    data = MRMIX_BMI_GSD_geom_point_dat, mapping=aes(x=theta_sim_numeric-2000, y=value, group=theta_sim_numeric), 
    inherit.aes = FALSE, size=0.2, colour=ggthemes_data$few$colors$Dark[3,2][[1]],
    side = "l", range_scale = .4, alpha = .2)+
  ## FINISH rainbow
  scale_x_continuous(labels = comma)+
  xlab("Size of Sample I")+
  ylab("Mean thetas")+
  labs(title = "Simulated Thetas with Empirical Standard Error", subtitle = "MRMIX estimate: BMI -> GSD")+
  theme_few() -> MRMIX_GSD_simplot_thetas
MRMIX_GSD_simplot_thetas

# SAVE
filename_IVW_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_IVW_GSD_simplot_thetas_NSIM_", nsim,".png")
ggsave(filename = filename_IVW_GSD_simplot_thetas, plot = IVW_GSD_simplot_thetas, 
       width = 15, height = 6, 
       units = "in"  # default
       )
filename_MODE_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_MODE_GSD_simplot_thetas_NSIM_", nsim,".png")
ggsave(filename = filename_MODE_GSD_simplot_thetas, plot = MODE_GSD_simplot_thetas, 
       width = 15, height = 6, 
       units = "in"  # default
       )
filename_CONMIX_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_CONMIX_GSD_simplot_thetas_NSIM_", nsim,".png")
ggsave(filename = filename_CONMIX_GSD_simplot_thetas, plot = CONMIX_GSD_simplot_thetas, 
       width = 15, height = 6, 
       units = "in"  # default
       )
filename_MRMIX_GSD_simplot_thetas <- paste0("../output/plots/", Sys.Date(), "_CONMIX_GSD_simplot_thetas_NSIM_", nsim,".png")
ggsave(filename = filename_MRMIX_GSD_simplot_thetas, plot = MRMIX_GSD_simplot_thetas, 
       width = 15, height = 6, 
       units = "in"  # default
       )
```
```{r theta-sim-distribution-GSD-arrange, fig.height=18, fig.width=10, warning=FALSE}
IVW_GSD_simplot_thetas$labels$title <- ""
MODE_GSD_simplot_thetas$labels$title <- ""
CONMIX_GSD_simplot_thetas$labels$title <- ""
MRMIX_GSD_simplot_thetas$labels$title <- ""
ggarrange(IVW_GSD_simplot_thetas,MODE_GSD_simplot_thetas,CONMIX_GSD_simplot_thetas, MRMIX_GSD_simplot_thetas,
          ncol = 1, nrow = 4) -> GSD_simplot_thetas_arrange

filename_GSD_simplot_thetas_arrange <- paste0("../output/plots/", Sys.Date(), "_GSD_simplot_thetas_NSIM_", nsim,".png")
ggsave(filename = filename_GSD_simplot_thetas_arrange, plot = GSD_simplot_thetas_arrange, 
       width = 10, height = 22, 
       units = "in"  # default
       )
```


### GSD Bias & MSE

```{r plot-performance-metrics-bias-mse-GSD, fig.height=10, fig.width=8}
## IVW
#BIAS
ggplot(IVW_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "IVW estimate: BMI -> GSD")+
  theme_few() -> biasplot_IVW_GSD
#MSE
ggplot(IVW_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "IVW estimate: BMI -> GSD")+
  theme_few() -> mseplot_IVW_GSD
ggarrange(biasplot_IVW_GSD, mseplot_IVW_GSD) -> bias_mse_plot_IVW_GSD

## MODE
ggplot(MODE_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "MODE estimate: BMI -> GSD")+
  theme_few() -> biasplot_MODE_GSD
ggplot(MODE_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
    geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "MODE estimate: BMI -> GSD")+
  theme_few() -> mseplot_MODE_GSD
ggarrange(biasplot_MODE_GSD, mseplot_MODE_GSD) -> bias_mse_plot_MODE_GSD

## CONMIX
ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "CONMIX estimate: BMI -> GSD")+
  theme_few() -> biasplot_CONMIX_GSD
ggplot(CONMIX_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "CONMIX estimate: BMI -> GSD")+
  theme_few() -> mseplot_CONMIX_GSD
ggarrange(biasplot_CONMIX_GSD, mseplot_CONMIX_GSD) -> bias_mse_plot_CONMIX_GSD

## MRMIX
ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=bias))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(-1.5,1.5))+
  xlab("Size of Sample I")+
  ylab("Bias")+
  labs(title = "MRMIX estimate: BMI -> GSD")+
  theme_few() -> biasplot_MRMIX_GSD
ggplot(MRMIX_GSD_performance_metrics, aes(x=N, y=mse))+ 
  geom_point()+
  geom_hline(yintercept=  0, col=ggthemes_data$few$colors$Dark[9,2][[1]], alpha=0.6, linetype=2, size=0.8)+
  scale_x_continuous(labels = comma)+
  scale_y_continuous(limits = c(0,1.8))+
  xlab("Size of Sample I")+
  ylab("MSE")+
  labs(title = "MRMIX estimate: BMI -> GSD")+
  theme_few() -> mseplot_MRMIX_GSD
ggarrange(biasplot_MRMIX_GSD, mseplot_MRMIX_GSD) -> bias_mse_plot_MRMIX_GSD

ggarrange(bias_mse_plot_IVW_GSD, bias_mse_plot_MODE_GSD, bias_mse_plot_CONMIX_GSD, bias_mse_plot_MRMIX_GSD
          ncol = 1, nrow = 4, common.legend = TRUE) -> bias_mse_plot_GSD
bias_mse_plot_GSD

filename_simplot_bias_mse <- paste0("../output/plots/", Sys.Date(), "_GSD_simplot_bias_mse_NSIM_", nsim, ".png")
ggsave(filename = filename_simplot_bias_mse, plot = bias_mse_plot_GSD, 
       width = 7, height = 12, 
       units = "in"  # default
       )
```


**Deleted all code for GBC simulation downstream analysis since no longer needed for thesis.**


# Parallelize performance estimates for different "TRUE" estimates

We use 5 BMI -> GSD reference estimates: IVW random model, MRMix, weighted median, Wald estimate MC4R variant, Wald estimate FTO variant

```{r}
selected_true_methods <- c("IVW", "MRMix", "median", "MC4R_Wald", "FTO_Wald")
true_estimates <- est.bmi.gsd %>% 
  filter(mr_methods %in% selected_true_methods)
```

```{r parallel-performance}

funs <- c(mean = mean, emp_se=emp_se, bias = bias, mse = mse)  # as above

for (true_method in 1:length(selected_true_methods)) {
  print(true_method)
  
  # define reference estimate for all 5 chosen methods in loop 
  args.gsd.method <- list(theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]], 
                          nsim=nrow(est_sim)
  )
  
  IVW_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_IVW_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, IVW_GSD_mean=mean, IVW_GSD_emp_se=emp_se, IVW_GSD_bias=bias, IVW_GSD_mse=mse)

  MODE_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_MODE_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, MODE_GSD_mean=mean, MODE_GSD_emp_se=emp_se, MODE_GSD_bias=bias, MODE_GSD_mse=mse)
  
  CONMIX_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_CONMIX_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, CONMIX_GSD_mean=mean, CONMIX_GSD_emp_se=emp_se, CONMIX_GSD_bias=bias, CONMIX_GSD_mse=mse)
  
  MRMIX_GSD_performance_metrics <- est_sim %>% 
    select(!!!(colname_MRMIX_theta_sim_BMI_GSD)) %>% 
    map_df(.f = ~ funs %>% 
             map(exec, .x, !!!args.gsd.method), .id = "var") %>% 
    cbind(sample_size_filtered) %>% 
    rename(N = sample_size_filtered, MRMIX_GSD_mean=mean, MRMIX_GSD_emp_se=emp_se, MRMIX_GSD_bias=bias, MRMIX_GSD_mse=mse)
  
  ## further metrics
  IVW_GSD_coverage_calc <- c()
  IVW_GSD_coverage_bias_eliminated_calc <- c()
  IVW_GSD_type1error_calc <- c()
  MODE_GSD_coverage_calc <- c()
  MODE_GSD_coverage_bias_eliminated_calc <- c()
  MODE_GSD_type1error_calc <- c()
  CONMIX_GSD_coverage_calc <- c()
  CONMIX_GSD_coverage_bias_eliminated_calc <- c()
  CONMIX_GSD_type1error_calc <- c()
  MRMIX_GSD_coverage_calc <- c()
  MRMIX_GSD_coverage_bias_eliminated_calc <- c()
  MRMIX_GSD_type1error_calc <- c()

  for (i in 1:length(sample_size_filtered)) {
    sample_size <- sample_size_filtered[i]
    # GSD additional metrics ----------------------------------------------------
    ## 1/ IVW
    col_IVW_GSD_theta <- paste0("IVW_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_IVW_GSD_theta_se <- paste0("IVW_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_IVW_GSD_mr_pval <- paste0("IVW_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    IVW_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_IVW_GSD_theta], thetaobs_se = est_sim[col_IVW_GSD_theta_se], theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    IVW_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_IVW_GSD_theta]], thetaobs_se = est_sim[[col_IVW_GSD_theta_se]])
    IVW_GSD_type1error_calc[i] <- type1error(est_sim[[col_IVW_GSD_mr_pval]])
    ## 2/ MODE
    col_MODE_GSD_theta <- paste0("MODE_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MODE_GSD_theta_se <- paste0("MODE_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MODE_GSD_mr_pval <- paste0("MODE_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    MODE_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MODE_GSD_theta], 
                                         thetaobs_se = est_sim[col_MODE_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    MODE_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MODE_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_MODE_GSD_theta_se]])
    MODE_GSD_type1error_calc[i] <- type1error(est_sim[[col_MODE_GSD_mr_pval]])
    ## 3/ CONMIX
    col_CONMIX_GSD_theta <- paste0("CONMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_CONMIX_GSD_theta_se <- paste0("CONMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_CONMIX_GSD_mr_pval <- paste0("CONMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    CONMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_CONMIX_GSD_theta], 
                                         thetaobs_se = est_sim[col_CONMIX_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    CONMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_CONMIX_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_CONMIX_GSD_theta_se]])
    CONMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_CONMIX_GSD_mr_pval]])  
      ## 4/ MRMIX
    col_MRMIX_GSD_theta <- paste0("MRMIX_theta_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MRMIX_GSD_theta_se <- paste0("MRMIX_theta_se_sim_BMI_GSD_", format(sample_size, scientific = FALSE))
    col_MRMIX_GSD_mr_pval <- paste0("MRMIX_mr_pval_BMI_GSD_", format(sample_size, scientific = FALSE))
    # calc
    MRMIX_GSD_coverage_calc[i] <- coverage(thetaobs = est_sim[col_MRMIX_GSD_theta], 
                                         thetaobs_se = est_sim[col_MRMIX_GSD_theta_se], 
                                         theta = est.bmi.gsd$estimate[est.bmi.gsd$mr_methods == selected_true_methods[true_method]])
    MRMIX_GSD_coverage_bias_eliminated_calc[i] <- coverage_bias_eliminated(thetaobs = est_sim[[col_MRMIX_GSD_theta]], 
                                                                         thetaobs_se = est_sim[[col_MRMIX_GSD_theta_se]])
    MRMIX_GSD_type1error_calc[i] <- type1error(est_sim[[col_MRMIX_GSD_mr_pval]])  
  }
  
  ## combine metrics
  IVW_GSD_performance_metrics <- IVW_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(IVW_GSD_coverage_calc, IVW_GSD_coverage_bias_eliminated_calc, IVW_GSD_type1error_calc)) %>% 
    as_tibble()
  MODE_GSD_performance_metrics <- MODE_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(MODE_GSD_coverage_calc, 
                     MODE_GSD_coverage_bias_eliminated_calc, MODE_GSD_type1error_calc)) %>% 
    as_tibble()
  CONMIX_GSD_performance_metrics <- CONMIX_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(CONMIX_GSD_coverage_calc, 
                     CONMIX_GSD_coverage_bias_eliminated_calc, CONMIX_GSD_type1error_calc)) %>% 
    as_tibble()
  MRMIX_GSD_performance_metrics <- MRMIX_GSD_performance_metrics %>% 
    select(N, everything()) %>% 
    cbind(data.frame(MRMIX_GSD_coverage_calc, 
                     MRMIX_GSD_coverage_bias_eliminated_calc, MRMIX_GSD_type1error_calc)) %>% 
    as_tibble()
  MRMIX_GSD_performance_metrics %>% 
    mutate(across(cols = where(is.numeric), round(3))) %>% 
    kable(caption = "MRMIX method on BMI -> GSD performance metrics") %>% 
    kable_styling(font_size=10)

}



```







